// @hero-design/react/src/components/SelectButton/StyledSelectButton.tsx
import styled, { css } from 'styled-components';

const StyledSelectButton = styled.label`
  display: inline-block;
  margin: 0;
  padding: 0;
`;

const SelectButtonGroupWrapper = styled.div<{
  themeSize: 'small' | 'medium' | 'large';
}>`
  display: inline-flex;
  flex-wrap: wrap;
  padding: 0;

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          margin: ${theme.__hd__.selectButton.space.groupSmallMargin};
          width: ${theme.__hd__.selectButton.sizes.groupSmallWidth};
          > * {
            margin: ${theme.__hd__.selectButton.space.smallMargin};
          }
        `;
      case 'medium':
        return css`
          margin: ${theme.__hd__.selectButton.space.groupMediumMargin};
          width: ${theme.__hd__.selectButton.sizes.groupMediumWidth};
          > * {
            margin: ${theme.__hd__.selectButton.space.mediumMargin};
          }
        `;
      case 'large':
        return css`
          margin: ${theme.__hd__.selectButton.space.groupLargeMargin};
          width: ${theme.__hd__.selectButton.sizes.groupLargeWidth};
          > * {
            margin: ${theme.__hd__.selectButton.space.largeMargin};
          }
        `;
    }
  }};
`;

const StyledInputText = styled.span<{
  themeSize: 'small' | 'medium' | 'large';
}>`
  box-sizing: border-box;
  line-height: 100%;
  display: flex;
  align-items: center;
  cursor: pointer;
  border-style: solid;
  border-color: ${({ theme }) =>
    theme.__hd__.selectButton.colors.defaultBorder};
  border-width: ${({ theme }) =>
    theme.__hd__.selectButton.borderWidths.default};
  border-radius: ${({ theme }) => theme.__hd__.selectButton.radii.default};
  font-weight: ${({ theme }) => theme.__hd__.selectButton.fontWeights.default};
  margin: 0;
  padding: 0;
  color: ${({ theme }) => theme.__hd__.selectButton.colors.defaultText};
  &:hover,
  &:focus {
    color: ${({ theme }) => theme.__hd__.selectButton.colors.hover};
    border-color: ${({ theme }) => theme.__hd__.selectButton.colors.hover};
  }

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          height: ${theme.__hd__.selectButton.sizes.small};
          padding: ${theme.__hd__.selectButton.space.smallPadding};
          font-size: ${theme.__hd__.selectButton.fontSizes.small};
        `;
      case 'medium':
        return css`
          height: ${theme.__hd__.selectButton.sizes.medium};
          padding: ${theme.__hd__.selectButton.space.mediumPadding};
          font-size: ${theme.__hd__.selectButton.fontSizes.medium};
        `;
      case 'large':
        return css`
          height: ${theme.__hd__.selectButton.sizes.large};
          padding: ${theme.__hd__.selectButton.space.largePadding};
          font-size: ${theme.__hd__.selectButton.fontSizes.large};
        `;
    }
  }};
`;

const StyledInput = styled.input`
  opacity: 0;
  width: 0;
  height: 0;
  position: absolute;
  margin: 0;
  padding: 0;

  &:disabled ~ ${StyledInputText} {
    color: ${({ theme }) => theme.__hd__.selectButton.colors.disabledText};
    background-color: ${({ theme }) =>
      theme.__hd__.selectButton.colors.disabledBackground};
    border-color: ${({ theme }) =>
      theme.__hd__.selectButton.colors.defaultBorder};
    pointer-events: none;
  }

  &:disabled:checked ~ ${StyledInputText} {
    color: ${({ theme }) =>
      theme.__hd__.selectButton.colors.disabledSelectedText};
    background-color: ${({ theme }) =>
      theme.__hd__.selectButton.colors.disabledSelectedBackground};
    border: none;
  }

  &:checked + ${StyledInputText} {
    border-color: ${({ theme }) =>
      theme.__hd__.selectButton.colors.selectedBackground};
    color: ${({ theme }) => theme.__hd__.selectButton.colors.selectedText};
    background: ${({ theme }) =>
      theme.__hd__.selectButton.colors.selectedBackground};
  }
`;

export {
  SelectButtonGroupWrapper,
  StyledSelectButton,
  StyledInput,
  StyledInputText,
};
