// @hero-design/react/src/components/Timeline/StyledTimeline.tsx
import styled, { css } from 'styled-components';

const TimelineItemLine = styled.div`
  position: absolute;
  top: ${({ theme }) => theme.__hd__.timeline.space.itemLinePositionTop};
  left: ${({ theme }) => theme.__hd__.timeline.space.itemLinePositionLeft};
  height: ${({ theme }) => theme.__hd__.timeline.sizes.itemLineHeight};
  border-left: ${({ theme }) => theme.__hd__.timeline.sizes.itemLineWidth} solid
    ${({ theme }) => theme.__hd__.timeline.colors.itemLineBg};
  margin: 0;
  padding: 0;
`;

const TimelineItemContentWrapper = styled.div`
  position: relative;
  margin: 0;
  padding: 0;
  margin-left: ${({ theme }) =>
    theme.__hd__.timeline.space.itemContentMarginLeft};
`;

const TimelineItemWrapper = styled.li`
  position: relative;
  margin: 0;
  padding: 0;
  padding-bottom: ${({ theme }) =>
    theme.__hd__.timeline.space.itemPaddingBottom};
  list-style: none;

  &:last-child {
    padding-bottom: 0;

    ${TimelineItemLine} {
      display: none;
    }
  }
`;

const TimelineWrapper = styled.ul`
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
`;

const TimelineItemMarker = styled.div<{
  themeIntent: 'success' | 'info' | 'warning' | 'danger' | 'error' | 'default';
}>`
  position: absolute;
  top: ${({ theme }) => theme.__hd__.timeline.space.itemMarkerPositionTop};
  width: ${({ theme }) => theme.__hd__.timeline.sizes.itemMarkerWidth};
  height: ${({ theme }) => theme.__hd__.timeline.sizes.itemMarkerHeight};
  background-color: ${({ theme }) => theme.__hd__.timeline.colors.itemMarkerBg};
  border-width: ${({ theme }) => theme.__hd__.timeline.borderWidths.itemMarker};
  border-style: solid;
  border-radius: 50%;
  box-sizing: border-box;
  margin: 0;
  padding: 0;

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'default':
        return css`
          border-color: ${theme.__hd__.timeline.colors.itemMarkerDefault};
        `;
      case 'success':
        return css`
          border-color: ${theme.__hd__.timeline.colors.itemMarkerSuccess};
        `;
      case 'info':
        return css`
          border-color: ${theme.__hd__.timeline.colors.itemMarkerInfo};
        `;
      case 'warning':
        return css`
          border-color: ${theme.__hd__.timeline.colors.itemMarkerWarning};
        `;
      case 'danger':
        return css`
          border-color: ${theme.__hd__.timeline.colors.itemMarkerDanger};
        `;
      case 'error':
        return css`
          border-color: ${theme.__hd__.timeline.colors.itemMarkerError};
        `;
    }
  }};
`;

const TimelineItemText = styled.div`
  font-size: ${({ theme }) => theme.__hd__.timeline.fontSizes.itemText};
  line-height: ${({ theme }) => theme.__hd__.timeline.lineHeights.itemText};
  font-weight: ${({ theme }) => theme.__hd__.timeline.fontWeights.itemText};
  color: ${({ theme }) => theme.__hd__.timeline.colors.itemText};
  margin: 0;
  padding: 0;
`;

const TimelineItemHelpText = styled.div`
  font-size: ${({ theme }) => theme.__hd__.timeline.fontSizes.itemHelpText};
  line-height: ${({ theme }) => theme.__hd__.timeline.lineHeights.itemHelpText};
  font-weight: ${({ theme }) => theme.__hd__.timeline.fontWeights.itemHelpText};
  color: ${({ theme }) => theme.__hd__.timeline.colors.itemHelpText};
  margin: 0;
  padding: 0;
`;

export {
  TimelineWrapper,
  TimelineItemWrapper,
  TimelineItemText,
  TimelineItemHelpText,
  TimelineItemContentWrapper,
  TimelineItemMarker,
  TimelineItemLine,
};
