// @hero-design/react/src/components/Form/Field/StyledField.tsx
import styled, { css } from 'styled-components';
import StyledInput, { InputWrapper } from '../../Input/StyledInput';
import StyledTextArea, { HiddenTextArea } from '../../Input/StyledTextArea';
import type { Size } from './types';

const FieldWrapper = styled.div<{
  themeStatus?: 'success' | 'warning' | 'error' | 'validating';
}>`
  margin: 0;
  padding: 0;

  ${({ themeStatus, theme }) => {
    switch (themeStatus) {
      case 'warning':
        return css`
          ${InputWrapper} {
            border-color: ${theme.__hd__.form.colors.warningBorder};
            &:focus-within {
              border-color: ${theme.__hd__.form.colors.warningBorder};
              box-shadow: ${theme.__hd__.form.shadows.warningFocus};
            }
          }

          ${StyledTextArea} {
            border-color: ${theme.__hd__.form.colors.warningBorder};
            &:focus-within {
              border-color: ${theme.__hd__.form.colors.warningBorder};
              box-shadow: ${theme.__hd__.form.shadows.warningFocus};
            }
          }
        `;
      case 'error':
        return css`
          ${InputWrapper} {
            border-color: ${theme.__hd__.form.colors.errorBorder};
            &:focus-within {
              border-color: ${theme.__hd__.form.colors.errorBorder};
              box-shadow: ${theme.__hd__.form.shadows.errorFocus};
            }
          }

          ${StyledInput} {
            color: ${theme.__hd__.form.colors.errorText};
          }

          ${StyledTextArea} {
            border-color: ${theme.__hd__.form.colors.errorBorder};
            &:focus-within {
              border-color: ${theme.__hd__.form.colors.errorBorder};
              box-shadow: ${theme.__hd__.form.shadows.errorFocus};
            }
          }

          ${StyledTextArea} {
            color: ${theme.__hd__.form.colors.errorText};
          }
        `;
      case 'success':
      case 'validating':
      case undefined:
        return css``;
    }
  }};
`;

const ValidateIconWrapper = styled.div<{
  themeSize: Size;
}>`
  margin: 0;
  padding: 0;
  display: flex;

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          padding-left: ${theme.__hd__.form.space.smallIconPadding};
        `;
      case 'medium':
        return css`
          padding-left: ${theme.__hd__.form.space.mediumIconPadding};
        `;
      case 'large':
        return css`
          padding-left: ${theme.__hd__.form.space.largeIconPadding};
        `;
    }
  }};
`;

const ValidateSuffixWrapper = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
`;

const TextAreaWrapper = styled.div<{ themeSize: Size }>`
  margin: 0;
  padding: 0px;
  position: relative;

  ${ValidateIconWrapper} {
    position: absolute;

    ${({ themeSize, theme }) => {
      switch (themeSize) {
        case 'small':
          return css`
            top: calc(
              ${theme.__hd__.form.space.smallIconPadding} +
                ${theme.__hd__.input.borderWidths.default}
            );
            right: calc(
              ${theme.__hd__.form.space.smallIconPadding} +
                ${theme.__hd__.input.borderWidths.default}
            );
            font-size: ${theme.__hd__.input.fontSizes.small};
          `;
        case 'medium':
          return css`
            top: calc(
              ${theme.__hd__.form.space.mediumIconPadding} +
                ${theme.__hd__.input.borderWidths.default}
            );
            right: calc(
              ${theme.__hd__.form.space.mediumIconPadding} +
                ${theme.__hd__.input.borderWidths.default}
            );
            font-size: ${theme.__hd__.input.fontSizes.medium};
          `;
        case 'large':
          return css`
            top: calc(
              ${theme.__hd__.form.space.largeIconPadding} +
                ${theme.__hd__.input.borderWidths.default}
            );
            right: calc(
              ${theme.__hd__.form.space.largeIconPadding} +
                ${theme.__hd__.input.borderWidths.default}
            );
            font-size: ${theme.__hd__.input.fontSizes.large};
          `;
      }
    }};
  }

  ${StyledTextArea}, ${HiddenTextArea} {
    ${({ themeSize, theme }) => {
      switch (themeSize) {
        case 'small':
          return css`
            padding-right: calc(
              ${theme.__hd__.form.space.smallIconPadding} * 2 +
                ${theme.__hd__.input.fontSizes.small}
            );
          `;
        case 'medium':
          return css`
            padding-right: calc(
              ${theme.__hd__.form.space.mediumIconPadding} * 2 +
                ${theme.__hd__.input.fontSizes.medium}
            );
          `;
        case 'large':
          return css`
            padding-right: calc(
              ${theme.__hd__.form.space.largeIconPadding} * 2 +
                ${theme.__hd__.input.fontSizes.large}
            );
          `;
      }
    }};
  }
`;

export {
  FieldWrapper,
  ValidateIconWrapper,
  ValidateSuffixWrapper,
  TextAreaWrapper,
};
