// @hero-design/react/src/components/Notification/StyledNotification.tsx
import styled from 'styled-components';

const StyledIconWrapper = styled.div`
  display: flex;
  align-items: center;

  margin: ${({ theme }) => theme.__hd__.notification.space.iconMarginTop}
    ${({ theme }) => theme.__hd__.notification.space.iconMarginRight} 0 0;
  padding: 0;
  line-height: ${({ theme }) => theme.__hd__.notification.lineHeights.icon};
`;

const StyledInner = styled.div`
  flex: 1;
  margin: 0;
  padding: 0;
`;

const StyledTitle = styled.div`
  color: ${({ theme }) => theme.__hd__.notification.colors.title};
  font-size: ${({ theme }) => theme.__hd__.notification.fontSizes.title};
  font-weight: ${({ theme }) => theme.__hd__.notification.fontWeights.title};
  line-height: ${({ theme }) => theme.__hd__.notification.lineHeights.title};
  margin: 0;
  padding: 0;
  margin-bottom: ${({ theme }) =>
    theme.__hd__.notification.space.titleMarginBottom};
`;

const StyledContent = styled.div`
  color: ${({ theme }) => theme.__hd__.notification.colors.content};
  font-size: ${({ theme }) => theme.__hd__.notification.fontSizes.content};
  font-weight: ${({ theme }) => theme.__hd__.notification.fontWeights.content};
  line-height: ${({ theme }) => theme.__hd__.notification.lineHeights.content};
  margin: 0;
  padding: 0;
`;

const StyledNotification = styled.div`
  display: flex;
  align-items: flex-start;
  margin: 0;

  background-color: ${({ theme }) => theme.__hd__.notification.colors.bg};
  border-radius: ${({ theme }) => theme.__hd__.notification.radii.default};
  padding: ${({ theme }) => theme.__hd__.notification.space.padding};
  box-shadow: ${({ theme }) => theme.__hd__.notification.shadows.default};
`;

const StyledCloseButton = styled.div`
  padding: 0;

  margin: ${({ theme }) => theme.__hd__.notification.space.iconMarginTop} 0 0
    auto;
  line-height: ${({ theme }) => theme.__hd__.notification.lineHeights.icon};
`;

export {
  StyledTitle,
  StyledInner,
  StyledContent,
  StyledIconWrapper,
  StyledCloseButton,
};

export default StyledNotification;
