// @hero-design/react/src/components/Slider/StyledSlider.tsx
import styled, { css } from 'styled-components';
import type {
  FlattenInterpolation,
  ThemeProps,
  DefaultTheme,
} from 'styled-components';

export const CommonSliderTrack = css`
  box-sizing: border-box;
  width: 100%;
  height: ${({ theme }) => theme.__hd__.slider.sizes.trackHeight};
  border-radius: ${({ theme }) => theme.__hd__.slider.radii.default};
`;

export const CommonSliderThumb = css`
  box-sizing: border-box;
  border: ${({ theme }) => theme.__hd__.slider.borderWidths.thumb} solid
    ${({ theme }) => theme.__hd__.slider.colors.thumbBorder};
  height: ${({ theme }) => theme.__hd__.slider.sizes.thumb};
  width: ${({ theme }) => theme.__hd__.slider.sizes.thumb};
  border-radius: 50%;
  background: ${({ theme }) => theme.__hd__.slider.colors.thumbBackground};
  cursor: pointer;
`;

export const CommonThumbBoxShadow = css`
  box-shadow: ${({ theme }) => theme.__hd__.slider.shadows.thumb};
`;

export const CommonTooltipAnimation = css`
  transition: all 0.2s ease-in-out;
`;

export const SliderRail = styled.div`
  background: ${({ theme }) => theme.__hd__.slider.colors.background};
  border-radius: ${({ theme }) => theme.__hd__.slider.radii.default};
  height: ${({ theme }) => theme.__hd__.slider.sizes.trackHeight};
  width: 100%;
  position: absolute;
  left: 0px;
  top: ${({ theme }) => theme.__hd__.slider.space.trackOffsetTop};
  margin: 0;
  padding: 0;
`;

export const SliderTrack = styled.div`
  background: ${({ theme }) => theme.__hd__.slider.colors.track};
  border-radius: ${({ theme }) => theme.__hd__.slider.radii.default};
  height: ${({ theme }) => theme.__hd__.slider.sizes.trackHeight};

  position: absolute;
  left: 0px;
  top: ${({ theme }) => theme.__hd__.slider.space.trackOffsetTop};
  margin: 0;
  padding: 0;
`;

export const SliderTooltip = styled.div<{ alwaysShowTooltip: boolean }>`
  display: ${(props): string => (props.alwaysShowTooltip ? 'initial' : 'none')};
  position: absolute;
  margin: 0;
  padding: 0;

  line-height: ${({ theme }) => theme.__hd__.tooltip.lineHeights.default};
  min-width: ${({ theme }) => theme.__hd__.slider.sizes.tooltipMinWidth};

  background: ${({ theme }) => theme.__hd__.tooltip.colors.dark.background};
  color: white;
  border-radius: ${({ theme }) => theme.__hd__.tooltip.radii.default};

  white-space: nowrap;
  font-size: ${({ theme }) => theme.__hd__.tooltip.fontSizes.default};
  text-align: center;

  padding: ${({ theme }) => theme.__hd__.tooltip.space.wrapperPadding};

  user-select: none;
  transform: translate(-50%, -150%);
  left: ${({ theme }) => theme.__hd__.slider.space.tooltipOffsetLeft};
  ${CommonTooltipAnimation};

  &:before {
    content: '';
    position: absolute;
    top: ${({ theme }) => theme.__hd__.slider.space.tooltipPointerOffsetTop};
    left: 50%;
    border: ${({ theme }) => theme.__hd__.tooltip.sizes.arrow} solid
      ${({ theme }) => theme.__hd__.tooltip.colors.dark.background};
    border-bottom: none;
    border-left-color: transparent;
    border-right-color: transparent;
    transform: translateX(-50%);
  }
`;

export const SliderThumb = styled.div`
  box-sizing: border-box;
  border-radius: 50%;
  background: ${({ theme }) => theme.__hd__.slider.colors.thumbBackground};
  height: ${({ theme }) => theme.__hd__.slider.sizes.thumb};
  width: ${({ theme }) => theme.__hd__.slider.sizes.thumb};
  border: ${({ theme }) => theme.__hd__.slider.borderWidths.thumb} solid
    ${({ theme }) => theme.__hd__.slider.colors.thumbBorder};
  outline: none;
  box-shadow: none;

  position: absolute;
  transform: translateX(-50%);
  margin: 0;
  padding: 0;

  &:focus,
  &:hover {
    ${CommonThumbBoxShadow};
    ${SliderTooltip} {
      display: initial;
    }
  }
`;

const DisabledSlider = css`
  pointer-events: none;
  ${SliderRail},
  ${SliderTrack},
  ${SliderThumb} {
    cursor: not-allowed;
  }

  ${SliderTrack} {
    background: ${({ theme }) => theme.__hd__.slider.colors.disabled};
  }

  ${SliderThumb} {
    border-color: ${({ theme }) => theme.__hd__.slider.colors.disabled};
  }
`;

const EnabledSlider = css`
  pointer-events: initial;

  ${SliderRail},
  ${SliderTrack},
  ${SliderThumb} {
    cursor: pointer;
  }
`;

export const SliderContainer = styled.div<{ disabled: boolean }>`
  height: ${({ theme }) => theme.__hd__.slider.sizes.inputHeight};
  width: 100%;
  position: relative;
  padding: 0;
  margin: 0;

  ${(props): FlattenInterpolation<ThemeProps<DefaultTheme>> =>
    props.disabled ? DisabledSlider : EnabledSlider};
`;
