// @hero-design/react/src/theme/global/colors/hero.ts
import palette from '@hero-design/colors';
import type { SystemPalette } from './types';

const heroSystemPalette: SystemPalette = {
  primary: palette.violet,
  lightPrimary: palette.purple,
  hoverPrimary: palette.violetLight30,
  activePrimary: palette.violetDark30,
  primaryBackground: palette.violetLight75,
  primaryLightBackground: palette.violetLight90,
  primaryNeutralBackground: palette.violetDark30,
  primaryDarkBackground: palette.violetDark45,
  info: palette.dodgerBlue,
  infoBorder: palette.blue,
  infoDarkBackground: palette.blueLight75,
  infoBackground: palette.blueLight90,
  danger: palette.red,
  dangerDark: palette.redDark15,
  hoverDanger: palette.redLight30,
  activeDanger: palette.redDark30,
  dangerDarkBackground: palette.redLight75,
  dangerBackground: palette.redLight90,
  warning: palette.orange,
  warningDark: palette.orangeDark15,
  hoverWarning: palette.orangeLight30,
  activeWarning: palette.orangeDark30,
  warningDarkBackground: palette.orangeLight75,
  warningBackground: palette.orangeLight90,
  success: palette.green,
  successDark: palette.greenDark15,
  hoverSuccess: palette.greenLight30,
  activeSuccess: palette.greenDark30,
  successDarkBackground: palette.grotesqueGreenLight60,
  successBackground: palette.grotesqueGreenLight90,
  error: palette.pink,
  errorDark: palette.pinkDark15,
  hoverError: palette.pinkLight30,
  activeError: palette.pinkDark30,
  errorDarkBackground: palette.pinkLight75,
  errorBackground: palette.pinkLight90,
  link: palette.violet,
  hoverLink: palette.violetLight30,
  activeLink: palette.violetDark30,
  text: palette.black,
  bodyText: palette.greyDark45,
  subduedText: palette.greyDark30,
  disabledText: palette.grey,
  lightText: palette.white,
  defaultBackground: palette.greyLight90,
  defaultLightBackground: palette.white,
  defaultDarkBackground: palette.black,
  defaultGreyBackground: palette.greyLight75,
  defaultDarkGreyBackground: palette.greyLight45,
  neutralBackground: palette.grey,
  neutralLightBackground: palette.greyLight60,
  neutralDarkBackground: palette.greyDark30,
  defaultBorder: palette.greyLight60,
  black: palette.black,
  primaryHeaderBackground: palette.smaltLight90,
  primaryRingBackground: palette.smalt,
  infoHeaderBackground: palette.violetLight60,
};

export default heroSystemPalette;
