// @hero-design/react/src/theme/global/colors/types.ts
import palette from '@hero-design/colors';

export type Palette = typeof palette;

export type SystemPalette = {
  primary: string;
  lightPrimary: string;
  hoverPrimary: string;
  activePrimary: string;
  primaryBackground: string;
  primaryLightBackground: string;
  primaryNeutralBackground: string;
  primaryDarkBackground: string;
  info: string;
  infoBorder: string;
  infoDarkBackground: string;
  infoBackground: string;
  danger: string;
  dangerDark: string;
  hoverDanger: string;
  activeDanger: string;
  dangerDarkBackground: string;
  dangerBackground: string;
  warning: string;
  warningDark: string;
  hoverWarning: string;
  activeWarning: string;
  warningDarkBackground: string;
  warningBackground: string;
  success: string;
  successDark: string;
  hoverSuccess: string;
  activeSuccess: string;
  successDarkBackground: string;
  successBackground: string;
  error: string;
  errorDark: string;
  hoverError: string;
  activeError: string;
  errorDarkBackground: string;
  errorBackground: string;
  link: string;
  hoverLink: string;
  activeLink: string;
  text: string;
  bodyText: string;
  subduedText: string;
  disabledText: string;
  lightText: string;
  defaultBackground: string;
  defaultLightBackground: string;
  defaultDarkBackground: string;
  defaultGreyBackground: string;
  defaultDarkGreyBackground: string;
  neutralBackground: string;
  neutralLightBackground: string;
  neutralDarkBackground: string;
  defaultBorder: string;
  black: string;
  primaryHeaderBackground: string;
  primaryRingBackground: string;
  infoHeaderBackground: string;
};
