// @hero-design/react/src/components/Chart/HorizontalBar/index.tsx
import React, {
  useRef,
  useEffect,
  useCallback,
  useState,
  useMemo,
} from 'react';
import * as d3 from 'd3';
import { useTheme } from 'styled-components';
import { CommonProps } from '../../common';
import {
  LegendConfig,
  NavigationConfig,
  Series,
  TickConfig,
  XAxisConfig,
  YAxisConfig,
  BarConfig,
  MatchedSegment,
  StyleConfig,
} from '../type';
import {
  StyledBarChartContainer,
  StyledLegendContainer,
} from './StyledHorizontalBarChart';

import { useCss, useResizeObserver } from '../../../utils/hooks';
import Typography from '../../Typography';
import {
  difference,
  filterElementHasNoColor,
  getParentElement,
  usePrevious,
} from '../utils';
import YAxis from './ChartContent/YAxis';
import XAxis from './ChartContent/XAxis';
import YAxisTitle from './ChartContent/YAxisTitle';
import LegendGroup from './ChartContent/LegendGroup';
import BarArea from './ChartContent/BarArea';
import GridArea from './ChartContent/GridArea';

import { useColorScale } from '../common/colorScale';
import Navigation from './ChartContent/Navigation';
import BarTooltip from './ChartContent/BarTooltip';

type DataValue = number | undefined;
export interface HorizontalBarChartProps extends CommonProps {
  /**
   * Data of the chart.
   */
  data: Array<Series<Array<DataValue>>>;
  /**
   * X axis config.
   */
  xAxisConfig?: Omit<XAxisConfig, 'labels'> & { tick?: TickConfig };
  /**
   * Y axis config.
   */
  yAxisConfig?: Omit<YAxisConfig, 'maxValue'>;
  /**
   * Legend config.
   */
  legendConfig?: LegendConfig;
  /**
   * Bar section config.
   */
  barConfig?: Omit<BarConfig<DataValue>, 'showTotal'>;
  /**
   * Navigation button config.
   */
  navigationConfig?: NavigationConfig;
  /**
   * Matched pattern to highlight the bars.
   */
  highlightedItems?: MatchedSegment;
  /**
   * * styleConfig use to custom the style of the chart.
   * * styleConfig must be an object:
   *    * color?: use to custom the legend colors.
   */
  styleConfig?: StyleConfig;
}

type ElementSize = {
  height: number;
  width: number;
};

const NegativeBarChart = ({
  data,
  xAxisConfig,
  yAxisConfig,
  legendConfig,
  navigationConfig = { visible: false },
  barConfig,
  'data-test-id': dataTestId,
  style,
  sx,
  id,
  className,
  highlightedItems,
  styleConfig,
}: HorizontalBarChartProps) => {
  const defaultMaxValue = useMemo(
    () => xAxisConfig?.maxValue || 0,
    [xAxisConfig?.maxValue]
  );

  const isNegativeBarChart = useMemo(
    () => data.some((d) => d.data.some((v) => v !== undefined && v < 0)),
    [data]
  );

  const minValue = useMemo(
    () => (isNegativeBarChart ? -defaultMaxValue : 0),
    [defaultMaxValue, isNegativeBarChart]
  );

  const defaultSize = barConfig?.size || 'large';

  const theme = useTheme();

  const {
    __hd__: { horizontalBarChart: horizontalBarChartTheme },
  } = theme;

  const {
    horizontalBarChartMarginBottom,
    horizontalBarChartMarginLeft,
    horizontalBarChartMarginRight,
    horizontalBarChartMarginTop,
    horizontalBarChartMarginLeftWithNavigation,
    gridChartMarginTop,
  } = horizontalBarChartTheme.space;

  const chartMargin = {
    top: horizontalBarChartMarginTop,
    right: horizontalBarChartMarginRight,
    bottom: horizontalBarChartMarginBottom,
    left: yAxisConfig?.title
      ? horizontalBarChartMarginLeft
      : horizontalBarChartMarginLeftWithNavigation,
  };

  const containerRef = useRef<HTMLDivElement>(null);
  const tooltipRef = useRef<HTMLDivElement>(null);

  const legendContainerRef = useRef<HTMLDivElement>(null);

  const [containerSize, setContainerSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });

  const [legendContainerSize, setLegendContainerSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });

  const [yAxisSize, setYAxisSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });
  const [yAxisTitleSize, setYAxisTitleSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });

  const [currentSelectedBar, setCurrentSelectedBar] = useState<
    | {
        legend: string;
        value: number;
        legendColor: string;
      }
    | undefined
  >(undefined);

  const legends = useMemo(() => data.map((item) => item.label), [data]);

  const [internalSelectedLegends, setInternalSelectedLegends] = useState<
    string[]
  >([]);

  const defaultLegends = useMemo(
    () =>
      legendConfig?.selectable
        ? legendConfig?.value || internalSelectedLegends
        : internalSelectedLegends,
    [internalSelectedLegends, legendConfig]
  );

  const selectedLegends = useMemo(
    () => legends.filter((l) => defaultLegends.includes(l)),
    [legends, defaultLegends]
  );

  const svgSizes: ElementSize = useMemo(
    () => ({
      width: containerSize.width,
      height: containerSize.height - legendContainerSize.height,
    }),
    [containerSize.height, containerSize.width, legendContainerSize.height]
  );

  const yAxisWidth = useMemo(
    () => Math.floor(yAxisSize.width),
    [yAxisSize.width]
  );

  const yAxisTitleWidth = useMemo(
    () => Math.floor(yAxisTitleSize.width),
    [yAxisTitleSize.width]
  );

  const boundsHeight = svgSizes.height - chartMargin.top - chartMargin.bottom;
  const boundsWidth =
    svgSizes.width -
    chartMargin.right -
    chartMargin.left -
    yAxisWidth -
    yAxisTitleWidth;

  const onResize = useCallback(() => {
    if (containerRef.current) {
      const { height, width } = containerRef.current.getBoundingClientRect();
      setContainerSize({ height, width });
    }

    if (legendContainerRef.current) {
      const { height, width } =
        legendContainerRef.current.getBoundingClientRect();
      setLegendContainerSize({ height, width });
    }
  }, []);

  useResizeObserver(onResize, getParentElement(containerRef.current));

  const prevLegends = usePrevious(legends);

  const legendsDiff = useMemo(
    () => difference(legends, prevLegends || []),
    [legends, prevLegends]
  );

  useEffect(() => {
    setInternalSelectedLegends((prev) => {
      return Array.from(new Set([...prev, ...legendsDiff]));
    });
  }, [legendsDiff]);

  const customColors = useMemo(
    () =>
      styleConfig?.series !== undefined
        ? filterElementHasNoColor(styleConfig.series)
        : undefined,
    [styleConfig?.series]
  );

  // Color Scale
  const colorScale = useColorScale(legends, customColors);

  const onMouseOverBar = useCallback((_e, barData) => {
    const { color, legend, value } = barData;
    setCurrentSelectedBar({
      legend,
      value,
      legendColor: color,
    });
  }, []);

  const onMouseOutBar = useCallback(() => {
    setCurrentSelectedBar(undefined);
  }, []);

  const onMouseMoveBar = useCallback((e) => {
    const tooltipElement = d3.select(tooltipRef.current);
    const { clientX, clientY } = e;

    tooltipElement
      .style('top', `${clientY - 12}px`)
      .style('left', `${clientX}px`);
  }, []);

  const barAreaData = useMemo(
    () =>
      data.map((d) => {
        return {
          ...d,
          color: colorScale(d.label),
        };
      }),
    [colorScale, data]
  );

  return (
    <StyledBarChartContainer
      ref={containerRef}
      data-test-id={dataTestId}
      style={{ ...style, ...useCss({ ...sx }) }}
      id={id}
      className={className}
    >
      {navigationConfig.visible && (
        <Navigation
          height={boundsHeight}
          yAxisTitleWidth={yAxisTitleWidth}
          navigationConfig={navigationConfig}
          yAxisConfig={yAxisConfig}
          data-test-id={dataTestId}
          disabledButtons={navigationConfig.disabledButtons}
        />
      )}
      <BarTooltip
        show={!!currentSelectedBar}
        data-test-id={dataTestId}
        ref={tooltipRef}
        color={currentSelectedBar?.legendColor}
        title={currentSelectedBar?.legend}
        value={currentSelectedBar?.value}
      />
      <svg width={svgSizes.width} height={svgSizes.height}>
        <g
          width={boundsWidth}
          height={boundsHeight}
          transform={`translate(${yAxisTitleWidth + chartMargin.left},${
            chartMargin.top + gridChartMarginTop
          })`}
        >
          <YAxisTitle
            height={boundsHeight}
            setSize={setYAxisTitleSize}
            yAxisWidth={yAxisWidth}
            yAxisConfig={yAxisConfig}
          />
        </g>
        <g
          width={boundsWidth}
          height={boundsHeight}
          transform={`translate(${yAxisTitleWidth + chartMargin.left},${
            chartMargin.top
          })`}
        >
          {typeof xAxisConfig?.tick?.interval !== 'undefined' && (
            <GridArea
              height={boundsHeight}
              width={boundsWidth}
              minValue={minValue}
              maxValue={defaultMaxValue}
              xAxisConfig={xAxisConfig}
              highlightMiddleLine={isNegativeBarChart}
            />
          )}
          <BarArea
            height={boundsHeight}
            width={boundsWidth}
            maxValue={defaultMaxValue}
            minValue={minValue}
            data-test-id={dataTestId}
            yAxisConfig={yAxisConfig}
            barConfig={barConfig}
            onMouseOver={onMouseOverBar}
            onMouseOut={onMouseOutBar}
            onMouseMove={onMouseMoveBar}
            highlightedItems={highlightedItems}
            selectedLegends={selectedLegends}
            data={barAreaData}
            isNegative={isNegativeBarChart}
            barSize={defaultSize}
          />
          <XAxis
            height={boundsHeight}
            width={boundsWidth}
            maxValue={defaultMaxValue}
            minValue={minValue}
            xAxisConfig={xAxisConfig}
          />
          <YAxis
            dataLength={data.length}
            height={boundsHeight}
            getAxisSize={setYAxisSize}
            yAxisConfig={yAxisConfig}
            barSize={defaultSize}
          />
        </g>
      </svg>

      <StyledLegendContainer ref={legendContainerRef}>
        {!!xAxisConfig?.title && (
          <Typography.Text fontSize={12} fontWeight="regular" intent="subdued">
            {xAxisConfig.title}
          </Typography.Text>
        )}
        <LegendGroup
          legends={legends}
          onChange={setInternalSelectedLegends}
          selectedLegends={selectedLegends}
          data-test-id={dataTestId ? `${dataTestId}-legend` : undefined}
          legendConfig={legendConfig}
          colorScale={colorScale}
        />
      </StyledLegendContainer>
    </StyledBarChartContainer>
  );
};

export default NegativeBarChart;
