// @hero-design/react/src/components/Chart/HorizontalBar/StyledHorizontalBarChart.tsx
import styled from 'styled-components';
import Box from '../../Box';
import { Arrow, TooltipWrapper } from '../../Tooltip/StyledTooltip';
import IconButton from '../../Button/IconButton';

const StyledBarChartContainer = styled.div`
  position: relative;
  height: 100%;
  width: 100%;

  svg {
    position: relative;
    z-index: 2;
  }
  .x-axis {
    text {
      font-weight: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.fontWeights.xAxisTick};
      font-size: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.fontSizes.xAxisTick};
      line-height: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.lineHeights.xAxisTick};
      color: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.colors.xAxisTickColor};
    }
  }
  .y-axis {
    text {
      font-weight: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.fontWeights.yAxisTick};
      font-size: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.fontSizes.yAxisTick};
      line-height: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.lineHeights.yAxisTick};
      color: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.colors.yAxisTickColor};
    }
  }

  .vertical-title {
    font-size: ${({ theme }) =>
      theme.__hd__.horizontalBarChart.fontSizes.verticalTitle};
    font-weight: ${({ theme }) =>
      theme.__hd__.horizontalBarChart.fontWeights.verticalTitle};
    fill: ${({ theme }) =>
      theme.__hd__.horizontalBarChart.colors.verticalTitleFillColor};
  }
`;

const StyledLegendContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  > *:not(:last-child) {
    margin-bottom: ${({ theme }) =>
      `${theme.__hd__.horizontalBarChart.space.legendContainerChildMarginBottom}px`};
  }
  ${TooltipWrapper} {
    background: ${({ theme }) =>
      theme.__hd__.horizontalBarChart.colors.tooltipBackgroundColor};
    box-shadow: ${({ theme }) =>
      theme.__hd__.horizontalBarChart.boxShadows.tooltip};
    padding: ${({ theme }) =>
      theme.__hd__.horizontalBarChart.space.moreTooltipPadding};
  }

  ${Arrow} {
    &::before {
      background: ${({ theme }) =>
        theme.__hd__.horizontalBarChart.colors.tooltipBackgroundColor};
    }
  }
`;
const StyledTooltipContentWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  & > *:not(:last-child) {
    margin-right: ${({ theme }) =>
      `${theme.__hd__.horizontalBarChart.space.tooltipChildMargin}px`};
  }
`;

const StyledNavigateButton = styled(IconButton)`
  width: ${({ theme }) =>
    `${theme.__hd__.horizontalBarChart.sizes.navigationButton}px`};
  height: ${({ theme }) =>
    `${theme.__hd__.horizontalBarChart.sizes.navigationButton}px`};
`;

const StyledNavigateButtonWrapper = styled(Box)<{
  themeTop: string | number;
  themePosition: 'right' | 'left';
  themeIsHasTitle: boolean;
  themeLeft?: number;
}>`
  position: absolute;
  display: grid;
  place-items: center;
  top: ${({ themeTop }) =>
    typeof themeTop === 'number' ? `${themeTop}px` : themeTop};
  ${({ themePosition, theme, themeIsHasTitle, themeLeft }) => {
    const positionMapping = {
      left: `left: ${
        themeIsHasTitle
          ? `${
              theme.__hd__.horizontalBarChart.space
                .navigateLeftPositionHasTitle + (themeLeft || 0)
            }px`
          : `${
              theme.__hd__.horizontalBarChart.space.navigateLeftPosition +
              (themeLeft || 0)
            }px`
      }`,
      right: `right: -${theme.__hd__.horizontalBarChart.space.navigateRightPosition}px`,
    };

    return positionMapping[themePosition];
  }};
  z-index: 3;
`;

export {
  StyledBarChartContainer,
  StyledLegendContainer,
  StyledTooltipContentWrapper,
  StyledNavigateButton,
  StyledNavigateButtonWrapper,
};
