// @hero-design/react/src/components/Checkbox/StyledCheckbox.tsx
import styled, { css } from 'styled-components';

const Label = styled.label`
  display: inline-flex;
  position: relative;
  margin: 0;
  padding: 0;
  vertical-align: middle;
`;

const LabelText = styled.span`
  margin: 0;
  padding: 0;
  padding-left: ${({ theme }) => theme.__hd__.checkbox.space.labelPaddingLeft};
  font-size: ${({ theme }) => theme.__hd__.checkbox.fontSizes.default};
  font-weight: ${({ theme }) => theme.__hd__.checkbox.fontWeights.default};
  line-height: ${({ theme }) => theme.__hd__.checkbox.lineHeights.default};
  min-height: ${({ theme }) => theme.__hd__.checkbox.lineHeights.default};
  color: ${({ theme }) => theme.__hd__.checkbox.colors.text};

  &::before {
    content: '';
    position: absolute;
    width: ${({ theme }) => theme.__hd__.checkbox.sizes.width};
    height: ${({ theme }) => theme.__hd__.checkbox.sizes.height};
    top: ${({ theme }) => theme.__hd__.checkbox.space.top};
    left: 0;
    border: ${({ theme }) => theme.__hd__.checkbox.borderWidths.default} solid
      ${({ theme }) => theme.__hd__.checkbox.colors.defaultBorder};
    border-radius: ${({ theme }) => theme.__hd__.checkbox.radii.default};
    background: ${({ theme }) =>
      theme.__hd__.checkbox.colors.defaultBackground};
    box-sizing: border-box;
  }

  &:hover {
    cursor: pointer;

    &::before {
      border-color: ${({ theme }) =>
        theme.__hd__.checkbox.colors.checkedBorder};
    }
  }
`;

const HiddenInput = styled.input`
  display: none;
  margin: 0;
  padding: 0;

  &:checked ~ ${LabelText} {
    &::before {
      border-color: ${({ theme }) =>
        theme.__hd__.checkbox.colors.checkedBorder};
      background: ${({ theme }) =>
        theme.__hd__.checkbox.colors.checkedBackground};
    }

    &::after {
      content: '';
      position: absolute;
      height: ${({ theme }) => theme.__hd__.checkbox.sizes.markHeight};
      width: ${({ theme }) => theme.__hd__.checkbox.sizes.markWidth};
      top: ${({ theme }) => theme.__hd__.checkbox.space.markTop};
      left: ${({ theme }) => theme.__hd__.checkbox.space.markLeft};
      border-width: 0
        ${({ theme }) => theme.__hd__.checkbox.borderWidths.default}
        ${({ theme }) => theme.__hd__.checkbox.borderWidths.default} 0;
      border-color: ${({ theme }) =>
        theme.__hd__.checkbox.colors.defaultCheckmark};
      border-style: solid;
      transform: translate(-50%, -30%) rotate(45deg);
      box-sizing: border-box;
    }
  }

  &:indeterminate ~ ${LabelText} {
    &::after {
      content: '';
      position: absolute;
      height: ${({ theme }) =>
        theme.__hd__.checkbox.sizes.indeterminateBoxHeight};
      width: ${({ theme }) =>
        theme.__hd__.checkbox.sizes.indeterminateBoxWidth};
      top: ${({ theme }) => theme.__hd__.checkbox.space.indeterminateBoxTop};
      left: ${({ theme }) => theme.__hd__.checkbox.space.indeterminateBoxLeft};
      background: ${({ theme }) =>
        theme.__hd__.checkbox.colors.indeterminateBoxBackground};
      border-radius: ${({ theme }) =>
        theme.__hd__.checkbox.radii.indeterminateBox};
    }
  }

  &:disabled ~ ${LabelText} {
    color: ${({ theme }) => theme.__hd__.checkbox.colors.disabledText};

    &:hover {
      cursor: not-allowed;
    }

    &::before {
      border-color: ${({ theme }) =>
        theme.__hd__.checkbox.colors.defaultBorder};
      background: ${({ theme }) =>
        theme.__hd__.checkbox.colors.disabledBackground};
    }

    &::after {
      border-color: ${({ theme }) =>
        theme.__hd__.checkbox.colors.disabledCheckmark};
    }
  }
`;

const GroupWrapper = styled.div<{ themeLayout: 'vertical' | 'horizontal' }>`
  display: flex;
  margin: 0;
  padding: 0;

  ${({ themeLayout, theme }) => {
    switch (themeLayout) {
      case 'vertical':
        return css`
          flex-direction: column;
          & > ${Label}:not(:first-child) {
            margin-top: ${theme.__hd__.checkbox.space.marginTop};
          }
        `;
      case 'horizontal':
        return css`
          & > ${Label}:not(:first-child) {
            margin-left: ${theme.__hd__.checkbox.space.marginLeft};
          }
        `;
    }
  }};
`;

const ButtonLabel = styled.label`
  display: inline-block;
  margin: 0;
  padding: 0;
`;

const ButtonLabelText = styled.span<{
  themeSize: 'small' | 'medium' | 'large';
}>`
  box-sizing: border-box;
  line-height: 100%;
  display: flex;
  align-items: center;
  cursor: pointer;
  border-style: solid;
  border-color: ${({ theme }) =>
    theme.__hd__.checkbox.colors.buttonDefaultBorder};
  border-width: ${({ theme }) => theme.__hd__.checkbox.borderWidths.default};
  border-radius: ${({ theme }) => theme.__hd__.checkbox.radii.default};
  font-weight: ${({ theme }) => theme.__hd__.checkbox.fontWeights.default};
  margin: 0;
  padding: 0;
  color: ${({ theme }) => theme.__hd__.checkbox.colors.buttonText};
  &:hover,
  &:focus {
    color: ${({ theme }) => theme.__hd__.checkbox.colors.buttonHover};
    border-color: ${({ theme }) => theme.__hd__.checkbox.colors.buttonHover};
  }

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          height: ${theme.__hd__.checkbox.sizes.buttonSmall};
          padding: ${theme.__hd__.checkbox.space.buttonSmallPadding};
          font-size: ${theme.__hd__.checkbox.fontSizes.buttonsmall};
        `;
      case 'medium':
        return css`
          height: ${theme.__hd__.checkbox.sizes.buttonMedium};
          padding: ${theme.__hd__.checkbox.space.buttonMediumPadding};
          font-size: ${theme.__hd__.checkbox.fontSizes.buttonMedium};
        `;
      case 'large':
        return css`
          height: ${theme.__hd__.checkbox.sizes.buttonLarge};
          padding: ${theme.__hd__.checkbox.space.buttonLargePadding};
          font-size: ${theme.__hd__.checkbox.fontSizes.buttonLarge};
        `;
    }
  }};
`;

const ButtonHiddenInput = styled.input`
  opacity: 0;
  width: 0;
  height: 0;
  position: absolute;
  margin: 0;
  padding: 0;

  &:disabled ~ ${ButtonLabelText} {
    color: ${({ theme }) => theme.__hd__.checkbox.colors.buttonDisabledText};
    background-color: ${({ theme }) =>
      theme.__hd__.checkbox.colors.buttonDisabledBackground};
    border-color: ${({ theme }) =>
      theme.__hd__.checkbox.colors.buttonDefaultBorder};
    pointer-events: none;
  }

  &:disabled:checked ~ ${ButtonLabelText} {
    color: ${({ theme }) =>
      theme.__hd__.checkbox.colors.buttonDisabledSelectedText};
    background-color: ${({ theme }) =>
      theme.__hd__.checkbox.colors.buttonDisabledSelectedBackground};
    border: none;
  }

  &:checked + ${ButtonLabelText} {
    border-color: ${({ theme }) =>
      theme.__hd__.checkbox.colors.buttonSelectedBackground};
    color: ${({ theme }) => theme.__hd__.checkbox.colors.buttonSelectedText};
    background: ${({ theme }) =>
      theme.__hd__.checkbox.colors.buttonSelectedBackground};
  }
`;

const ButtonGroupWrapper = styled.div<{
  themeSize: 'small' | 'medium' | 'large';
}>`
  display: inline-flex;
  flex-wrap: wrap;
  padding: 0;

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          margin: ${theme.__hd__.checkbox.space.buttonGroupSmallMargin};
          width: ${theme.__hd__.checkbox.sizes.buttonGroupSmallWidth};
          > * {
            margin: ${theme.__hd__.checkbox.space.buttonSmallMargin};
          }
        `;
      case 'medium':
        return css`
          margin: ${theme.__hd__.checkbox.space.buttonGroupMediumMargin};
          width: ${theme.__hd__.checkbox.sizes.buttonGroupMediumWidth};
          > * {
            margin: ${theme.__hd__.checkbox.space.buttonMediumMargin};
          }
        `;
      case 'large':
        return css`
          margin: ${theme.__hd__.checkbox.space.buttonGroupLargeMargin};
          width: ${theme.__hd__.checkbox.sizes.buttonGroupLargeWidth};
          > * {
            margin: ${theme.__hd__.checkbox.space.buttonLargeMargin};
          }
        `;
    }
  }};
`;

export {
  Label,
  LabelText,
  HiddenInput,
  GroupWrapper,
  ButtonLabel,
  ButtonLabelText,
  ButtonHiddenInput,
  ButtonGroupWrapper,
};
