// @hero-design/react/src/components/Input/StyledInput.tsx
import styled, { css } from 'styled-components';
import { StyledTag, CloseButton } from '../Tag/StyledTag';
import type { InputState, Size } from './utils';

const InputWrapper = styled.div<{
  themeSize: Size;
  themeState: InputState;
}>`
  display: flex;
  align-items: center;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  border: ${({ theme }) => theme.__hd__.input.borderWidths.default} solid
    ${({ theme }) => theme.__hd__.input.colors.border};
  border-radius: ${({ theme }) => theme.__hd__.input.radii.default};
  line-height: 100%;
  font-weight: ${({ theme }) => theme.__hd__.input.fontWeights.default};

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          height: ${theme.__hd__.input.sizes.small};
          font-size: ${theme.__hd__.input.fontSizes.small};
        `;
      case 'medium':
        return css`
          height: ${theme.__hd__.input.sizes.medium};
          font-size: ${theme.__hd__.input.fontSizes.medium};
        `;
      case 'large':
        return css`
          height: ${theme.__hd__.input.sizes.large};
          font-size: ${theme.__hd__.input.fontSizes.large};
        `;
    }
  }};

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'enabled':
        return css`
          background: ${theme.__hd__.input.colors.background};
          &:hover {
            border-color: ${theme.__hd__.input.colors.focusBorder};
          }
          &:focus-within {
            outline: none;
            border-color: ${theme.__hd__.input.colors.focusBorder};
            box-shadow: ${theme.__hd__.input.shadows.focus};
          }
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.input.colors.disabledText};
          border-color: ${theme.__hd__.input.colors.border};
          background: ${theme.__hd__.input.colors.disabledBackground};
          cursor: not-allowed;
          & ${StyledTag} {
            color: ${theme.__hd__.input.colors.disabledText};
            border-color: ${theme.__hd__.input.colors.border};
            background: ${theme.__hd__.input.colors.disabledBackground};

            & > ${CloseButton} {
              color: ${theme.__hd__.input.colors.disabledText};
              pointer-events: none;
              cursor: not-allowed;
            }
          }
        `;
      case 'invalid':
        return css`
          background: ${theme.__hd__.input.colors.background};
          border-color: ${theme.__hd__.input.colors.invalidBorder};
          &:focus-within {
            outline: none;
            border-color: ${theme.__hd__.input.colors.invalidBorder};
            box-shadow: ${theme.__hd__.input.shadows.invalidFocus};
          }
        `;
    }
  }};
`;

const StyledInput = styled.input<{
  themePaddingLeft: Size | 'none';
  themePaddingRight: Size | 'none';
}>`
  flex: 1;
  height: 100%;
  overflow: hidden;
  margin: 0;
  padding: 0;
  border: none;
  border-radius: ${({ theme }) => theme.__hd__.input.radii.default};
  font-weight: inherit;
  font-size: inherit;

  color: ${({ theme }) => theme.__hd__.input.colors.text};
  background-color: ${({ theme }) => theme.__hd__.input.colors.background};

  &:focus {
    outline: none;
  }

  &:disabled {
    cursor: not-allowed;
    color: ${({ theme }) => theme.__hd__.input.colors.disabledText};
    background-color: ${({ theme }) =>
      theme.__hd__.input.colors.disabledBackground};
  }

  &::placeholder {
    color: ${({ theme }) => theme.__hd__.input.colors.disabledText};
  }

  ${({ themePaddingLeft, theme }) => {
    switch (themePaddingLeft) {
      case 'small':
        return css`
          padding-left: ${theme.__hd__.input.space.smallPadding};
        `;
      case 'medium':
        return css`
          padding-left: ${theme.__hd__.input.space.mediumPadding};
        `;
      case 'large':
        return css`
          padding-left: ${theme.__hd__.input.space.largePadding};
        `;
      case 'none':
        return css`
          padding-left: 0;
        `;
    }
  }};

  ${({ themePaddingRight, theme }) => {
    switch (themePaddingRight) {
      case 'small':
        return css`
          padding-right: ${theme.__hd__.input.space.smallPadding};
        `;
      case 'medium':
        return css`
          padding-right: ${theme.__hd__.input.space.mediumPadding};
        `;
      case 'large':
        return css`
          padding-right: ${theme.__hd__.input.space.largePadding};
        `;
      case 'none':
        return css`
          padding-right: 0;
        `;
    }
  }};
`;

const AffixWrapper = styled.span<{
  themeMarginLeft: 'auto' | 'none';
  themeSize: Size;
}>`
  margin: 0;
  padding: 0;

  > * {
    display: block;
  }

  ${({ themeMarginLeft }) => {
    switch (themeMarginLeft) {
      case 'auto':
        return css`
          margin-left: auto;
        `;
      case 'none':
        return css``;
    }
  }};

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          padding-left: ${theme.__hd__.input.space.smallPadding};
          padding-right: ${theme.__hd__.input.space.smallPadding};
        `;
      case 'medium':
        return css`
          padding-left: ${theme.__hd__.input.space.mediumPadding};
          padding-right: ${theme.__hd__.input.space.mediumPadding};
        `;
      case 'large':
        return css`
          padding-left: ${theme.__hd__.input.space.largePadding};
          padding-right: ${theme.__hd__.input.space.largePadding};
        `;
    }
  }};
`;

export { InputWrapper, AffixWrapper };

export default StyledInput;
