// @hero-design/react/src/components/Comment/StyledComment.tsx
import styled, { css } from 'styled-components';

import StyledTextArea from '../Input/StyledTextArea';

const Avatar = styled.div`
  margin: 0;
  padding: 0;
  margin-right: ${({ theme }) => theme.__hd__.comment.space.avatarMarginRight};

  img {
    width: ${({ theme }) => theme.__hd__.comment.sizes.avatarSize};
    height: ${({ theme }) => theme.__hd__.comment.sizes.avatarSize};
    border-radius: 50%;
    margin: 0;
    padding: 0;
  }
`;

const AuthorWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const Author = styled.div`
  display: flex;
  align-items: center;
  font-weight: ${({ theme }) => theme.__hd__.comment.fontWeights.author};
  font-size: ${({ theme }) => theme.__hd__.comment.fontSizes.author};
  line-height: ${({ theme }) => theme.__hd__.comment.lineHeights.author};
  color: ${({ theme }) => theme.__hd__.comment.colors.author};
  margin: 0;
  padding: 0;
`;

const AuthorDetails = styled.div`
  margin: 0;
  padding: 0;
  font-weight: ${({ theme }) => theme.__hd__.comment.fontWeights.authorDetails};
  font-size: ${({ theme }) => theme.__hd__.comment.fontSizes.authorDetails};
  line-height: ${({ theme }) => theme.__hd__.comment.lineHeights.authorDetails};
  color: ${({ theme }) => theme.__hd__.comment.colors.authorDetails};
`;

const TimeWrapper = styled.div`
  white-space: nowrap;
  font-weight: ${({ theme }) => theme.__hd__.comment.fontWeights.time};
  font-size: ${({ theme }) => theme.__hd__.comment.fontSizes.time};
  line-height: ${({ theme }) => theme.__hd__.comment.lineHeights.time};
  color: ${({ theme }) => theme.__hd__.comment.colors.time};
  margin: 0;
  padding: 0;
`;

const TimeSeparator = styled.span`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.comment.space.timeSeparatorPadding};
`;

const ContentWrapper = styled.div`
  position: relative;
  margin: 0;
  padding: 0;
  font-weight: ${({ theme }) => theme.__hd__.comment.fontWeights.content};
  font-size: ${({ theme }) => theme.__hd__.comment.fontSizes.content};
  line-height: ${({ theme }) => theme.__hd__.comment.lineHeights.content};
  flex-grow: 1;
  color: ${({ theme }) => theme.__hd__.comment.colors.content};
`;

const ContentDetail = styled.div`
  white-space: pre-wrap;
  margin: 0;
  padding: 0;
  margin-top: ${({ theme }) => theme.__hd__.comment.space.contentMarginTop};
`;

const ActionsWrapper = styled.div<{ themeShown: boolean }>`
  position: absolute;
  top: 0;
  right: 0;
  margin: 0;
  padding: 0;

  ${({ themeShown }) => {
    switch (themeShown) {
      case true:
        return css`
          display: block;
        `;
      case false:
        return css`
          display: none;
        `;
    }
  }};
`;

const ReactionsWrapper = styled.div`
  margin: 0;
  padding: 0;
  margin-top: ${({ theme }) => theme.__hd__.comment.space.reactionMarginTop};
`;

const CommentWrapper = styled.div`
  margin: 0;
  padding: 0;

  &:not(:last-child) {
    margin-bottom: ${({ theme }) => theme.__hd__.comment.space.marginBottom};
  }
`;

const CommentInner = styled.div`
  display: flex;
  padding: ${({ theme }) => theme.__hd__.comment.space.padding};
  border-radius: ${({ theme }) => theme.__hd__.comment.radii.container};
  margin: 0;
  &:hover {
    background-color: ${({ theme }) => theme.__hd__.comment.colors.highlight};
  }
`;

const NestedCommentWrapper = styled.div`
  margin: ${({ theme }) => theme.__hd__.comment.space.nestedMargin};
  padding: 0;
`;

const EditorContainer = styled.div`
  display: flex;
  margin: 0;
  padding: 0;

  ${Avatar} {
    margin-left: ${({ theme }) => theme.__hd__.comment.space.avatarMarginLeft};
  }
`;

const EditorWrapper = styled.div`
  margin: 0;
  padding: 0;
  border: ${({ theme }) => theme.__hd__.input.borderWidths.default} solid
    ${({ theme }) => theme.__hd__.input.colors.border};
  border-radius: ${({ theme }) => theme.__hd__.input.radii.default};
  width: 0;
  flex-grow: 1;
  flex-shrink: 1;
  color: ${({ theme }) => theme.__hd__.comment.colors.content};

  &:hover {
    border-color: ${({ theme }) => theme.__hd__.input.colors.focusBorder};
  }

  &:focus-within {
    outline: none;
    border-color: ${({ theme }) => theme.__hd__.input.colors.focusBorder};
    box-shadow: ${({ theme }) => theme.__hd__.input.shadows.focus};
  }

  ${StyledTextArea} {
    border: none;
    box-shadow: none;
  }
`;

const EditorActionsWrapper = styled.div`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.comment.space.editorActionsPadding};
  display: flex;
  flex-direction: row-reverse;
  align-items: flex-end;
  justify-content: space-between;
`;

export {
  Avatar,
  AuthorWrapper,
  Author,
  AuthorDetails,
  TimeWrapper,
  TimeSeparator,
  ContentWrapper,
  ContentDetail,
  ActionsWrapper,
  ReactionsWrapper,
  CommentWrapper,
  CommentInner,
  NestedCommentWrapper,
  EditorContainer,
  EditorWrapper,
  EditorActionsWrapper,
};
