// @hero-design/react/src/components/Progress/StyledProgressCircle.tsx
import styled, { css } from 'styled-components';
import { ProgressIntent } from './types';
import { getColorFromProgressIntent } from './utils';

const Wrapper = styled.div<{
  themeIntent: ProgressIntent;
  themeSize: 'small' | 'medium' | 'large';
}>`
  box-sizing: border-box;
  color: ${({ theme, themeIntent }) =>
    getColorFromProgressIntent({
      theme,
      intent: themeIntent,
      getter: (definedThemeIntent) =>
        theme.__hd__.progress.colors.circleText[definedThemeIntent],
      fallbackColor: theme.colors.bodyText,
    })};
  display: inline-block;
  position: relative;
  margin: 0;
  padding: 0;
  svg {
    overflow: initial;
  }

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          width: ${theme.__hd__.progress.sizes.circleSmall};
          height: ${theme.__hd__.progress.sizes.circleSmall};
          font-size: ${theme.__hd__.progress.fontSizes.circleSmall};
        `;
      case 'medium':
        return css`
          width: ${theme.__hd__.progress.sizes.circleMedium};
          height: ${theme.__hd__.progress.sizes.circleMedium};
          font-size: ${theme.__hd__.progress.fontSizes.circleMedium};
        `;
      case 'large':
        return css`
          width: ${theme.__hd__.progress.sizes.circleLarge};
          height: ${theme.__hd__.progress.sizes.circleLarge};
          font-size: ${theme.__hd__.progress.fontSizes.circleLarge};
        `;
    }
  }};
`;

const Info = styled.div`
  display: flex;
  justify-content: center;
  position: absolute;
  top: 50%;
  left: 50%;
  width: 100%;
  transform: translate(-50%, -50%);
  margin: 0;
  padding: 0;
`;

export { Wrapper, Info };
