// @hero-design/react/src/components/Sidebar/StyledSideBar.tsx
import styled, { css } from 'styled-components';
import { CountNumberBadge } from '../Badge/StyledBadge';

const scrollBarStyle = css`
    overflow-y: scroll;
    scrollbar-width: thin;
    scrollbar-color: ${({ theme }) =>
      `${theme.__hd__.sideBar.colors.scrollBarThumb} transparent`};

    &::-webkit-scrollbar {
      background-color: transparent;
      width: ${({ theme }) => theme.__hd__.sideBar.sizes.scrollBar};
    }
    &::-webkit-scrollbar-thumb {
      background-color: ${({ theme }) =>
        theme.__hd__.sideBar.colors.scrollBarThumb};
      border-radius: ${({ theme }) =>
        `calc(${theme.__hd__.sideBar.sizes.scrollBar} / 2)`};
      border-width: ${({ theme }) =>
        theme.__hd__.sideBar.borderWidths.scrollBar};
      border-style: solid;
      border-color: transparent;
      background-clip: content-box;
    }
    &::-webkit-scrollbar-track {
      margin: ${({ theme }) => theme.__hd__.sideBar.space.scrollBarMargin};
    }
  }
`;

const SideBarContainer = styled.div`
  display: flex;
`;

const NavBar = styled.nav`
  margin: 0;
  padding: 0;
  display: flex;
  align-items: stretch;
  overflow: hidden;
  position: relative;
`;

const SideBarBackground = styled.div`
  margin: 0;
  padding: 0;
  position: absolute;
  z-index: 2;
  background-color: ${({ theme }) =>
    theme.__hd__.sideBar.colors.sideBarBackground};
  border-radius: ${({ theme }) =>
    `0px ${theme.__hd__.sideBar.radii.sideBarWrapperRight} ${theme.__hd__.sideBar.radii.sideBarWrapperRight} 0px`};
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.sideBarWidth};
  height: 100dvh;
`;

const ScrollWrapper = styled.div<{
  themeShowPanel: boolean;
}>`
  margin: 0;
  padding: 0;
  max-height: 100dvh;
  overflow-x: hidden;
  overflow-y: hidden;
  z-index: 2;
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.sideBarScrollWrapperWidth};

  @media (hover: hover) {
    &:hover {
      ${scrollBarStyle};
    }
  }

  @media (pointer: coarse) {
    ${scrollBarStyle};
  }
`;

const SideBarWrapper = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
  text-align: center;
  min-height: 100dvh;
  display: flex;
  flex-direction: column;
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.sideBarWidth};
  border-radius: ${({ theme }) =>
    `0px ${theme.__hd__.sideBar.radii.sideBarWrapperRight} ${theme.__hd__.sideBar.radii.sideBarWrapperRight} 0px`};
`;

const PrimaryItemIconWrapper = styled.div`
  position: relative;
  border-radius: ${({ theme }) =>
    `calc(${theme.__hd__.sideBar.sizes.itemIconHeight} / 2)`};
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.itemIconWidth};
  height: ${({ theme }) => theme.__hd__.sideBar.sizes.itemIconHeight};
`;

const PrimaryItem = styled.li<{
  themeActive: boolean;
  themeDisabled?: boolean;
}>`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.sideBar.space.sideBarItemPadding};
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.itemWidth};
  color: ${({ theme }) => theme.__hd__.sideBar.colors.sideBarText};
  border-radius: ${({ theme }) => theme.__hd__.sideBar.radii.sideBarItem};
  cursor: pointer;

  &:hover {
    > ${PrimaryItemIconWrapper} {
      background-color: ${({ theme, themeDisabled }) =>
        !themeDisabled
          ? `${theme.__hd__.sideBar.colors.sidebarItemIconBackground}66`
          : 'none'};
    }
  }

  & > ${PrimaryItemIconWrapper}::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: ${({ theme }) =>
      theme.__hd__.sideBar.colors.sidebarItemIconBackground};
    border-radius: ${({ theme }) =>
      `calc(${theme.__hd__.sideBar.sizes.itemIconHeight} / 2)`};
    transform: scaleX(0);
  }

  ${({ themeActive, themeDisabled }) => {
    if (themeDisabled) {
      return css`
        cursor: not-allowed;
      `;
    }
    switch (themeActive) {
      case true:
        return css`
          & > ${PrimaryItemIconWrapper}::before {
            transform: scaleX(1) !important;
            transition: transform 0.3s;
          }
        `;
      case false:
        return css``;
    }
  }};
`;

const HeaderItem = styled(PrimaryItem)`
  font-size: ${({ theme }) => theme.__hd__.sideBar.fontSizes.headerIcon};
  color: ${({ theme }) => theme.__hd__.sideBar.colors.sideBarText};
  background: transparent;
`;

const ItemLinkWraper = styled.a`
  display: flex;
  text-decoration: none;
`;

const ItemsContainer = styled.div`
  margin: 0;
  padding: ${({ theme }) =>
    theme.__hd__.sideBar.space.sideBarItemsContainerPadding};

  > *:not(:last-child) {
    margin-bottom: ${({ theme }) =>
      theme.__hd__.sideBar.space.sideBarItemMarginBottom};
  }

  ${CountNumberBadge} {
    background-color: ${({ theme }) =>
      theme.__hd__.sideBar.colors.sideBarCountBackground};
    border-color: ${({ theme }) =>
      theme.__hd__.sideBar.colors.sideBarCountBackground};
    color: ${({ theme }) => theme.__hd__.sideBar.colors.sideBarCountText};
    transform: none;
    right: unset;
    top: 0;
    left: 52px;
    font-size: ${({ theme }) =>
      theme.__hd__.sideBar.fontSizes.sideBarNotiBadge};
    padding: ${({ theme }) => theme.__hd__.sideBar.space.sideBarNotiBadge};
    height: ${({ theme }) => theme.__hd__.sideBar.sizes.panelItemBadge};
    min-width: ${({ theme }) => theme.__hd__.sideBar.sizes.panelItemBadge};
    z-index: 1;
    display: flex;
    align-items: center;
    justify-content: center;
  }
`;

const HeaderItemsContainer = styled(ItemsContainer)`
  padding: ${({ theme }) =>
    theme.__hd__.sideBar.space.sideBarHeaderItemsContainerPadding};
`;

const PanelWrapper = styled.div<{ themeActive: boolean }>`
  z-index: 1;
  transition: transform 0.3s ease;

  ${({ theme, themeActive }) => {
    switch (themeActive) {
      case true:
        return css`
          margin-left: -${theme.__hd__.sideBar.sizes.scrollBar};
          transform: translateX(
            -${theme.__hd__.sideBar.radii.sideBarWrapperRight}
          );
        `;
      case false:
        return css`
          transform: translateX(-100%);
        `;
    }
  }};
`;

export {
  SideBarContainer,
  NavBar,
  ScrollWrapper,
  SideBarBackground,
  SideBarWrapper,
  ItemsContainer,
  PrimaryItem,
  HeaderItem,
  HeaderItemsContainer,
  ItemLinkWraper,
  PanelWrapper,
  PrimaryItemIconWrapper,
};
