// @hero-design/react/src/components/Sidebar/StyledPanel.tsx
import styled, { css } from 'styled-components';
import { StyledBadge } from '../Badge/StyledBadge';

const PanelBackground = styled.div`
  margin: 0;
  padding: 0;
  position: absolute;
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.panelWidth};
  height: 100dvh;
  background-color: ${({ theme }) =>
    theme.__hd__.sideBar.colors.panelBackground};
  box-shadow: ${({ theme }) =>
    theme.__hd__.sideBar.shadows.panelContainerShadow};
  border-radius: ${({ theme }) => theme.__hd__.sideBar.radii.panelContainter};
  z-index: 1;
`;

const BackButtonWrapper = styled.div<{ themeVisibility: boolean }>`
  display: flex;
  align-items: center;
  cursor: pointer;
  padding: ${({ theme }) => theme.__hd__.sideBar.space.panelBackButtonPadding};

  > *:first-child {
    margin-right: ${({ theme }) =>
      theme.__hd__.sideBar.space.backIconMarginRight};
  }

  ${({ themeVisibility }) => ({
    visibility: themeVisibility ? 'visible' : 'hidden',
  })};
`;

const PanelSubheaderWrapper = styled.div`
  display: flex;
  align-items: center;
  padding: ${({ theme }) => theme.__hd__.sideBar.space.panelSubHeaderPadding};
`;

const PanelItemWrapper = styled.div<{ themeActive: boolean }>`
  margin: 0;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: ${({ theme }) => theme.__hd__.sideBar.space.panelItemPadding};
  border-radius: ${({ theme }) => theme.__hd__.sideBar.radii.panelItem};
  cursor: pointer;
  box-sizing: border-box;

  &:hover {
    background-color: ${({ theme }) =>
      `${theme.__hd__.sideBar.colors.panelItemHoverBackground}0d`};
  }

  ${({ theme, themeActive }) => {
    switch (themeActive) {
      case true:
        return css`
          background-color: ${theme.__hd__.sideBar.colors
            .panelItemActiveBackground};
        `;
      case false:
        return css``;
    }
  }};

  > ${StyledBadge} {
    display: flex;
    align-items: center;
    justify-content: center;
    height: ${({ theme }) => theme.__hd__.sideBar.sizes.panelItemBadge};
    min-width: ${({ theme }) => theme.__hd__.sideBar.sizes.panelItemBadge};
  }
`;

const ItemLinkWrapper = styled.a`
  display: block;
  text-decoration: none;
`;

const PanelItemsContainer = styled.div`
  margin: 0;
  padding: 0;

  > ${ItemLinkWrapper}:not(:first-child) {
    margin-top: ${({ theme }) => theme.__hd__.sideBar.space.panelItemMarginTop};
  }
`;

const PanelScroller = styled.div`
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 2;
  max-height: 100dvh;
  overflow-x: hidden;
  overflow-y: hidden;
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.panelWrapperWidth};
  border-radius: ${({ theme }) => theme.__hd__.sideBar.radii.panelContainter};

  &:hover {
    overflow-y: scroll;
    scrollbar-width: thin;
    scrollbar-color: ${({ theme }) =>
      `${theme.__hd__.sideBar.colors.scrollBarThumb} transparent`};

    &::-webkit-scrollbar {
      background-color: transparent;
      width: ${({ theme }) => theme.__hd__.sideBar.sizes.scrollBar};
    }
    &::-webkit-scrollbar-thumb {
      background-color: ${({ theme }) =>
        theme.__hd__.sideBar.colors.scrollBarThumb};
      border-radius: ${({ theme }) =>
        `calc(${theme.__hd__.sideBar.sizes.scrollBar} / 2)`};
      border-width: ${({ theme }) =>
        theme.__hd__.sideBar.borderWidths.scrollBar};
      border-style: solid;
      border-color: transparent;
      background-clip: content-box;
    }
    &::-webkit-scrollbar-track {
      margin: ${({ theme }) => theme.__hd__.sideBar.space.scrollBarMargin};
    }
  }
`;

const PanelContainer = styled.div`
  min-height: 100dvh;
  padding: ${({ theme }) => theme.__hd__.sideBar.space.panelContainerPadding};
  border-radius: ${({ theme }) => theme.__hd__.sideBar.radii.panelContainter};
  width: ${({ theme }) => theme.__hd__.sideBar.sizes.panelWidth};

  > ${PanelItemsContainer}:nth-child(n + 4) {
    margin-top: ${({ theme }) =>
      theme.__hd__.sideBar.space.panelItemContainerMarginTop};
  }
`;

const PanelHeaderWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: ${({ theme }) => theme.__hd__.sideBar.sizes.panelHeaderHeight};
`;

export {
  PanelBackground,
  PanelScroller,
  PanelContainer,
  BackButtonWrapper,
  PanelItemsContainer,
  ItemLinkWrapper,
  PanelItemWrapper,
  PanelHeaderWrapper,
  PanelSubheaderWrapper,
};
