// @hero-design/colors/src/defaultMobileDarkPalette/index.ts
import colorScales from '../colorScales';

const {
  aliceBlue,
  antiFlashWhite,
  apple,
  blazingBonfire,
  bigStone,
  flagstone,
  deepSaffron,
  deepIntoTheJungle,
  emerald,
  grey,
  honeydew,
  indianDance,
  linen,
  maasstrichtBlue,
  mellowApricot,
  moussaka,
  pastelRed,
  pineGreen,
  seashell,
  silverChalice,
  sonicSilver,
  ultramarineBlue,
  vermilion,
  vodka,
  theatreGold,
  uniformGreen,
  mintLeaf,
  blueOblivion,
  pastelBlue,
} = colorScales;

const palette = {
  black: '#000000',
  white: '#ffffff',

  greyLight95: grey.lighten95,
  greyLight90: grey.lighten90,
  greyLight75: grey.lighten75,
  greyLight45: grey.lighten45,
  greyLight30: grey.lighten30,
  grey: grey.base,
  greyDark20: grey.darken20,
  greyDark30: grey.darken30,

  maasstrichtBlueLight90: maasstrichtBlue.lighten90,
  maasstrichtBlueLight80: maasstrichtBlue.lighten80,
  maasstrichtBlueLight70: maasstrichtBlue.lighten70,
  maasstrichtBlueLight60: maasstrichtBlue.lighten60,
  maasstrichtBlueLight50: maasstrichtBlue.lighten50,
  maasstrichtBlueLight30: maasstrichtBlue.lighten30,
  maasstrichtBlueLight15: maasstrichtBlue.lighten15,
  maasstrichtBlueLight10: maasstrichtBlue.lighten10,
  maasstrichtBlue: maasstrichtBlue.base,

  pastelRed: pastelRed.base,
  pastelRedLight20: pastelRed.lighten20,

  mellowApricot: mellowApricot.base,
  mellowApricotLight20: mellowApricot.lighten20,

  moussaka: moussaka.base,

  emerald: emerald.base,
  emeraldLight20: emerald.lighten20,
  emeraldDark20: emerald.darken20,

  vodka: vodka.base,
  vodkaLight20: vodka.lighten20,

  vermilion: vermilion.base,
  vermilionLight20: vermilion.lighten20,
  vermilionDark10: vermilion.darken10,

  deepSaffron: deepSaffron.base,
  deepSaffronDark20: deepSaffron.darken20,

  indianDance: indianDance.base,

  theatreGold: theatreGold.base,
  theatreGoldLight20: theatreGold.lighten20,

  pineGreen: pineGreen.base,
  pineGreenLight20: pineGreen.lighten20,
  pineGreenDark10: pineGreen.darken10,

  ultramarineBlue: ultramarineBlue.base,
  ultramarineBlueLight20: ultramarineBlue.lighten20,
  ultramarineBlueDark10: ultramarineBlue.darken10,

  sonicSilver: sonicSilver.base,
  sonicSilverLight50: sonicSilver.lighten45,
  sonicSilverDark10: sonicSilver.darken10,
  linen: linen.base,
  seashell: seashell.base,
  honeydew: honeydew.base,
  aliceBlue: aliceBlue.base,
  antiFlashWhite: antiFlashWhite.base,
  apple: apple.base,

  silverChalice: silverChalice.base,
  silverChaliceLight20: silverChalice.lighten20,

  uniformGreen: uniformGreen.base,
  blazingBonfire: blazingBonfire.base,
  deepIntoTheJungle: deepIntoTheJungle.base,
  mintLeaf: mintLeaf.base,
  blueOblivion: blueOblivion.base,
  pastelBlue: pastelBlue.base,

  bigStone: bigStone.base,
  flagstone: flagstone.base,
};

export default palette;
