// @hero-design/react/src/components/Empty/index.tsx
import React from 'react';
import type { ReactElement } from 'react';
import { useCss } from '../../utils/hooks';
import SvgEmpty from './assets/empty.svg';
import {
  EmptyWrapper,
  EmptyContentWrapper,
  EmptyImg,
  EmptyDescriptionWrapper,
  EmptyTitleWrapper,
  EmptyTextGroupWrapper,
  EmptyExtraWrapper,
} from './StyledEmpty';
import type { CommonProps } from '../common';

export interface EmptyProps extends CommonProps {
  /**
   * Extra action buttons.
   */
  extra?: ReactElement;
  /**
   *  Empty image source URL.
   */
  imageSrc?: string;
  /**
   * Size of the Empty.
   */
  size?: 'small' | 'medium';
  /**
   *  Text under image.
   */
  text: string | ReactElement;
  /**
   *  Optional heading text under image.
   */
  title?: string | ReactElement;
}

const Empty = ({
  imageSrc,
  text,
  title,
  extra,
  size = 'medium',
  id,
  className,
  style,
  sx = {},
  'data-test-id': dataTestId,
}: EmptyProps): ReactElement => {
  return (
    <EmptyWrapper
      id={id}
      className={className}
      style={{ ...style, ...useCss(sx) }}
      data-test-id={dataTestId}
    >
      <EmptyContentWrapper themeSize={size}>
        {imageSrc ? <EmptyImg src={imageSrc} alt="empty-img" /> : <SvgEmpty />}
        <EmptyTextGroupWrapper>
          {title !== undefined && (
            <EmptyTitleWrapper>{title}</EmptyTitleWrapper>
          )}
          <EmptyDescriptionWrapper>{text}</EmptyDescriptionWrapper>
        </EmptyTextGroupWrapper>
        {React.isValidElement(extra) === true && (
          <EmptyExtraWrapper>{extra}</EmptyExtraWrapper>
        )}
      </EmptyContentWrapper>
    </EmptyWrapper>
  );
};

export default Empty;
