// @hero-design/react/src/components/Empty/StyledEmpty.tsx
import styled, { css } from 'styled-components';

const EmptyWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: 0;
  padding: 0;
`;

const EmptyContentWrapper = styled.div<{ themeSize: 'small' | 'medium' }>`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.empty.space.wrapperPadding};

  @media (min-width: ${({ theme }) => theme.breakpoints.medium}) {
    ${({ themeSize, theme }) => {
      switch (themeSize) {
        case 'small':
          return css`
            max-width: ${theme.__hd__.empty.sizes.wrapperSmallMaxWidth};
          `;
        case 'medium':
          return css`
            max-width: ${theme.__hd__.empty.sizes.wrapperMediumMaxWidth};
          `;
      }
    }};
  }
`;

const EmptyImg = styled.img`
  margin: 0;
  padding: 0;
`;

const EmptyDescriptionWrapper = styled.div`
  display: flex;
  justify-content: center;
  text-align: center;
  color: ${({ theme }) => theme.__hd__.empty.colors.text};
  font-size: ${({ theme }) => theme.__hd__.empty.fontSizes.text};
  line-height: ${({ theme }) => theme.__hd__.empty.lineHeights.text};
  margin: 0;
`;

const EmptyTitleWrapper = styled.div`
  display: flex;
  justify-content: center;
  text-align: center;
  color: ${({ theme }) => theme.__hd__.empty.colors.text};
  font-size: ${({ theme }) => theme.__hd__.empty.fontSizes.title};
  line-height: ${({ theme }) => theme.__hd__.empty.lineHeights.title};
  font-weight: ${({ theme }) => theme.__hd__.empty.fontWeights.title};
  margin: 0;
  margin-bottom: ${({ theme }) => theme.__hd__.empty.space.titleBottomMargin};
`;

const EmptyExtraWrapper = styled.div`
  margin: 0;
  padding: 0;
  margin-top: ${({ theme }) => theme.__hd__.empty.space.extraMarginTop};
`;

const EmptyTextGroupWrapper = styled.div`
  margin: 0;
  padding: 0;
  margin-top: ${({ theme }) => theme.__hd__.empty.space.textGroupMarginTop};
`;

export {
  EmptyWrapper,
  EmptyContentWrapper,
  EmptyImg,
  EmptyDescriptionWrapper,
  EmptyTitleWrapper,
  EmptyExtraWrapper,
  EmptyTextGroupWrapper,
};
