// @hero-design/react/src/components/TimePicker/StyledTimePicker.tsx
import styled, { css } from 'styled-components';

import { DropdownWrapper } from '../Dropdown/StyledDropdown';

const animationDuration = '0.25s';

const PanelColumnWrapper = styled.ul`
  flex: 1 0 auto;
  width: ${({ theme }) => theme.__hd__.timePicker.sizes.columnWidth};
  margin: 0;
  padding: 0;
  text-align: left;
  overflow-x: hidden;
  overflow-y: auto;
  list-style: none;
  border-left-style: solid;
  border-left-width: ${({ theme }) =>
    theme.__hd__.timePicker.borderWidths.column};
  border-left-color: ${({ theme }) =>
    theme.__hd__.timePicker.colors.columnBorder};

  &:focus {
    outline: none;
  }
  &:after {
    display: block;
    height: ${({ theme }) => theme.__hd__.timePicker.sizes.columnAfterHeight};
    content: '';
  }
  &:first-child {
    border: none;
  }
`;

const PanelCellWrapper = styled.li`
  margin: 0;
  padding: 0;
  transition: background ${animationDuration};
  background: ${({ theme }) => theme.__hd__.timePicker.colors.cellBackground};

  &:hover {
    background: ${({ theme }) => theme.__hd__.timePicker.colors.cellHover};
  }
`;

const PanelCellInner = styled.div<{ themeActive: boolean }>`
  display: block;
  width: 100%;
  height: ${({ theme }) => theme.__hd__.timePicker.sizes.cellHeight};
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.timePicker.space.cellPadding};
  color: ${({ theme }) => theme.__hd__.timePicker.colors.cell};
  line-height: ${({ theme }) => theme.__hd__.timePicker.lineHeights.cell};
  font-size: ${({ theme }) => theme.__hd__.timePicker.fontSizes.cell};
  cursor: pointer;

  ${({ themeActive, theme }) => {
    switch (themeActive) {
      case true:
        return css`
          background-color: ${theme.__hd__.timePicker.colors.cellActive};
          font-weight: ${theme.__hd__.timePicker.fontWeights.cellActive};
        `;
      case false:
        return css``;
    }
  }};
`;

const PanelContainer = styled.div`
  display: flex;
  flex: auto;
  height: ${({ theme }) => theme.__hd__.timePicker.sizes.columnHeight};
  width: 100%;
  table-layout: fixed;
  border-collapse: collapse;
  margin: 0;
  padding: 0;

  &:focus {
    outline: none;
  }
`;

const TimePickerWrapper = styled.div`
  margin: 0;
  padding: 0;

  & > ${DropdownWrapper} {
    display: block;
  }
`;

export {
  PanelColumnWrapper,
  PanelCellWrapper,
  PanelCellInner,
  PanelContainer,
  TimePickerWrapper,
};
