// @hero-design/react/src/components/common/ThemeInterfaces.ts
import type { CSSProperties } from 'react';
import type { Space } from '../../theme/global/space';
import type { FontSizes, LineHeights } from '../../theme/global/typography';

interface ThemeScales {
  borderStyles: 'dotted' | 'dashed' | 'solid';
  borderWidths: 'base';
  colors:
    | 'black'
    | 'blue'
    | 'blue-dark-15'
    | 'blue-dark-30'
    | 'blue-dark-45'
    | 'blue-dark-60'
    | 'blue-dark-75'
    | 'blue-dark-90'
    | 'blue-light-15'
    | 'blue-light-30'
    | 'blue-light-45'
    | 'blue-light-60'
    | 'blue-light-75'
    | 'blue-light-90'
    | 'dodger-blue'
    | 'dodger-blue-dark-15'
    | 'dodger-blue-dark-30'
    | 'dodger-blue-light-30'
    | 'dodger-blue-light-45'
    | 'dodger-blue-light-60'
    | 'dodger-blue-light-75'
    | 'dodger-blue-light-90'
    | 'green'
    | 'green-dark-15'
    | 'green-dark-30'
    | 'green-dark-75'
    | 'green-light-30'
    | 'green-light-75'
    | 'green-light-90'
    | 'grey'
    | 'grey-dark-15'
    | 'grey-dark-30'
    | 'grey-dark-45'
    | 'grey-dark-60'
    | 'grey-dark-75'
    | 'grey-light-45'
    | 'grey-light-60'
    | 'grey-light-75'
    | 'grey-light-85'
    | 'grey-light-90'
    | 'grey-light-95'
    | 'grotesque-green'
    | 'grotesque-green-dark-30'
    | 'grotesque-green-dark-45'
    | 'grotesque-green-light-30'
    | 'grotesque-green-light-60'
    | 'grotesque-green-light-75'
    | 'grotesque-green-light-90'
    | 'orange'
    | 'orange-dark-15'
    | 'orange-dark-30'
    | 'orange-dark-75'
    | 'orange-light-30'
    | 'orange-light-75'
    | 'orange-light-90'
    | 'pink'
    | 'pink-dark-15'
    | 'pink-dark-30'
    | 'pink-dark-45'
    | 'pink-dark-75'
    | 'pink-light-30'
    | 'pink-light-45'
    | 'pink-light-75'
    | 'pink-light-90'
    | 'purple'
    | 'purple-dark-15'
    | 'purple-dark-30'
    | 'purple-light-30'
    | 'red'
    | 'red-dark-15'
    | 'red-dark-30'
    | 'red-dark-45'
    | 'red-dark-60'
    | 'red-dark-75'
    | 'red-dark-90'
    | 'red-light-15'
    | 'red-light-30'
    | 'red-light-45'
    | 'red-light-60'
    | 'red-light-75'
    | 'red-light-90'
    | 'smalt'
    | 'smalt-dark-30'
    | 'smalt-dark-75'
    | 'smalt-light-30'
    | 'smalt-light-45'
    | 'smalt-light-75'
    | 'smalt-light-90'
    | 'violet'
    | 'violet-dark-15'
    | 'violet-dark-30'
    | 'violet-dark-45'
    | 'violet-dark-60'
    | 'violet-dark-75'
    | 'violet-dark-90'
    | 'violet-light-15'
    | 'violet-light-30'
    | 'violet-light-45'
    | 'violet-light-60'
    | 'violet-light-75'
    | 'violet-light-90'
    | 'white'
    | 'yellow'
    | 'yellow-dark-15'
    | 'yellow-dark-30'
    | 'yellow-dark-75'
    | 'yellow-light-30'
    | 'yellow-light-45'
    | 'yellow-light-60'
    | 'yellow-light-75'
    | 'yellow-light-90'
    // Config for system colors
    | 'primary'
    | 'light-primary'
    | 'hover-primary'
    | 'active-primary'
    | 'primary-background'
    | 'primary-light-background'
    | 'primary-neutral-background'
    | 'primary-dark-background'
    | 'info'
    | 'info-border'
    | 'info-dark-background'
    | 'info-background'
    | 'danger'
    | 'danger-dark'
    | 'hover-danger'
    | 'active-danger'
    | 'danger-dark-background'
    | 'danger-background'
    | 'warning'
    | 'warning-dark'
    | 'hover-warning'
    | 'active-warning'
    | 'warning-dark-background'
    | 'warning-background'
    | 'success'
    | 'success-dark'
    | 'hover-success'
    | 'active-success'
    | 'success-dark-background'
    | 'success-background'
    | 'error'
    | 'error-dark'
    | 'hover-error'
    | 'active-error'
    | 'error-dark-background'
    | 'error-background'
    | 'link'
    | 'hover-link'
    | 'active-link'
    | 'text'
    | 'body-text'
    | 'subdued-text'
    | 'disabled-text'
    | 'light-text'
    | 'default-background'
    | 'default-light-background'
    | 'default-dark-background'
    | 'default-grey-background'
    | 'default-dark-grey-background'
    | 'neutral-background'
    | 'neutral-light-background'
    | 'neutral-dark-background'
    | 'default-border'
    | 'primary-header-background'
    | 'primary-ring-background'
    | 'info-header-background';
  fontSizes: keyof FontSizes;
  fontWeights: 'bold' | 'light' | 'regular' | 'semi-bold';
  lineHeights: keyof LineHeights;
  radii: 'base' | 'medium' | 'large';
  shadows: 'default' | 'primary' | 'danger' | 'warning' | 'divider';
  space: keyof Space;
}

interface SystemProperties {
  /**
   * Background color.
   */
  bgColor?: ThemeScales['colors'];
  /**
   * Border color.
   */
  borderColor?: ThemeScales['colors'];
  /**
   * Border radius.
   */
  borderRadius?: ThemeScales['radii'];
  /**
   * Border style.
   */
  borderStyle?: ThemeScales['borderStyles'];
  /**
   * Border width.
   */
  borderWidth?: ThemeScales['borderWidths'];
  /**
   * Box shadow.
   */
  boxShadow?: ThemeScales['shadows'];
  /**
   * Text color.
   */
  color?: ThemeScales['colors'];
  /**
   * Font size.
   */
  fontSize?: ThemeScales['fontSizes'];
  /**
   * Font weight.
   */
  fontWeight?: ThemeScales['fontWeights'];
  /**
   * Line height.
   */
  lineHeight?: ThemeScales['lineHeights'];
  /**
   * Margin.
   */
  m?: ThemeScales['space'];
  /**
   * Margin-bottom.
   */
  mb?: ThemeScales['space'];
  /**
   * Margin-left.
   */
  ml?: ThemeScales['space'];
  /**
   * Margin-right.
   */
  mr?: ThemeScales['space'];
  /**
   * Margin-top.
   */
  mt?: ThemeScales['space'];
  /**
   * Padding.
   */
  p?: ThemeScales['space'];
  /**
   * Padding-bottom.
   */
  pb?: ThemeScales['space'];
  /**
   * Padding-left.
   */
  pl?: ThemeScales['space'];
  /**
   * Padding-right.
   */
  pr?: ThemeScales['space'];
  /**
   * Padding-top.
   */
  pt?: ThemeScales['space'];
}

type SxKeys = keyof SystemProperties | keyof CSSProperties;
type SxValues = string | number;
type SxObject = Partial<Record<SxKeys, SxValues>>;

export type { ThemeScales, SystemProperties, SxKeys, SxValues, SxObject };
