// @hero-design/react/src/components/Tabs/StyledTabs.tsx
import styled, { css } from 'styled-components';
import { StyledMenu } from '../Menu/StyledMenu';

const TabWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const TabList = styled.div`
  position: relative;
  display: flex;
  margin: 0;
  padding: 0;

  &::before {
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    content: '';
    border-bottom: ${({ theme }) => theme.__hd__.tabs.borderWidths.default}
      solid ${({ theme }) => theme.__hd__.tabs.colors.border};
  }

  &:focus {
    outline: none;
  }
`;

const Tab = styled.button<{ themeVariant: 'basic' | 'active' | 'disabled' }>`
  position: relative;
  display: inline-flex;
  align-items: center;
  white-space: nowrap;
  background-color: ${({ theme }) => theme.__hd__.tabs.colors.background};
  border-radius: ${({ theme }) => theme.__hd__.tabs.radii.default};
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.tabs.space.padding};
  font-size: ${({ theme }) => theme.__hd__.tabs.fontSizes.default};
  line-height: ${({ theme }) => theme.__hd__.tabs.lineHeights.default};
  font-weight: ${({ theme }) => theme.__hd__.tabs.fontWeights.default};
  border: ${({ theme }) => theme.__hd__.tabs.borderWidths.default} solid
    ${({ theme }) => theme.__hd__.tabs.colors.border};
  margin-right: ${({ theme }) => theme.__hd__.tabs.space.marginRight};
  box-sizing: border-box;
  border-bottom: ${({ theme }) => theme.__hd__.tabs.borderWidths.default} solid
    ${({ theme }) => theme.__hd__.tabs.colors.border};
  &:focus {
    outline: none;
  }

  ${({ themeVariant, theme }) => {
    switch (themeVariant) {
      case 'basic':
        return css`
          color: ${theme.__hd__.tabs.colors.text};
          &:hover {
            color: ${theme.__hd__.tabs.colors.hoverText};
          }
          &:focus {
            color: ${theme.__hd__.tabs.colors.focusText};
          }
        `;
      case 'active':
        return css`
          background-color: ${theme.__hd__.tabs.colors.activeBackground};
          border-bottom: none;
          color: ${theme.__hd__.tabs.colors.activeText};
          &:focus {
            border-color: ${theme.__hd__.tabs.colors.focusBorder};
          }
        `;

      case 'disabled':
        return css`
          color: ${theme.__hd__.tabs.colors.disabledText};
        `;
    }
  }};
`;

const TruncatedTab = styled(Tab)`
  height: 100%;
  max-width: ${({ theme }) => theme.__hd__.tabs.sizes.maxWidth};
`;
const DropdownTabWraper = styled.div`
  height: 100%;
  display: flex;
`;

const TabTitle = styled.span`
  overflow: hidden;
  text-overflow: ellipsis;
`;

const TabTitleIcon = styled.span`
  display: inline-flex;
  margin: 0;
  padding: 0;
  margin-right: ${({ theme }) => theme.__hd__.tabs.space.iconMarginRight};
`;

const TabTitleArrowIcon = styled.span`
  display: inline-flex;
  margin: 0;
  padding: 0;
  margin-left: ${({ theme }) => theme.__hd__.tabs.space.iconMarginLeft};
`;

const TabPanel = styled.div`
  position: relative;
  border: ${({ theme }) => theme.__hd__.tabs.borderWidths.default} solid
    ${({ theme }) => theme.__hd__.tabs.colors.border};
  border-top: none;
  border-radius: ${({ theme }) => theme.__hd__.tabs.radii.panel};
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.tabs.space.panelPadding};
  color: ${({ theme }) => theme.__hd__.tabs.colors.panelText};
  background: ${({ theme }) => theme.__hd__.tabs.colors.panelBackground};
  font-size: ${({ theme }) => theme.__hd__.tabs.fontSizes.panel};
  line-height: ${({ theme }) => theme.__hd__.tabs.lineHeights.panel};
  font-weight: ${({ theme }) => theme.__hd__.tabs.fontWeights.panel};
`;

const OverflownMenu = styled(StyledMenu)`
  max-height: ${({ theme }) => theme.__hd__.tabs.sizes.moreTabMaxHeight};
  overflow-y: scroll;
`;

export {
  TabWrapper,
  TabList,
  Tab,
  TruncatedTab,
  TabTitle,
  TabTitleIcon,
  TabTitleArrowIcon,
  TabPanel,
  DropdownTabWraper,
  OverflownMenu,
};
