// @hero-design/react/src/components/Modal/StyledModal.tsx
import styled, { keyframes, css } from 'styled-components';

const fadeIn = keyframes`
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
`;

const zoomIn = keyframes`
  from {
    transform: scale(0.5);
  }

  to {
    transform: scale(1);
  }
`;

const ModalContainer = styled.div<{
  themeOpen: boolean;
}>`
  position: fixed;
  z-index: 9999;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  margin: 0;
  padding: 0;

  ${({ themeOpen }) => {
    switch (themeOpen) {
      case true:
        return css`
          display: block;
        `;
      case false:
        return css`
          display: none;
        `;
    }
  }};
`;

const ModalOverlay = styled.div`
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  overflow: auto;
  background-color: ${({ theme }) =>
    theme.__hd__.modal.colors.containerBackground};
  animation: ${fadeIn} 0.2s linear;
`;

const ModalContentWrapper = styled.div`
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  min-height: 100vh;
  pointer-events: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  padding-top: ${({ theme }) => theme.__hd__.modal.space.wrapperPaddingY};
  padding-bottom: ${({ theme }) => theme.__hd__.modal.space.wrapperPaddingY};
`;

const ModalContent = styled.div<{
  themeSize: 'small' | 'medium' | 'large' | 'xlarge' | 'xxlarge';
}>`
  display: flex;
  flex-direction: column;
  max-height: 100%;
  background-color: ${({ theme }) =>
    theme.__hd__.modal.colors.contentBackground};
  border-radius: ${({ theme }) => theme.__hd__.modal.radii.content};
  pointer-events: all;
  animation: ${zoomIn} 0.2s linear;
  min-width: ${({ theme }) => theme.__hd__.modal.sizes.contentSmallMinWidth};
  max-width: ${({ theme }) => theme.__hd__.modal.sizes.contentSmallMaxWidth};
  width: 100%;
  margin: 0;
  padding: 0;

  @media (min-width: ${({ theme }) => theme.breakpoints.medium}) {
    ${({ themeSize, theme }) => {
      switch (themeSize) {
        case 'small':
          return css`
            min-width: ${theme.__hd__.modal.sizes.contentSmallMinWidth};
            max-width: ${theme.__hd__.modal.sizes.contentSmallMaxWidth};
          `;
        case 'medium':
          return css`
            min-width: ${theme.__hd__.modal.sizes.contentMediumMinWidth};
            max-width: ${theme.__hd__.modal.sizes.contentMediumMaxWidth};
          `;
        case 'large':
        case 'xlarge':
        case 'xxlarge':
          return css`
            min-width: ${theme.__hd__.modal.sizes.contentLargeMinWidth};
            max-width: ${theme.__hd__.modal.sizes.contentLargeMaxWidth};
          `;
      }
    }};
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.large}) {
    ${({ themeSize, theme }) => {
      switch (themeSize) {
        case 'xlarge':
        case 'xxlarge':
          return css`
            min-width: ${theme.__hd__.modal.sizes.contentExtraLargeMinWidth};
            max-width: ${theme.__hd__.modal.sizes.contentExtraLargeMaxWidth};
          `;
        default:
          return css``;
      }
    }};
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.xlarge}) {
    ${({ themeSize, theme }) => {
      switch (themeSize) {
        case 'xxlarge':
          return css`
            min-width: ${theme.__hd__.modal.sizes
              .contentDoubleExtraLargeMinWidth};
            max-width: ${theme.__hd__.modal.sizes
              .contentDoubleExtraLargeMaxWidth};
          `;
        default:
          return css``;
      }
    }};
  }
`;

const ModalHeader = styled.div<{
  themeVariant?: 'basic' | 'primary';
}>`
  border-radius: ${({ theme }) => theme.__hd__.modal.radii.header};
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.modal.space.headerPadding};
  font-size: ${({ theme }) => theme.__hd__.modal.fontSizes.header};
  font-weight: ${({ theme }) => theme.__hd__.modal.fontWeights.header};
  line-height: ${({ theme }) => theme.__hd__.modal.lineHeights.header};
  display: flex;

  ${({ themeVariant, theme }) => {
    switch (themeVariant) {
      case 'basic':
        return css`
          background-color: ${theme.__hd__.modal.colors.headerBasicBackground};
          color: ${theme.__hd__.modal.colors.headerBasic};
        `;
      case 'primary':
        return css`
          background-color: ${theme.__hd__.modal.colors
            .headerPrimaryBackground};
          color: ${theme.__hd__.modal.colors.headerPrimary};
        `;
      case undefined:
        return css``;
    }
  }};
`;

const ModalBody = styled.div<{ themeIsPopUp: boolean }>`
  font-size: ${({ theme }) => theme.__hd__.modal.fontSizes.body};
  line-height: ${({ theme }) => theme.__hd__.modal.lineHeights.body};
  font-weight: ${({ theme }) => theme.__hd__.modal.fontWeights.body};
  color: ${({ theme }) => theme.__hd__.modal.colors.body};
  margin: 0;
  overflow: auto;

  ${({ themeIsPopUp, theme }) => {
    switch (themeIsPopUp) {
      case true:
        return css`
          padding: ${theme.__hd__.modal.space.popUpBodyPadding};
        `;
      case false:
        return css`
          padding: ${theme.__hd__.modal.space.bodyPadding};
        `;
    }
  }};
`;

const CloseButton = styled.button`
  display: flex;
  align-items: center;
  border: none;
  background: transparent;
  font-size: inherit;
  margin: 0;
  margin-left: auto;
  padding: 0;
  color: inherit;

  &:hover {
    cursor: pointer;
  }

  &:focus {
    outline: none;
  }
`;

const ModalFooter = styled.div`
  text-align: right;
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.modal.space.footerPadding};
`;

const HeaderFooterWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const PopUpModalBodyContainer = styled.div`
  display: flex;
  margin: 0;
  padding: 0;
`;

const PopUpModalBodyTextWrapper = styled.div`
  flex: 1;
  margin: 0;
  padding: 0;
`;

const TitleSpacing = styled.div`
  margin: 0;
  padding: 0;
  margin-bottom: ${({ theme }) =>
    theme.__hd__.modal.space.popUpTitleMarginBottom};
`;

const IconWrapper = styled.div`
  font-size: ${({ theme }) => theme.__hd__.modal.fontSizes.popUpIcon};
  margin: 0;
  padding: 0;
  margin-right: ${({ theme }) => theme.__hd__.modal.space.popUpIconMarginRight};
`;

export {
  ModalContainer,
  ModalOverlay,
  ModalContentWrapper,
  ModalContent,
  ModalHeader,
  ModalBody,
  CloseButton,
  ModalFooter,
  HeaderFooterWrapper,
  PopUpModalBodyContainer,
  PopUpModalBodyTextWrapper,
  TitleSpacing,
  IconWrapper,
};
