// @hero-design/react/src/components/Chart/BarChart/index.tsx
import React, {
  useRef,
  useEffect,
  useCallback,
  useState,
  useMemo,
} from 'react';
import * as d3 from 'd3';
import { useTheme } from 'styled-components';
import { CommonProps } from '../../common';
import {
  LegendConfig,
  NavigationConfig,
  Series,
  TickConfig,
  XAxisConfig,
  YAxisConfig,
  BarConfig,
  MatchedSegment,
  StyleConfig,
} from '../type';
import {
  StyledBarChartContainer,
  StyledLegendContainer,
} from './StyledBarChart';

import { useCss, useResizeObserver } from '../../../utils/hooks';
import Typography from '../../Typography';

import {
  difference,
  filterElementHasNoColor,
  getParentElement,
  totalOfMaxValuesFromEachRows,
  usePrevious,
} from '../utils';

import { useColorScale } from '../common/colorScale';
import { createNiceScale } from '../niceNumbers';
import BarArea from './ChartContent/BarArea';

import GridArea from './ChartContent/GridArea';
import LegendGroup from './ChartContent/LegendGroup';
import Navigation from './ChartContent/Navigation';
import XAxis from './ChartContent/XAxis';
import YAxisTitle from './ChartContent/YAxisTitle';
import YAxis from './ChartContent/YAxis';
import BarTooltip from './ChartContent/BarTooltip';

type DataValue = number | undefined;
export interface BarChartProps extends CommonProps {
  /**
   * Data of the chart.
   */
  data: Array<Series<Array<DataValue>>>;
  /**
   * X axis config.
   */
  xAxisConfig?: Omit<XAxisConfig, 'labels'> & { tick?: TickConfig };
  /**
   * Y axis config.
   */
  yAxisConfig?: Omit<YAxisConfig, 'maxValue'>;
  /**
   * Legend config.
   */
  legendConfig?: LegendConfig;
  /**
   * Bar section config.
   */
  barConfig?: BarConfig<DataValue>;
  /**
   * Navigation button config.
   */
  navigationConfig?: NavigationConfig;
  /**
   * Matched pattern to highlight the bars.
   */
  highlightedItems?: MatchedSegment;
  /**
   * styleConfig use to custom the style of the chart.
   * * styleConfig must be an object:
   *    * color?: use to custom the legend colors.
   */
  styleConfig?: StyleConfig;
  /**
   * Tooltip renderer.
   */
  tooltipRenderer?: (props: {
    label?: string;
    value?: DataValue;
    legendColor?: string;
    yAxisLabel?: string;
  }) => React.ReactNode;
}

type ElementSize = {
  height: number;
  width: number;
};

const BarChart = ({
  data,
  xAxisConfig,
  yAxisConfig,
  legendConfig,
  navigationConfig = { visible: false },
  barConfig,
  'data-test-id': dataTestId,
  style,
  sx,
  id,
  className,
  highlightedItems,
  styleConfig,
  tooltipRenderer,
}: BarChartProps) => {
  const niceValues = useMemo(() => {
    const maxDataValue = totalOfMaxValuesFromEachRows(data);
    const maxDataValuesWithShownTotal =
      typeof barConfig?.showTotal !== 'undefined'
        ? maxDataValue + 1
        : maxDataValue;

    return createNiceScale(0, maxDataValuesWithShownTotal);
  }, [barConfig?.showTotal, data]);

  const maxValue = Number(xAxisConfig?.maxValue ?? niceValues.niceMax);

  const isNegativeBarChart = useMemo(
    () => data.some((d) => d.data.some((v) => v !== undefined && v < 0)),
    [data]
  );

  const minValue = isNegativeBarChart ? -maxValue : 0;

  const defaultStep = xAxisConfig?.step || niceValues.tickSpacing;
  const tickInterval = xAxisConfig?.tick?.interval || defaultStep;

  const defaultBarSize = barConfig?.size || 'large';

  const theme = useTheme();

  const {
    __hd__: { barChart: barChartTheme },
  } = theme;

  const {
    barChartMarginBottom,
    barChartMarginLeft,
    barChartMarginRight,
    barChartMarginTop,
    barChartMarginLeftWithNavigation,
    gridChartMarginTop,
    legendContainerMarginTop,
    chartContentMarginLeft,
    navigationButtonMarginLeft,
  } = barChartTheme.space;

  const chartMargin = {
    top: barChartMarginTop,
    right: barChartMarginRight,
    bottom: barChartMarginBottom,
    left: yAxisConfig?.title
      ? barChartMarginLeft
      : barChartMarginLeftWithNavigation,
  };

  const containerRef = useRef<HTMLDivElement>(null);
  const tooltipRef = useRef<HTMLDivElement>(null);
  const barsContainerRef = useRef<SVGGElement>(null);
  const legendContainerRef = useRef<HTMLDivElement>(null);
  const axesRef = useRef<SVGGElement>(null);

  const [containerSize, setContainerSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });

  const [legendContainerSize, setLegendContainerSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });

  const [yAxisSize, setYAxisSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });
  const [yAxisTitleSize, setYAxisTitleSize] = useState<ElementSize>({
    height: 0,
    width: 0,
  });

  const [currentSelectedBar, setCurrentSelectedBar] = useState<
    | {
        legend: string;
        value: number;
        legendColor: string;
        percent: number;
        yAxisLabel: string;
      }
    | undefined
  >(undefined);

  const [internalSelectedLegends, setInternalSelectedLegends] = useState<
    string[]
  >([]);

  const legends = useMemo(() => data.map((item) => item.label), [data]);

  const defaultLegends = useMemo(
    () =>
      legendConfig?.selectable
        ? legendConfig?.value || internalSelectedLegends
        : internalSelectedLegends,
    [internalSelectedLegends, legendConfig]
  );

  const selectedLegends = useMemo(
    () => legends.filter((l) => defaultLegends.includes(l)),
    [legends, defaultLegends]
  );

  const svgSizes: ElementSize = useMemo(
    () => ({
      width: containerSize.width,
      height:
        containerSize.height -
        legendContainerSize.height -
        legendContainerMarginTop,
    }),
    [
      containerSize.height,
      containerSize.width,
      legendContainerMarginTop,
      legendContainerSize.height,
    ]
  );

  const yAxisWidth = useMemo(
    () => Math.floor(yAxisSize.width),
    [yAxisSize.width]
  );

  const yAxisTitleWidth = useMemo(
    () => Math.floor(yAxisTitleSize.width),
    [yAxisTitleSize.width]
  );

  const boundsHeight = svgSizes.height - chartMargin.top - chartMargin.bottom;

  const boundsWidth =
    svgSizes.width -
    chartMargin.right -
    chartMargin.left -
    yAxisWidth -
    yAxisTitleWidth;

  const handleOnResize = useCallback(() => {
    if (containerRef.current) {
      const { height, width } = containerRef.current.getBoundingClientRect();
      setContainerSize({ height, width });
    }

    if (legendContainerRef.current) {
      const { height, width } =
        legendContainerRef.current.getBoundingClientRect();
      setLegendContainerSize({ height, width });
    }
  }, []);

  useResizeObserver(handleOnResize, getParentElement(containerRef.current));

  const customColors = useMemo(
    () =>
      styleConfig?.series !== undefined
        ? filterElementHasNoColor(styleConfig.series)
        : undefined,
    [styleConfig?.series]
  );

  // Color Scale
  const colorScale = useColorScale(legends, customColors);

  const prevLegends = usePrevious(legends);

  const legendsDiff = useMemo(
    () => difference(legends, prevLegends || []),
    [legends, prevLegends]
  );

  useEffect(() => {
    setInternalSelectedLegends((prev) => {
      return Array.from(new Set([...prev, ...legendsDiff]));
    });
  }, [legendsDiff]);

  const onMouseOverBar = useCallback((_e, barData) => {
    const { color, legend, value, percent, yAxisLabel } = barData;
    setCurrentSelectedBar({
      legend,
      value,
      legendColor: color,
      percent,
      yAxisLabel,
    });
  }, []);

  const onMouseOutBar = useCallback(() => {
    setCurrentSelectedBar(undefined);
  }, []);

  const onMouseMoveBar = useCallback((e) => {
    const tooltipElement = d3.select(tooltipRef.current);
    const { clientX, clientY } = e;

    tooltipElement
      .style('top', `${Math.round(clientY - 12)}px`)
      .style('left', `${Math.round(clientX)}px`);
  }, []);

  const barAreaData = useMemo(
    () =>
      data.map((d) => {
        return {
          ...d,
          color: colorScale(d.label),
        };
      }),
    [colorScale, data]
  );

  return (
    <StyledBarChartContainer
      ref={containerRef}
      data-test-id={dataTestId}
      style={{ ...style, ...useCss({ ...sx }) }}
      id={id}
      className={className}
    >
      {navigationConfig.visible && (
        <Navigation
          top={boundsHeight / 2}
          left={
            yAxisConfig?.title
              ? chartContentMarginLeft - navigationButtonMarginLeft
              : 0
          }
          onClickNavigation={navigationConfig.onNavigate}
          data-test-id={dataTestId}
          disabledButtons={navigationConfig.disabledButtons}
        />
      )}
      <BarTooltip
        show={!!currentSelectedBar}
        data-test-id={dataTestId}
        ref={tooltipRef}
        color={currentSelectedBar?.legendColor}
        title={currentSelectedBar?.legend}
        value={currentSelectedBar?.value}
        percent={currentSelectedBar?.percent}
        tooltipRenderer={tooltipRenderer}
        yAxisLabel={currentSelectedBar?.yAxisLabel}
      />
      <svg width={svgSizes.width} height={svgSizes.height}>
        <g
          ref={barsContainerRef}
          transform={`translate(${yAxisWidth + chartMargin.left},${
            chartMargin.top + gridChartMarginTop
          })`}
        >
          <YAxisTitle
            height={boundsHeight}
            setSize={setYAxisTitleSize}
            yAxisWidth={yAxisWidth}
            yAxisConfig={yAxisConfig}
          />
        </g>
        <g
          ref={axesRef}
          transform={`translate(${
            yAxisWidth + yAxisTitleWidth + chartContentMarginLeft
          },${chartMargin.top})`}
        >
          <GridArea
            height={boundsHeight}
            width={boundsWidth}
            minValue={minValue}
            maxValue={maxValue}
            tickInterval={tickInterval}
            highlightMiddleLine={isNegativeBarChart}
            data-test-id={dataTestId ? `${dataTestId}-grid-line` : undefined}
          />
          <BarArea
            height={boundsHeight}
            width={boundsWidth}
            maxValue={maxValue}
            minValue={minValue}
            data-test-id={dataTestId}
            yAxisConfig={yAxisConfig}
            barConfig={barConfig}
            onMouseOver={onMouseOverBar}
            onMouseOut={onMouseOutBar}
            onMouseMove={onMouseMoveBar}
            highlightedItems={highlightedItems}
            selectedLegends={selectedLegends}
            data={barAreaData}
            barSize={defaultBarSize}
          />
          <XAxis
            height={boundsHeight}
            width={boundsWidth}
            maxValue={maxValue}
            minValue={minValue}
            defaultStep={defaultStep}
          />
          <YAxis
            dataLength={data.length}
            height={boundsHeight}
            getAxisSize={setYAxisSize}
            yAxisConfig={yAxisConfig}
            barSize={defaultBarSize}
          />
        </g>
      </svg>

      <StyledLegendContainer ref={legendContainerRef}>
        {!!xAxisConfig?.title && (
          <Typography.Text fontSize={12} fontWeight="regular" intent="subdued">
            {xAxisConfig.title}
          </Typography.Text>
        )}
        <LegendGroup
          legends={legends}
          onChange={setInternalSelectedLegends}
          selectedLegends={selectedLegends}
          data-test-id={dataTestId ? `${dataTestId}-legend` : undefined}
          legendConfig={legendConfig}
          colorScale={colorScale}
        />
      </StyledLegendContainer>
    </StyledBarChartContainer>
  );
};

export default BarChart;
