// @hero-design/react/src/components/Tooltip/StyledTooltip.tsx
import styled, { css, keyframes } from 'styled-components';

const Arrow = styled.div<{ variant: 'dark' | 'light' }>`
  height: ${({ theme }) => theme.__hd__.tooltip.sizes.arrow};
  width: ${({ theme }) => theme.__hd__.tooltip.sizes.arrow};
  position: absolute;
  margin: 0;
  padding: 0;

  &::before {
    box-sizing: border-box;
    position: absolute;
    transform: rotate(45deg);
    height: ${({ theme }) => theme.__hd__.tooltip.sizes.arrow};
    width: ${({ theme }) => theme.__hd__.tooltip.sizes.arrow};
    background: ${({ theme, variant }) =>
      theme.__hd__.tooltip.colors[variant].background};
    content: '';
  }
`;

const delayShowFrame = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

const TooltipWrapper = styled.div<{
  variant: 'dark' | 'light';
  themeDelay?: number;
}>`
  z-index: 9999;
  color: ${({ theme, variant }) => theme.__hd__.tooltip.colors[variant].text};
  background: ${({ theme, variant }) =>
    theme.__hd__.tooltip.colors[variant].background};
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.tooltip.space.wrapperPadding};
  font-size: ${({ theme }) => theme.__hd__.tooltip.fontSizes.default};
  font-weight: ${({ theme }) => theme.__hd__.tooltip.fontWeights.default};
  line-height: ${({ theme }) => theme.__hd__.tooltip.lineHeights.default};
  border-radius: ${({ theme }) => theme.__hd__.tooltip.radii.default};
  text-align: left;
  border-width: ${({ theme }) => theme.__hd__.tooltip.borderWidths.default};
  border-style: solid;
  border-color: ${({ theme, variant }) =>
    theme.__hd__.tooltip.colors[variant].border};
  box-sizing: border-box;
  box-shadow: ${({ theme, variant }) => theme.__hd__.tooltip.shadows[variant]};

  ${({ themeDelay }) => {
    if (themeDelay) {
      return css`
        opacity: 0;
        animation: ${delayShowFrame} 0s linear ${themeDelay}ms 1 forwards;
      `;
    }
  }};
`;

const TooltipContainer = styled.div`
  display: inline-block;
  margin: 0;
  padding: 0;
`;

export { TooltipContainer, TooltipWrapper, Arrow };
