// @hero-design/react/src/components/Menu/StyledMenu.tsx
import styled, { css } from 'styled-components';

const StyledMenuItem = styled.li<{
  themeIntent: 'text' | 'danger';
  themeState: 'default' | 'disabled' | 'active' | 'focused';
}>`
  display: flex;
  flex: 1;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.menu.space.itemPadding};
  font-size: ${({ theme }) => theme.__hd__.menu.fontSizes.item};
  font-weight: ${({ theme }) => theme.__hd__.menu.fontWeights.item};
  line-height: ${({ theme }) => theme.__hd__.menu.lineHeights.item};

  & > *:not(:first-child):last-child {
    margin-left: ${({ theme }) =>
      theme.__hd__.menu.space.rightElementMarginLeft};
  }

  &:hover {
    background: ${({ theme }) => theme.__hd__.menu.colors.itemActiveBackground};
  }

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'text':
        return css`
          color: ${theme.__hd__.menu.colors.item};
        `;
      case 'danger':
        return css`
          color: ${theme.__hd__.menu.colors.itemDanger};
        `;
    }
  }};

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'default':
        return css``;
      case 'disabled':
        return css`
          color: ${theme.__hd__.menu.colors.itemDisabled};
          cursor: not-allowed;
          pointer-events: none;
          &:hover {
            background: ${theme.__hd__.menu.colors.background};
          }
        `;
      case 'focused':
        return css`
          background: ${theme.__hd__.menu.colors.itemActiveBackground};
        `;
      case 'active':
        return css`
          background: ${theme.__hd__.menu.colors.itemActiveBackground};
          font-weight: ${theme.__hd__.menu.fontWeights.itemActive};
        `;
    }
  }};
`;

const LeftElement = styled.div`
  display: flex;
  align-items: center;
  flex: 1;
  margin: 0;
  padding: 0;
`;

const MenuIcon = styled.span`
  display: inline-flex;
  margin: 0;
  padding: 0;
  margin-right: ${({ theme }) => theme.__hd__.menu.space.iconMarginRight};
  font-size: ${({ theme }) => theme.__hd__.menu.fontSizes.icon};
  line-height: ${({ theme }) => theme.__hd__.menu.lineHeights.icon};
`;

const StyledMenu = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
  background: ${({ theme }) => theme.__hd__.menu.colors.background};
`;

export { StyledMenuItem, LeftElement, MenuIcon, StyledMenu };
