// @hero-design/react/src/components/Filters/StyledFilters.tsx
import styled, { css } from 'styled-components';

import { StyledIcon } from '../Icon/StyledIcon';
import { DropdownWrapper } from '../Dropdown/StyledDropdown';
import StyledInput, { InputWrapper } from '../Input/StyledInput';
import { SelectWrapper } from '../Select/StyledSelect';
import { InputContainer } from '../Select/SingleSelect/Styled';
import { FilterSize } from './types';

const LabelWrapper = styled.span`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.filters.space.filterButtonPadding};
  padding-left: 0;
  font-size: ${({ theme }) => theme.__hd__.filters.fontSizes.filterButton};
  font-weight: ${({ theme }) => theme.__hd__.filters.fontWeights.filterBy};
  line-height: 100%;
  white-space: nowrap;
`;

const FilterWrapper = styled.div<{ themeActive: boolean }>`
  margin: 0;
  padding: 0;
  & > ${SelectWrapper} > ${DropdownWrapper} > ${InputWrapper} {
    font-weight: ${({ theme }) =>
      theme.__hd__.filters.fontWeights.filterButton};

    & > ${InputContainer} > ${StyledInput} {
      &::placeholder {
        font-weight: ${({ theme }) =>
          theme.__hd__.filters.fontWeights.filterButton};
        color: ${({ theme }) => theme.__hd__.filters.colors.filterButtonText};
      }
    }

    &:hover {
      border-color: ${({ theme }) =>
        theme.__hd__.filters.colors.filterButtonHoverBorder};
    }

    &:focus-within {
      border-color: ${({ theme }) =>
        theme.__hd__.filters.colors.filterButtonHoverBorder};
      box-shadow: none;
    }
  }

  ${({ theme, themeActive }) =>
    themeActive
      ? css`
          & > ${SelectWrapper} > ${DropdownWrapper} > ${InputWrapper} {
            background-color: ${theme.__hd__.filters.colors
              .filterButtonActiveBackground};

            & > ${InputContainer} > ${StyledInput} {
              background-color: ${theme.__hd__.filters.colors
                .filterButtonActiveBackground};
            }
          }
        `
      : css``};
`;

const StyledDropdownButton = styled.button<{
  themeActive: boolean;
  themeSize: FilterSize;
}>`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.filters.space.filterButtonPadding};
  height: ${({ theme, themeSize }) =>
    theme.__hd__.filters.sizes.filterButton[themeSize]};
  border: ${({ theme }) =>
    `${theme.__hd__.filters.borderWidths.filterButton} solid`};
  font-size: ${({ theme, themeSize }) =>
    theme.__hd__.filters.fontSizes.filterButton[themeSize]};
  font-weight: ${({ theme }) => theme.__hd__.filters.fontWeights.filterButton};
  line-height: ${({ theme, themeSize }) =>
    theme.__hd__.filters.lineHeights.filterButton[themeSize]};
  display: inline-flex;
  align-items: center;
  background-color: ${({ theme }) =>
    theme.__hd__.filters.colors.filterButtonBackground};
  cursor: pointer;
  position: relative;
  border-radius: ${({ theme }) => theme.__hd__.filters.radii.filterButton};
  color: ${({ theme }) => theme.__hd__.filters.colors.filterButtonText};
  border-color: ${({ theme }) =>
    theme.__hd__.filters.colors.filterButtonBorder};

  &:hover {
    z-index: 1;
    border-color: ${({ theme }) =>
      theme.__hd__.filters.colors.filterButtonHoverBorder};
  }

  ${({ theme, themeActive }) =>
    themeActive
      ? css`
          z-index: 1;
          background-color: ${theme.__hd__.filters.colors
            .filterButtonActiveBackground};
        `
      : css`
          background-color: ${theme.__hd__.filters.colors
            .filterButtonBackground};
        `};

  > ${StyledIcon} {
    padding-left: ${({ theme }) =>
      theme.__hd__.filters.space.filterButtonIconPadding};
  }
`;

const StyledTagList = styled.div`
  margin: 0;
  padding: 0;
  margin-top: ${({ theme }) => theme.__hd__.filters.space.tagListMarginTop};
`;

const FiltersWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const StyledFilterList = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  flex-wrap: wrap;
  gap: ${({ theme }) => theme.__hd__.filters.space.filterButtonGap};
`;

const StyledFiltersHeader = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  justify-content: space-between;
  margin-bottom: ${({ theme }) => theme.space.small}px;
`;

const StyledFilterListWithAddons = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
`;

export {
  LabelWrapper,
  FilterWrapper,
  StyledDropdownButton,
  StyledTagList,
  FiltersWrapper,
  StyledFilterList,
  StyledFiltersHeader,
  StyledFilterListWithAddons,
};
