// @hero-design/colors/src/defaultMobilePalette/index.ts
import colorScales from '../colorScales';

const {
  aliceBlue,
  amaranthaRed,
  antiFlashWhite,
  apple,
  cumberlandFog,
  deepSaffron,
  emerald,
  gentianFlower,
  grey,
  honeydew,
  linen,
  maasstrichtBlue,
  mellowApricot,
  orangeBrown,
  pastelRed,
  pineGreen,
  seashell,
  silverChalice,
  sonicSilver,
  ultramarineBlue,
  vermilion,
  vodka,
  windsorGrey,
  black,
} = colorScales;

const palette = {
  black: black.base,
  white: '#ffffff',

  greyLight95: grey.lighten95,
  greyLight90: grey.lighten90,
  greyLight75: grey.lighten75,
  greyLight45: grey.lighten45,
  greyLight30: grey.lighten30,
  grey: grey.base,

  maasstrichtBlueLight90: maasstrichtBlue.lighten90,
  maasstrichtBlueLight80: maasstrichtBlue.lighten80,
  maasstrichtBlueLight50: maasstrichtBlue.lighten50,
  maasstrichtBlueLight30: maasstrichtBlue.lighten30,
  maasstrichtBlue: maasstrichtBlue.base,

  pastelRed: pastelRed.base,
  pastelRedLight20: pastelRed.lighten20,

  mellowApricot: mellowApricot.base,
  mellowApricotLight20: mellowApricot.lighten20,

  emerald: emerald.base,
  emeraldLight20: emerald.lighten20,
  emeraldDark20: emerald.darken20,

  vodka: vodka.base,
  vodkaLight20: vodka.lighten20,

  vermilion: vermilion.base,

  deepSaffron: deepSaffron.base,
  deepSaffronDark20: deepSaffron.darken20,

  pineGreen: pineGreen.base,
  ultramarineBlue: ultramarineBlue.base,
  sonicSilver: sonicSilver.base,
  linen: linen.base,
  seashell: seashell.base,
  honeydew: honeydew.base,
  aliceBlue: aliceBlue.base,
  antiFlashWhite: antiFlashWhite.base,
  apple: apple.base,

  silverChalice: silverChalice.base,
  silverChaliceLight20: silverChalice.lighten20,

  amaranthaRed: amaranthaRed.base,
  orangeBrown: orangeBrown.base,
  gentianFlower: gentianFlower.base,
  windsorGrey: windsorGrey.base,
  cumberlandFog: cumberlandFog.base,
};

export default palette;
