// @hero-design/react/src/components/Tabs/SubTabs/StyledSubTabs.tsx
import styled, { css } from 'styled-components';
import { StyledMenu } from '../../Menu/StyledMenu';

const TabWrapper = styled.div`
  margin: 0;
  padding: 0;
  position: relative;
`;

const TabList = styled.div`
  position: relative;
  margin: 0;
  padding: 0;
  overflow: hidden;
  display: flex;

  &:focus {
    outline: none;
  }

  border-bottom: ${({ theme }) => theme.__hd__.tabs.borderWidths.default} solid
    ${({ theme }) => theme.__hd__.tabs.colors.border};
`;

const TabListWrapper = styled.div`
  margin: 0;
  padding: 0;
  position: relative;
`;

const ActiveTabIndicator = styled.div`
  position: absolute;
  bottom: 0;
  height: ${({ theme }) => theme.__hd__.tabs.sizes.activeTabIndicatorHeight};
  background: ${({ theme }) => theme.__hd__.tabs.colors.focusBorder};
  transition: width 0.3s, left 0.3s, right 0.3s;
  margin: 0;
  padding: 0;
`;

const Tab = styled.button<{
  themeVariant: 'basic' | 'active' | 'disabled';
  themeSize: 'small' | 'medium';
}>`
  &:focus {
    outline: none;
  }

  position: relative;
  background: transparent;
  border: none;
  font-size: ${({ theme, themeSize }) =>
    theme.__hd__.tabs.fontSizes.subTabTitles[themeSize]};
  line-height: ${({ theme, themeSize }) =>
    theme.__hd__.tabs.lineHeights.subTabTitles[themeSize]};
  font-weight: ${({ theme }) => theme.__hd__.tabs.fontWeights.subTabTitle};
  margin: 0;
  padding: ${({ theme, themeSize }) =>
    theme.__hd__.tabs.space.subPadding[themeSize]};
  display: inline-flex;
  align-items: center;
  white-space: nowrap;
  ${({ themeVariant, theme }) => {
    switch (themeVariant) {
      case 'basic':
        return css`
          color: ${theme.__hd__.tabs.colors.text};
          &:hover {
            color: ${theme.__hd__.tabs.colors.hoverText};
          }
          &:focus {
            color: ${theme.__hd__.tabs.colors.focusText};
          }
        `;
      case 'active':
        return css`
          color: ${theme.__hd__.tabs.colors.activeText};
          &:focus {
            color: ${theme.__hd__.tabs.colors.subFocusText};
            & ~ ${ActiveTabIndicator} {
              background-color: ${theme.__hd__.tabs.colors.subFocusText};
            }
          }
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.tabs.colors.disabledText};
        `;
    }
  }};
`;

const TabPanel = styled.div<{
  themeSize: 'small' | 'medium';
}>`
  margin: 0;
  padding: ${({ theme, themeSize }) =>
    theme.__hd__.tabs.space.subTabPanelPadding[themeSize]};
`;

const TabTitleIcon = styled.span`
  display: inline-flex;
  margin: 0;
  padding: 0;
  margin-right: ${({ theme }) => theme.__hd__.tabs.space.iconMarginRight};
`;

const TruncatedTab = styled(Tab)`
  height: 100%;
  max-width: ${({ theme }) => theme.__hd__.tabs.sizes.maxWidth};
`;

const DropdownTabWraper = styled.div`
  height: 100%;
  display: flex;
`;

const TabTitle = styled.span`
  overflow: hidden;
  text-overflow: ellipsis;
`;

const OverflownMenu = styled(StyledMenu)`
  max-height: ${({ theme }) => theme.__hd__.tabs.sizes.moreTabMaxHeight};
  overflow-y: scroll;
`;

const TabTitleArrowIcon = styled.span`
  display: inline-flex;
  margin: 0;
  padding: 0;
  margin-left: ${({ theme }) => theme.__hd__.tabs.space.iconMarginLeft};
`;

export {
  TabWrapper,
  TabList,
  Tab,
  TabPanel,
  ActiveTabIndicator,
  TabTitleIcon,
  TruncatedTab,
  DropdownTabWraper,
  TabTitle,
  OverflownMenu,
  TabTitleArrowIcon,
  TabListWrapper,
};
