// @hero-design/react/src/components/Chart/PieChart/StyledPieChart.tsx
import styled from 'styled-components';

const StyledChartContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: fit-content;
  background-color: ${({ theme }) => theme.__hd__.chart.colors.pie.background};
`;

const StyledPieChartContainer = styled.div<{
  themeSize: 'small' | 'medium' | 'large';
}>`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledIconContainer = styled.div`
  background-color: ${({ theme }) =>
    theme.__hd__.chart.colors.pie.iconBackground};
  border-radius: ${({ theme }) =>
    theme.__hd__.chart.radii.pie.iconContainerRadius};
  width: ${({ theme }) => theme.__hd__.chart.sizes.pie.iconContainerSize};
  height: ${({ theme }) => theme.__hd__.chart.sizes.pie.iconContainerSize};
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledInnerContentContainer = styled.div`
  position: absolute;
  top: 15%;
  bottom: 15%;
  right: 15%;
  left: 15%;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 9999px;
  overflow: hidden;
`;

export {
  StyledPieChartContainer,
  StyledInnerContentContainer,
  StyledChartContainer,
  StyledIconContainer,
};
