// @hero-design/react/src/components/Alert/StyledAlert.tsx
import styled, { css } from 'styled-components';

const StyledIconWrapper = styled.div<{
  themeSize: 'default' | 'compact';
}>`
  display: flex;
  align-items: center;
  margin: 0;
  padding: 0;
  color: inherit;

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'default':
        return css`
          margin-top: ${theme.__hd__.alert.space.iconDefaultMarginTop};
          margin-right: ${theme.__hd__.alert.space.iconDefaultMarginRight};
          line-height: ${theme.__hd__.alert.lineHeights.iconDefault};
        `;
      case 'compact':
        return css`
          margin-top: ${theme.__hd__.alert.space.iconCompactMarginTop};
          margin-right: ${theme.__hd__.alert.space.iconCompactMarginRight};
          line-height: ${theme.__hd__.alert.lineHeights.iconCompact};
        `;
    }
  }};
`;

const StyledInner = styled.div`
  flex: 1;
  margin: 0;
  padding: 0;
`;

const StyledTitle = styled.div`
  margin: 0;
  padding: 0;
  font-size: ${({ theme }) => theme.__hd__.alert.fontSizes.title};
  font-weight: ${({ theme }) => theme.__hd__.alert.fontWeights.title};
  line-height: ${({ theme }) => theme.__hd__.alert.lineHeights.title};
  margin-bottom: ${({ theme }) => theme.__hd__.alert.space.titleMarginBottom};
`;

const StyledContent = styled.div`
  margin: 0;
  padding: 0;
  font-size: ${({ theme }) => theme.__hd__.alert.fontSizes.content};
  line-height: ${({ theme }) => theme.__hd__.alert.lineHeights.content};
`;

const StyledAlert = styled.div<{
  themeIntent: 'success' | 'info' | 'warning' | 'danger' | 'error';
  themeSize: 'default' | 'compact';
}>`
  display: flex;
  align-items: flex-start;
  border-style: solid;
  margin: 0;

  font-weight: ${({ theme }) => theme.__hd__.alert.fontWeights.text};
  border-radius: ${({ theme }) => theme.__hd__.alert.radii.wrapper};
  color: ${({ theme }) => theme.__hd__.alert.colors.text};
  border-width: ${({ theme }) => theme.__hd__.alert.borderWidths.wrapper};

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'default':
        return css`
          padding: ${theme.__hd__.alert.space.defaultPadding};
        `;
      case 'compact':
        return css`
          padding: ${theme.__hd__.alert.space.compactPadding};
        `;
    }
  }};

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'success':
        return css`
          border-color: ${theme.__hd__.alert.colors.borderSuccess};
          background-color: ${theme.__hd__.alert.colors.bgSuccess};
        `;
      case 'info':
        return css`
          border-color: ${theme.__hd__.alert.colors.borderInfo};
          background-color: ${theme.__hd__.alert.colors.bgInfo};
        `;
      case 'warning':
        return css`
          border-color: ${theme.__hd__.alert.colors.borderWarning};
          background-color: ${theme.__hd__.alert.colors.bgWarning};
        `;
      case 'danger':
        return css`
          border-color: ${theme.__hd__.alert.colors.borderDanger};
          background-color: ${theme.__hd__.alert.colors.bgDanger};
        `;
      case 'error':
        return css`
          border-color: ${theme.__hd__.alert.colors.borderError};
          background-color: ${theme.__hd__.alert.colors.bgError};
        `;
    }
  }};

  ${StyledTitle}, ${StyledContent} {
    ${({ themeIntent, theme }) => {
      switch (themeIntent) {
        case 'success':
          return css`
            color: ${theme.__hd__.alert.colors.textSuccess};
          `;
        case 'info':
          return css`
            color: ${theme.__hd__.alert.colors.textInfo};
          `;
        case 'warning':
          return css`
            color: ${theme.__hd__.alert.colors.textWarning};
          `;
        case 'danger':
          return css`
            color: ${theme.__hd__.alert.colors.textDanger};
          `;
        case 'error':
          return css`
            color: ${theme.__hd__.alert.colors.textError};
          `;
      }
    }};
  }
`;

const StyledCloseButton = styled.div<{
  themeSize: 'default' | 'compact';
}>`
  margin: 0;
  padding: 0;
  margin-left: auto;
  color: inherit;

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'default':
        return css`
          margin-top: ${theme.__hd__.alert.space.iconDefaultMarginTop};
          line-height: ${theme.__hd__.alert.lineHeights.iconDefault};
        `;
      case 'compact':
        return css`
          margin-right: ${theme.__hd__.alert.space.iconCompactMarginRight};
          line-height: ${theme.__hd__.alert.lineHeights.iconCompact};
        `;
    }
  }};
`;

export {
  StyledAlert,
  StyledTitle,
  StyledInner,
  StyledContent,
  StyledIconWrapper,
  StyledCloseButton,
};
