// @hero-design/react/src/components/Typography/Text/StyledText.tsx
import styled, { css } from 'styled-components';
import type { LabelHTMLAttributes } from 'react';

export interface StyledProps {
  themeFontSize: 10 | 12 | 14 | 16 | 18;
  themeFontWeight: 'light' | 'regular' | 'semi-bold' | 'bold';
  themeIntent:
    | 'main'
    | 'body'
    | 'subdued'
    | 'disabled'
    | 'primary'
    | 'success'
    | 'danger'
    | 'warning'
    | 'error'
    | 'white';
}

const textCss = css<StyledProps>`
  margin: 0;
  padding: 0;

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'main':
        return css`
          color: ${theme.__hd__.typography.colors.main};
        `;
      case 'body':
        return css`
          color: ${theme.__hd__.typography.colors.body};
        `;
      case 'subdued':
        return css`
          color: ${theme.__hd__.typography.colors.subdued};
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.typography.colors.disabled};
        `;
      case 'primary':
        return css`
          color: ${theme.__hd__.typography.colors.primary};
        `;
      case 'success':
        return css`
          color: ${theme.__hd__.typography.colors.success};
        `;
      case 'danger':
        return css`
          color: ${theme.__hd__.typography.colors.danger};
        `;
      case 'warning':
        return css`
          color: ${theme.__hd__.typography.colors.warning};
        `;
      case 'error':
        return css`
          color: ${theme.__hd__.typography.colors.error};
        `;
      case 'white':
        return css`
          color: ${theme.__hd__.typography.colors.white};
        `;
    }
  }};

  ${({ themeFontWeight, theme }) => {
    switch (themeFontWeight) {
      case 'light':
        return css`
          font-weight: ${theme.__hd__.typography.fontWeights.light};
        `;
      case 'regular':
        return css`
          font-weight: ${theme.__hd__.typography.fontWeights.regular};
        `;
      case 'semi-bold':
        return css`
          font-weight: ${theme.__hd__.typography.fontWeights.semiBold};
        `;
      case 'bold':
        return css`
          font-weight: ${theme.__hd__.typography.fontWeights.bold};
        `;
    }
  }};

  ${({ themeFontSize, theme }) => {
    switch (themeFontSize) {
      case 10:
        return css`
          font-size: ${theme.__hd__.typography.fontSizes.xsmall};
          line-height: ${theme.__hd__.typography.lineHeights.xsmall};
        `;
      case 12:
        return css`
          font-size: ${theme.__hd__.typography.fontSizes.small};
          line-height: ${theme.__hd__.typography.lineHeights.small};
        `;
      case 14:
        return css`
          font-size: ${theme.__hd__.typography.fontSizes.medium};
          line-height: ${theme.__hd__.typography.lineHeights.medium};
        `;
      case 16:
        return css`
          font-size: ${theme.__hd__.typography.fontSizes.large};
          line-height: ${theme.__hd__.typography.lineHeights.large};
        `;
      case 18:
        return css`
          font-size: ${theme.__hd__.typography.fontSizes.xlarge};
          line-height: ${theme.__hd__.typography.lineHeights.xlarge};
        `;
    }
  }};
`;

// Suport inline elements:
// * Inline elements:
// * * <strong>
// * * <b>
// * * <label>
// * * <a>
// * * <span>
// * * <br>
// * * <em>
// * * <i>
const nestedInlineStyle = css`
  > strong,
  > b,
  > label {
    margin: 0;
    padding: 0;

    font-weight: ${({ theme }) =>
      theme.__hd__.typography.fontWeights.nestedStrongTag};
  }

  > em,
  > i,
  > span,
  > br {
    margin: 0;
    padding: 0;
  }

  > a {
    margin: 0;
    padding: 0;

    text-decoration: none;

    color: ${({ theme }) => theme.__hd__.typography.colors.nestedLink};
    &:hover,
    &:focus {
      color: ${({ theme }) => theme.__hd__.typography.colors.hoverNestedLink};
    }
  }
`;
const StyledParagraphBlock = styled.p<StyledProps>`
  ${textCss};
  ${nestedInlineStyle};
`;

const StyledDivBlock = styled.div<StyledProps>`
  ${textCss};
  ${nestedInlineStyle};

  > p {
    ${textCss};
    ${nestedInlineStyle};
  }

  > ul {
    margin: 0;
    padding: 0;
    padding-inline-start: ${({ theme }) =>
      theme.__hd__.typography.space.nestedUnorderedListPaddingLeft};

    > li {
      margin: 0;
      padding: 0;
    }
  }
`;

const StyledSpanInline = styled.span<StyledProps>`
  ${textCss};
  ${nestedInlineStyle};
`;

const StyledLabelInline = styled.label<
  StyledProps & LabelHTMLAttributes<HTMLLabelElement>
>`
  ${textCss};
  ${nestedInlineStyle};
`;

export {
  StyledParagraphBlock,
  StyledDivBlock,
  StyledSpanInline,
  StyledLabelInline,
};
