// @hero-design/react/src/components/Result/StyledResult.tsx
import styled from 'styled-components';

const IconWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const ResultWrapper = styled.div`
  text-align: center;
  margin: 0;
  padding: 0;

  ${IconWrapper} > i {
    font-size: ${({ theme }) => theme.__hd__.result.fontSizes.icon};
  }
`;

const ResultImg = styled.img`
  margin: 0;
  padding: 0;
`;

const TitleWrapper = styled.div`
  color: ${({ theme }) => theme.__hd__.result.colors.default};
  font-size: ${({ theme }) => theme.__hd__.result.fontSizes.title};
  font-weight: ${({ theme }) => theme.__hd__.result.fontWeights.title};
  line-height: ${({ theme }) => theme.__hd__.result.lineHeights.title};
  margin: ${({ theme }) => theme.__hd__.result.space.titleMargin};
  padding: 0;
`;

const SubTitleWrapper = styled.div`
  color: ${({ theme }) => theme.__hd__.result.colors.default};
  font-size: ${({ theme }) => theme.__hd__.result.fontSizes.subTitle};
  font-weight: ${({ theme }) => theme.__hd__.result.fontWeights.subTitle};
  line-height: ${({ theme }) => theme.__hd__.result.lineHeights.subTitle};
  margin: 0;
  padding: 0;
`;

const ExtraWrapper = styled.div`
  margin: 0;
  padding: 0;
  margin-top: ${({ theme }) => theme.__hd__.result.space.extraMargin};
`;

export {
  ResultWrapper,
  ResultImg,
  TitleWrapper,
  SubTitleWrapper,
  ExtraWrapper,
  IconWrapper,
};
