// @hero-design/react/src/theme/getTheme.ts
import { defaultScale, heroSystemPalette, getGlobalTheme } from './global';

import { getAlertTheme } from './components/alert';
import { getBadgeTheme } from './components/badge';
import { getBannerTheme } from './components/banner';
import { getBreadcrumbTheme } from './components/breadcrumb';
import { getButtonTheme } from './components/button';
import { getCardTheme } from './components/card';
import { getCarouselTheme } from './components/carousel';
import { getColumnChartTheme } from './components/columnChart';
import { getCheckboxTheme } from './components/checkbox';
import { getCommentTheme } from './components/comment';
import { getDatePickerTheme } from './components/datePicker';
import { getDividerTheme } from './components/divider';
import { getDropdownTheme } from './components/dropdown';
import { getEmptyTheme } from './components/empty';
import { getFileTheme } from './components/file';
import { getFormTheme } from './components/form';
import { getIconTheme } from './components/icon';
import { getInPageNavigationTheme } from './components/inPageNavigation';
import { getInputTheme } from './components/input';
import { getLegendTheme } from './components/legend';
import { getMenuTheme } from './components/menu';
import { getModalTheme } from './components/modal';
import { getNotificationTheme } from './components/notification';
import { getPageHeaderTheme } from './components/pageHeader';
import { getPaginationTheme } from './components/pagination';
import { getPortletTheme } from './components/portlet';
import { getProgressTheme } from './components/progress';
import { getRadioTheme } from './components/radio';
import { getRateTheme } from './components/rate';
import { getResultTheme } from './components/result';
import { getSelectTheme } from './components/select';
import { getSelectButtonTheme } from './components/selectButton';
import { getSideBarTheme } from './components/sideBar';
import { getSliderTheme } from './components/slider';
import { getSpinnerTheme } from './components/spinner';
import { getStatisticTheme } from './components/statistic';
import { getStepsTheme } from './components/steps';
import { getSwitchTheme } from './components/switch';
import { getTableTheme } from './components/table';
import { getFiltersTheme } from './components/filters';
import { getTabsTheme } from './components/tabs';
import { getTagTheme } from './components/tag';
import { getTimePickerTheme } from './components/timePicker';
import { getTimelineTheme } from './components/timeline';
import { getTooltipTheme } from './components/tooltip';
import { getTypographyTheme } from './components/typography';
import { getWidgetTheme } from './components/widget';
import { getChartTheme } from './components/chart';
import { getLineChartTheme } from './components/lineChart';

import type { Theme } from './types';
import { getContextPanelTheme } from './components/contextPanel';
import { getBarChart } from './components/barChart';
import { getHorizontalBarChart } from './components/horizontalBarChart';

const getTheme = (
  scale = defaultScale,
  systemPalette = heroSystemPalette
): Theme => {
  const globalTheme = getGlobalTheme(scale, systemPalette);
  return {
    ...globalTheme,
    __hd__: {
      alert: getAlertTheme(globalTheme),
      badge: getBadgeTheme(globalTheme),
      banner: getBannerTheme(globalTheme),
      barChart: getBarChart(globalTheme),
      horizontalBarChart: getHorizontalBarChart(globalTheme),
      breadcrumb: getBreadcrumbTheme(globalTheme),
      button: getButtonTheme(globalTheme),
      card: getCardTheme(globalTheme),
      carousel: getCarouselTheme(globalTheme),
      checkbox: getCheckboxTheme(globalTheme),
      comment: getCommentTheme(globalTheme),
      columnChart: getColumnChartTheme(globalTheme),
      lineChart: getLineChartTheme(globalTheme),
      contextPanel: getContextPanelTheme(globalTheme),
      datePicker: getDatePickerTheme(globalTheme),
      divider: getDividerTheme(globalTheme),
      dropdown: getDropdownTheme(globalTheme),
      empty: getEmptyTheme(globalTheme),
      file: getFileTheme(globalTheme),
      form: getFormTheme(globalTheme),
      icon: getIconTheme(globalTheme),
      inPageNavigation: getInPageNavigationTheme(globalTheme),
      input: getInputTheme(globalTheme),
      legend: getLegendTheme(globalTheme),
      menu: getMenuTheme(globalTheme),
      modal: getModalTheme(globalTheme),
      notification: getNotificationTheme(globalTheme),
      pageHeader: getPageHeaderTheme(globalTheme),
      pagination: getPaginationTheme(globalTheme),
      portlet: getPortletTheme(globalTheme),
      progress: getProgressTheme(globalTheme),
      radio: getRadioTheme(globalTheme),
      rate: getRateTheme(globalTheme),
      result: getResultTheme(globalTheme),
      select: getSelectTheme(globalTheme),
      selectButton: getSelectButtonTheme(globalTheme),
      sideBar: getSideBarTheme(globalTheme),
      slider: getSliderTheme(globalTheme),
      spinner: getSpinnerTheme(globalTheme),
      statistic: getStatisticTheme(globalTheme),
      steps: getStepsTheme(globalTheme),
      switch: getSwitchTheme(globalTheme),
      table: getTableTheme(globalTheme),
      filters: getFiltersTheme(globalTheme),
      tabs: getTabsTheme(globalTheme),
      tag: getTagTheme(globalTheme),
      timePicker: getTimePickerTheme(globalTheme),
      timeline: getTimelineTheme(globalTheme),
      tooltip: getTooltipTheme(globalTheme),
      typography: getTypographyTheme(globalTheme),
      widget: getWidgetTheme(globalTheme),
      chart: getChartTheme(globalTheme),
    },
  };
};

export default getTheme;
