// @hero-design/react/src/components/Portlet/StyledPortlet.tsx
import styled from 'styled-components';
import { StyledIconButton } from '../Button/StyledButton';
import { SpinnerContainer } from '../Spinner/StyledSpinner';

const StyledPortletWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-self: flex-start;
  align-content: stretch;
  border-radius: ${({ theme }) => theme.__hd__.portlet.radii.wrapper};
  border-width: ${({ theme }) => theme.__hd__.portlet.borderWidths.wrapper};
  padding: ${({ theme }) => theme.__hd__.portlet.space.wrapperPadding};
  background-color: ${({ theme }) =>
    theme.__hd__.portlet.colors.wrapperBackground};

  & > ${SpinnerContainer} {
    min-height: 0;
  }
`;

const StyledHeaderWrapper = styled.div<{ themeOpen: boolean }>`
  display: flex;
  align-content: center;
  align-items: flex-start;
  justify-content: space-between;
  margin-bottom: ${({ theme, themeOpen }) =>
    themeOpen ? theme.__hd__.portlet.space.headerWrapperMargin : 0};
`;

const StyledActionWrapper = styled.div`
  display: flex;
  margin-left: ${({ theme }) => theme.__hd__.portlet.space.actionWrapperMargin};

  & > *:not(:first-child) {
    margin-left: ${({ theme }) =>
      theme.__hd__.portlet.space.actionButtonMargin};
  }

  ${StyledIconButton} {
    display: flex;
  }
`;

const StyledTitleWrapper = styled.div`
  display: flex;
  align-content: center;
  align-items: center;
  flex-wrap: wrap;

  & > *:first-child {
    margin-right: ${({ theme }) =>
      theme.__hd__.portlet.space.headerTitleMargin};
  }
`;

export {
  StyledPortletWrapper,
  StyledTitleWrapper,
  StyledHeaderWrapper,
  StyledActionWrapper,
};
