// @hero-design/react/src/components/ContextPanel/StyledContextPanel.tsx
import styled, { DefaultTheme, keyframes } from 'styled-components';
import Box from '../Box';
import IconButton from '../Button/IconButton';

const PanelKeyframes = keyframes`
from{
    transform: translateX(100%);
}
to{
    transform: translateX(0);
}
`;

const OverlayKeyframes = (theme: DefaultTheme) => keyframes`
from{
    background-color: transparent;
}
to{
    background-color: ${theme.__hd__.contextPanel.colors.overlayBackground};
}
`;

const StyledContextPanelContainer = styled(Box)`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  overflow: hidden;
  z-index: 9999;
`;

const StyledContextPanelOverlay = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  animation: ${({ theme }) => OverlayKeyframes(theme)} 0.2s linear forwards;
`;

const StyledContextPanel = styled(Box)<{
  themeOpen: boolean;
  themeHasFooter: boolean;
}>`
  position: absolute;
  height: 100vh;
  width: max-content;
  max-width: 100%;
  right: 0;
  top: 0;
  background-color: ${({ theme }) =>
    theme.__hd__.contextPanel.colors.contextPanelBackground};
  border-radius: ${({ theme }) =>
    theme.__hd__.contextPanel.borderRadius.contextPanel};
  padding: ${({ theme }) =>
    theme.__hd__.contextPanel.spaces.contextPanelPadding};
  will-change: transform;
  animation: ${PanelKeyframes} 0.2s linear;
  display: flex;
  flex-direction: column;
  justify-content: ${({ themeHasFooter }) =>
    themeHasFooter ? 'space-between' : 'flex-start'};
  overflow: hidden;
  overflow-y: auto;
  scrollbar-width: thin;
  scrollbar-color: ${({ theme }) =>
    `${theme.__hd__.contextPanel.colors.scrollbarThumbBackground} transparent`};

  &::-webkit-scrollbar {
    width: ${({ theme }) => theme.__hd__.contextPanel.sizes.scrollbarWidth};
  }
  &::-webkit-scrollbar-track {
    background: transparent;
    margin: ${({ theme }) =>
      theme.__hd__.contextPanel.spaces.scrollbarTrackMargin};
  }

  &::-webkit-scrollbar-thumb {
    background: ${({ theme }) =>
      theme.__hd__.contextPanel.colors.scrollbarThumbBackground};
    border-radius: ${({ theme }) =>
      theme.__hd__.contextPanel.borderRadius.scrollbar};
  }
`;

const StyledContextPanelHeader = styled(Box)`
  display: grid;
  margin-bottom: ${({ theme }) =>
    theme.__hd__.contextPanel.spaces.contextPanelHeaderMarginBottom};
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  align-items: center;
  font-size: ${({ theme }) => theme.__hd__.contextPanel.fontSizes.header};
  font-weight: ${({ theme }) => theme.__hd__.contextPanel.fontWeights.header};
  line-height: ${({ theme }) => theme.__hd__.contextPanel.lineHeights.header};
`;

const StyledContextPanelBody = styled(Box)<{
  themeHasFooter: boolean;
  themeHasHeader: boolean;
}>`
  flex: 1;
  font-size: ${({ theme }) => theme.__hd__.contextPanel.fontSizes.body};
  font-weight: ${({ theme }) => theme.__hd__.contextPanel.fontWeights.body};
  line-height: ${({ theme }) => theme.__hd__.contextPanel.lineHeights.body};
  margin-top: ${({ themeHasHeader, theme }) =>
    themeHasHeader
      ? theme.__hd__.contextPanel.spaces.contextPanelBodyMarginTop
      : '0'};
  margin-bottom: ${({ themeHasFooter, theme }) =>
    themeHasFooter
      ? theme.__hd__.contextPanel.spaces.contextPanelBodyMarginBottom
      : '0'};
`;

const StyledContextPanelFooter = styled(Box)`
  position: sticky;
  bottom: 0;
  left: 0;
  width: 100%;
  padding: ${({ theme }) =>
    theme.__hd__.contextPanel.spaces.contextPanelFooterPadding};
  background-color: ${({ theme }) =>
    theme.__hd__.contextPanel.colors.contextPanelBackground};
`;

const StyledContextPanelCloseIcon = styled(IconButton)`
  justify-self: flex-end;
`;

export {
  StyledContextPanelOverlay,
  StyledContextPanel,
  StyledContextPanelContainer,
  StyledContextPanelHeader,
  StyledContextPanelBody,
  StyledContextPanelFooter,
  StyledContextPanelCloseIcon,
};
