// @hero-design/colors/src/colorScales/index.ts
import createColorScales from './createColorScales';
import mixColor from './mixColor';

const BASE_COLORS = {
  aliceBlue: '#ecf0ff',
  amaranthaRed: '#cb300a',
  antiFlashWhite: '#f1f2f3',
  apple: '#e3602a',
  bigStone: '#344348',
  black: '#000000',
  blue1: '#5a68e2',
  blue: '#1dbeee',
  blueOblivion: '#23438f',
  blazingBonfire: '#ffa234',
  cornflowerBlue: '#5d80f0',
  cumberlandFog: '#dadbde',
  currant: '#48000a',
  darkBlue: '#25006e',
  deepSaffron: '#ffa234',
  deepIntoTheJungle: '#004e48',
  emerald: '#5ace7d',
  flagstone: '#acacaf',
  frenchSky: '#8999ff',
  gentianFlower: '#355bfb',
  gold: '#ffd500',
  green: '#01b39c',
  grey: '#a3a6ac',
  grotesqueGreen: '#6af293',
  hitPink: '#ffad79',
  honeydew: '#f0fef4',
  indianDance: '#f89076',
  jordyBlue: '#92abfc',
  linen: '#fcebe7',
  maasstrichtBlue: '#001f23',
  mauve: '#c38cee',
  moussaka: '#6f2a17',
  mellowApricot: '#ffbe71',
  mintLeaf: '#01cbb0',
  nightBlue: '#353957',
  orangeBrown: '#ac5d00',
  pastelRed: '#f46363',
  pastelBlue: '#a1b5fc',
  paua: '#130066',
  peach: '#f8ac7d',
  pineGreen: '#017d6d',
  pink: '#fe56aa',
  policeBlue: '#2e456d',
  purple: '#be83cf',
  royalBlue: '#5a68e2',
  scarletGum: '#401960',
  seashell: '#fff6eb',
  silverChalice: '#abacaf',
  smalt: '#003580',
  sonicSilver: '#737479',
  ultramarineBlue: '#4568fb',
  vermilion: '#de350b',
  violet1: '#8505a2',
  violet: '#7622d7',
  vodka: '#b5c3fd',
  theatreGold: '#a56822',
  uniformGreen: '#4d4628',
  yellow: '#fadb14',
  windsorGrey: '#606065',
} as const;

const colorScales = Object.entries(BASE_COLORS).reduce(
  (acc, [key, value]) => ({ ...acc, [key]: createColorScales(value) }),
  {} as Record<keyof typeof BASE_COLORS, ReturnType<typeof createColorScales>>
);

export { mixColor };

export default colorScales;
