// @hero-design/react/src/components/Widget/StyledWidget.tsx
import styled, { css } from 'styled-components';

const WidgetContainer = styled.div<{ themeActive: boolean }>`
  margin: 0;
  padding: 0;

  width: 100%;
  height: 100%;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  overflow: hidden;
  position: relative;
  background: ${({ theme }) => theme.__hd__.widget.colors.containerBackground};
  border: ${({ theme }) =>
    `${theme.__hd__.widget.borderWidths.container} solid ${theme.__hd__.widget.colors.containerBorder}`};
  border-radius: ${({ theme }) => theme.__hd__.widget.radii.container};

  ${({ theme, themeActive }) => {
    if (themeActive) {
      return css`
        border-color: ${theme.__hd__.widget.colors.containerActiveBorder};
      `;
    }
  }};
`;

const Header = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  align-items: center;
  padding: ${({ theme }) => theme.__hd__.widget.space.headerPadding};
`;

const HeaderLeft = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  align-items: center;
  gap: ${({ theme }) => theme.__hd__.widget.space.headerItemGap};
`;

const HeaderRight = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  align-items: center;
  gap: ${({ theme }) => theme.__hd__.widget.space.headerItemGap};
  margin-left: auto;
`;

const Content = styled.div`
  margin: 0;
  padding: 0;
  flex-grow: 1;
  overflow-y: auto;
  padding: ${({ theme }) => theme.__hd__.widget.space.contentPadding};
`;

const ResizeHandle = styled.div<{ themeActive: boolean }>`
  margin: 0;
  padding: 0;
  display: flex;
  position: absolute;
  bottom: ${({ theme }) => theme.__hd__.widget.space.resizeHandleOffset};
  right: ${({ theme }) => theme.__hd__.widget.space.resizeHandleOffset};
  padding: ${({ theme }) => theme.__hd__.widget.space.resizeHandlePadding};

  ${({ themeActive }) => {
    if (themeActive) {
      return css`
        cursor: nwse-resize;
      `;
    }
  }};
`;

export {
  WidgetContainer,
  Header,
  HeaderLeft,
  HeaderRight,
  Content,
  ResizeHandle,
};
