// @hero-design/react/src/components/Chart/ColumnChart/ChartContent/BarTooltip.tsx
import React, { forwardRef } from 'react';
import { StyledTooltip } from '../../StyledChart';

import { CommonProps } from '../../../common';
import { StyledTooltipContentWrapper } from '../StyledColumnChart';
import { DataValue } from '../../type';
import DefaultTooltip from './DefaultTooltip';

interface BarTooltipProps extends CommonProps {
  show?: boolean;
  color?: string;
  title: string;
  value?: string | number;
  xAxisLabel?: string;
  percent?: string;
  tooltipRenderer?: (props: {
    label?: string;
    value?: DataValue;
    legendColor?: string;
    xAxisLabel?: string;
  }) => React.ReactNode;
}

const BarTooltip = forwardRef<HTMLDivElement, BarTooltipProps>(
  (
    {
      show,
      'data-test-id': dataTestId,
      color,
      title,
      value,
      tooltipRenderer,
      xAxisLabel,
      percent,
    },
    ref
  ) => {
    const renderTooltip = React.useCallback(() => {
      return tooltipRenderer ? (
        tooltipRenderer({
          label: title,
          value: Number(value),
          legendColor: color,
          xAxisLabel,
        })
      ) : (
        <DefaultTooltip
          label={title}
          value={Number(value)}
          legendColor={color}
          percent={percent}
        />
      );
    }, [color, percent, title, tooltipRenderer, value, xAxisLabel]);

    return (
      <StyledTooltip
        ref={ref}
        style={{ position: 'fixed', display: show ? 'block' : 'none' }}
        data-test-id={`${dataTestId}-tooltip`}
        variant="light"
      >
        <StyledTooltipContentWrapper>
          {renderTooltip()}
        </StyledTooltipContentWrapper>
      </StyledTooltip>
    );
  }
);

export default BarTooltip;
