// @hero-design/react/src/components/Chart/ColumnChart/ChartContent/Rect.tsx
import React, { memo, useCallback, useLayoutEffect, useRef } from 'react';
import * as d3 from 'd3';
import { useTheme } from 'styled-components';
import { CommonProps } from '../../../common';
import { BarConfig } from '../../type';

interface RectProps extends CommonProps {
  height: number;
  color: string;
  x?: number;
  y?: number;
  onMouseOver?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  onMouseMove?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  onMouseOut?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  onClick?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  isHovered?: boolean;
  isHighlighted?: boolean;
  cursor?: SVGGElement['style']['cursor'];
  size?: BarConfig<unknown>['size'];
}

const HOVER_EFFECT_DURATION = 150;

const Rect = ({
  height,
  x,
  y,
  color,
  onMouseOver,
  onMouseMove,
  onMouseOut,
  onClick,
  'data-test-id': dataTestId,
  className,
  isHovered,
  isHighlighted,
  cursor,
  size = 'large',
}: RectProps) => {
  const theme = useTheme();

  const rectRef = useRef<SVGRectElement>(null);

  const {
    __hd__: { columnChart: columnChartTheme },
  } = theme;

  const { barDropShadow } = columnChartTheme.boxShadows;

  const { barWidth: barWidthMapping } = columnChartTheme.sizes;

  const { bar: barRadiusMapping } = columnChartTheme.radii;

  const { barHoverTransform: barHoverTransformMapping } =
    columnChartTheme.space;

  const barWidth = barWidthMapping[size];

  const barBorderRadius = barRadiusMapping[size];

  const barHoverTransform = barHoverTransformMapping[size];

  const handleIsHoveredEvent = useCallback(
    (
      rectElement: d3.Selection<
        SVGRectElement | null,
        unknown,
        null,
        undefined
      >,
      hovered?: boolean
    ) => {
      if (hovered) {
        rectElement
          .transition()
          .duration(HOVER_EFFECT_DURATION)
          .attr('x', x || 0)
          .attr('width', barWidth)
          .attr('rx', barBorderRadius)
          .attr('ry', barBorderRadius);
        return;
      }

      if (hovered === false) {
        rectElement
          .transition()
          .duration(HOVER_EFFECT_DURATION)
          .attr('x', (x || 0) + barHoverTransform)
          .attr('width', barWidthMapping.inactive)
          .attr('rx', barRadiusMapping.inactive)
          .attr('ry', barRadiusMapping.inactive);
        return;
      }

      if (hovered === undefined) {
        rectElement
          .transition()
          .duration(HOVER_EFFECT_DURATION)
          .attr('x', x || 0)
          .attr('width', barWidth)
          .attr('rx', barBorderRadius)
          .attr('ry', barBorderRadius);
      }
    },
    [
      x,
      barWidth,
      barBorderRadius,
      barHoverTransform,
      barWidthMapping.inactive,
      barRadiusMapping.inactive,
    ]
  );

  const handleHighlightedEvent = useCallback(
    (
      rectElement: d3.Selection<
        SVGRectElement | null,
        unknown,
        null,
        undefined
      >,
      highlighted: boolean,
      hovered?: boolean
    ) => {
      if (highlighted) {
        rectElement
          .transition()
          .duration(HOVER_EFFECT_DURATION)
          .attr('x', x || 0)
          .attr('width', barWidth)
          .attr('rx', barBorderRadius)
          .attr('ry', barBorderRadius);
        return;
      }

      if (!highlighted && hovered) {
        rectElement
          .transition()
          .attr('x', x || 0)
          .attr('width', barWidth)
          .attr('rx', barBorderRadius)
          .attr('ry', barBorderRadius);
        return;
      }

      if (!highlighted && !hovered) {
        rectElement
          .transition()
          .attr('x', (x || 0) + barHoverTransform)
          .attr('width', barWidthMapping.inactive)
          .attr('rx', barRadiusMapping.inactive)
          .attr('ry', barRadiusMapping.inactive);
      }
    },
    [
      barBorderRadius,
      barHoverTransform,
      barRadiusMapping.inactive,
      barWidth,
      barWidthMapping.inactive,
      x,
    ]
  );

  useLayoutEffect(() => {
    const rectElement = d3.select(rectRef.current);

    if (typeof isHighlighted === 'undefined') {
      handleIsHoveredEvent(rectElement, isHovered);
      return;
    }

    if (typeof isHighlighted !== 'undefined') {
      handleHighlightedEvent(rectElement, isHighlighted, isHovered);
    }
  }, [handleHighlightedEvent, handleIsHoveredEvent, isHighlighted, isHovered]);

  return (
    <rect
      ref={rectRef}
      onMouseOver={onMouseOver}
      onMouseMove={onMouseMove}
      onMouseOut={onMouseOut}
      onClick={onClick}
      className={className}
      data-test-id={dataTestId}
      y={y}
      width={barWidth}
      x={x}
      height={height < 0 ? 0 : Math.round(height)}
      fill={color}
      filter={
        isHovered || isHighlighted ? `drop-shadow(${barDropShadow})` : undefined
      }
      cursor={cursor}
      rx={barBorderRadius}
      ry={barBorderRadius}
    />
  );
};

export default memo(Rect);
