// @hero-design/react/src/components/Table/Filters/StyledFilters.tsx
import styled, { css } from 'styled-components';

import { StyledIcon } from '../../Icon/StyledIcon';

const StyledDropdownButton = styled.button<{ themeActive: boolean }>`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.table.space.filterButtonPadding};
  height: ${({ theme }) => theme.__hd__.table.sizes.filterButton};
  border: ${({ theme }) =>
    `${theme.__hd__.table.borderWidths.filterButton} solid`};
  font-size: ${({ theme }) => theme.__hd__.table.fontSizes.filterButton};
  font-weight: ${({ theme }) => theme.__hd__.table.fontWeights.filterButton};
  line-height: ${({ theme }) => theme.__hd__.table.lineHeights.filterButton};
  display: inline-flex;
  align-items: center;
  background-color: ${({ theme }) =>
    theme.__hd__.table.colors.filterButtonBackground};
  cursor: pointer;
  position: relative;
  max-width: 240px;
  margin-right: ${({ theme }) =>
    theme.__hd__.table.space.filterButtonMarginRight};
  margin-bottom: ${({ theme }) =>
    theme.__hd__.table.space.filterButtonMarginBottom};
  border-radius: ${({ theme }) => theme.__hd__.table.radii.filterButton};
  border-color: ${({ theme }) => theme.__hd__.table.colors.filterButtonBorder};

  &:hover {
    z-index: 1;
    border-color: ${({ theme }) =>
      theme.__hd__.table.colors.filterButtonHoverBorder};
  }

  ${({ theme, themeActive }) =>
    themeActive
      ? css`
          z-index: 1;
          background-color: ${theme.__hd__.table.colors
            .filterButtonActiveBackground};
        `
      : css`
          background-color: ${theme.__hd__.table.colors.filterButtonBackground};
        `};

  > ${StyledIcon} {
    padding-left: ${({ theme }) =>
      theme.__hd__.table.space.filterButtonIconPadding};
  }
`;

const StyledDropdownContent = styled.div`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.table.space.filterDropdownPadding};
  overflow: hidden;
`;

const StyledLabeledFilter = styled.div`
  margin: 0;
  padding: 0;
  font-size: ${({ theme }) => theme.__hd__.table.fontSizes.filterButton};
  font-weight: ${({ theme }) => theme.__hd__.table.fontWeights.filterLabel};
  color: ${({ theme }) => theme.__hd__.table.colors.filterLabelText};
`;

const FilterByWrapper = styled.span`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.table.space.filterButtonPadding};
  padding-left: 0;
  font-size: ${({ theme }) => theme.__hd__.table.fontSizes.filterButton};
  font-weight: ${({ theme }) => theme.__hd__.table.fontWeights.filterBy};
  line-height: 100%;
  white-space: nowrap;
`;

const FiltersWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const StyledLabel = styled.span`
  margin: 0;
  padding: 0;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`;

const StyledGroupedFilterContent = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
  max-width: 300px;

  > ${StyledLabeledFilter}:not(:last-child) {
    padding-bottom: ${({ theme }) =>
      theme.__hd__.table.space.filterDropdownPadding};
  }
`;

const StyledLabeledFilterContent = styled.div`
  margin: 0;
  padding: 0;
  padding-top: ${({ theme }) =>
    theme.__hd__.table.space.filterLabelContentPadding};
`;

const StyledFilterList = styled.div`
  margin: 0;
  padding: 0;
`;

const StyledFiltersHeader = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  justify-content: space-between;
  margin-bottom: ${({ theme }) => theme.space.small}px;

  @media (min-width: ${({ theme }) => theme.breakpoints.large}) {
    display: none;
  }
`;

const StyledFilterListWithAddons = styled.div`
  margin: 0;
  padding: 0;
  display: flex;

  @media (max-width: calc(${({ theme }) => theme.breakpoints.large} - 1px)) {
    > :not(${StyledFilterList}) {
      display: none;
    }
  }
`;

const StyledTagList = styled.div`
  margin: 0;
  padding: 0;
`;

export {
  StyledDropdownContent,
  StyledDropdownButton,
  StyledGroupedFilterContent,
  StyledLabeledFilter,
  StyledLabel,
  StyledLabeledFilterContent,
  StyledFilterList,
  StyledFiltersHeader,
  StyledFilterListWithAddons,
  StyledTagList,
  FilterByWrapper,
  FiltersWrapper,
};
