// @hero-design/colors/src/defaultWebPalette/index.ts
import colorScales from '../colorScales';

const {
  blue,
  ultramarineBlue,
  green,
  grey,
  grotesqueGreen,
  deepSaffron,
  pink,
  purple,
  vermilion,
  smalt,
  violet,
  yellow,
} = colorScales;

const palette = {
  black: '#000000',
  white: '#ffffff',
  blue: blue.base,
  blueDark15: blue.darken15,
  blueDark30: blue.darken30,
  blueDark45: blue.darken45,
  blueDark60: blue.darken60,
  blueDark75: blue.darken75,
  blueDark90: blue.darken90,
  blueLight15: blue.lighten15,
  blueLight30: blue.lighten30,
  blueLight45: blue.lighten45,
  blueLight60: blue.lighten60,
  blueLight75: blue.lighten75,
  blueLight90: blue.lighten90,
  dodgerBlue: ultramarineBlue.base,
  dodgerBlueDark15: ultramarineBlue.darken15,
  dodgerBlueDark30: ultramarineBlue.darken30,
  dodgerBlueLight30: ultramarineBlue.lighten30,
  dodgerBlueLight45: ultramarineBlue.lighten45,
  dodgerBlueLight60: ultramarineBlue.lighten60,
  dodgerBlueLight75: ultramarineBlue.lighten75,
  dodgerBlueLight90: ultramarineBlue.lighten90,
  green: green.base,
  greenDark15: green.darken15,
  greenDark30: green.darken30,
  greenDark75: green.darken75,
  greenLight30: green.lighten30,
  greenLight75: green.lighten75,
  greenLight90: green.lighten90,
  grey: grey.base,
  greyDark15: grey.darken15,
  greyDark30: grey.darken30,
  greyDark45: grey.darken45,
  greyDark60: grey.darken60,
  greyDark75: grey.darken75,
  greyLight45: grey.lighten45,
  greyLight60: grey.lighten60,
  greyLight75: grey.lighten75,
  greyLight85: grey.lighten85,
  greyLight90: grey.lighten90,
  greyLight95: grey.lighten95,
  grotesqueGreen: grotesqueGreen.base,
  grotesqueGreenDark30: grotesqueGreen.darken30,
  grotesqueGreenDark45: grotesqueGreen.darken45,
  grotesqueGreenLight30: grotesqueGreen.lighten30,
  grotesqueGreenLight60: grotesqueGreen.lighten60,
  grotesqueGreenLight75: grotesqueGreen.lighten75,
  grotesqueGreenLight90: grotesqueGreen.lighten90,
  orange: deepSaffron.base,
  orangeDark15: deepSaffron.darken15,
  orangeDark30: deepSaffron.darken30,
  orangeDark75: deepSaffron.darken75,
  orangeLight30: deepSaffron.lighten30,
  orangeLight75: deepSaffron.lighten75,
  orangeLight90: deepSaffron.lighten90,
  pink: pink.base,
  pinkDark15: pink.darken15,
  pinkDark30: pink.darken30,
  pinkDark45: pink.darken45,
  pinkDark75: pink.darken75,
  pinkLight30: pink.lighten30,
  pinkLight45: pink.lighten45,
  pinkLight75: pink.lighten75,
  pinkLight90: pink.lighten90,
  purple: purple.base,
  purpleDark15: purple.darken15,
  purpleDark30: purple.darken30,
  purpleLight30: purple.lighten30,
  red: vermilion.base,
  redDark15: vermilion.darken15,
  redDark30: vermilion.darken30,
  redDark45: vermilion.darken45,
  redDark60: vermilion.darken60,
  redDark75: vermilion.darken75,
  redDark90: vermilion.darken90,
  redLight15: vermilion.lighten15,
  redLight30: vermilion.lighten30,
  redLight45: vermilion.lighten45,
  redLight60: vermilion.lighten60,
  redLight75: vermilion.lighten75,
  redLight90: vermilion.lighten90,
  smalt: smalt.base,
  smaltDark30: smalt.darken30,
  smaltDark75: smalt.darken75,
  smaltLight30: smalt.lighten30,
  smaltLight45: smalt.lighten45,
  smaltLight75: smalt.lighten75,
  smaltLight90: smalt.lighten90,
  violet: violet.base,
  violetDark15: violet.darken15,
  violetDark30: violet.darken30,
  violetDark45: violet.darken45,
  violetDark60: violet.darken60,
  violetDark75: violet.darken75,
  violetDark90: violet.darken90,
  violetLight15: violet.lighten15,
  violetLight30: violet.lighten30,
  violetLight45: violet.lighten45,
  violetLight60: violet.lighten60,
  violetLight75: violet.lighten75,
  violetLight90: violet.lighten90,
  yellow: yellow.base,
  yellowDark15: yellow.darken15,
  yellowDark30: yellow.darken30,
  yellowDark75: yellow.darken75,
  yellowLight30: yellow.lighten30,
  yellowLight45: yellow.lighten45,
  yellowLight60: yellow.lighten60,
  yellowLight75: yellow.lighten75,
  yellowLight90: yellow.lighten90,
};

export default palette;
