// @hero-design/react/src/components/Chart/HorizontalBar/ChartContent/Rect.tsx
import React, { memo, useLayoutEffect, useRef } from 'react';
import * as d3 from 'd3';
import { useTheme } from 'styled-components';
import { CommonProps } from '../../../common';
import { BarConfig, DataValue } from '../../type';

interface RectProps extends CommonProps {
  width: number;
  color: string;
  x?: number;
  y?: number;
  radius?: number;
  onMouseOver?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  onMouseMove?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  onMouseOut?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  onClick?: (e: React.MouseEvent<SVGRectElement, MouseEvent>) => void;
  active?: boolean;
  isHighlighted?: boolean;
  size?: BarConfig<DataValue>['size'];
}

const Rect = ({
  width,
  x,
  y,
  color,
  onMouseOver,
  onMouseMove,
  onMouseOut,
  onClick,
  'data-test-id': dataTestId,
  className,
  radius,
  active,
  isHighlighted,
  size = 'large',
}: RectProps) => {
  const HOVER_EFFECT_DURATION = 150;
  const theme = useTheme();

  const rectRef = useRef<SVGRectElement>(null);

  const {
    __hd__: { horizontalBarChart: horizontalBarChartTheme },
  } = theme;

  const { barDropShadow } = horizontalBarChartTheme.boxShadows;

  const { barHeight: barHeightMapping } = horizontalBarChartTheme.sizes;

  const barHeight = barHeightMapping[size];

  const { bar: barBorderRadiusMapping } = horizontalBarChartTheme.radii;

  const { barHoverTransform: barHoverTransformMapping } =
    horizontalBarChartTheme.space;

  const barHoverTransform = barHoverTransformMapping[size];

  useLayoutEffect(() => {
    const rectElement = d3.select(rectRef.current);

    if (active && typeof isHighlighted === 'undefined') {
      rectElement
        .transition()
        .duration(HOVER_EFFECT_DURATION)
        .attr('y', y || 0)
        .attr('height', barHeight)
        .attr('rx', radius || 0)
        .attr('ry', radius || 0);
      return;
    }

    if (active === false && typeof isHighlighted === 'undefined') {
      rectElement
        .transition()
        .duration(HOVER_EFFECT_DURATION)
        .attr('y', (y || 0) + barHoverTransform)
        .attr('height', barHeightMapping.inactive)
        .attr('rx', barBorderRadiusMapping.inactive)
        .attr('ry', barBorderRadiusMapping.inactive);
      return;
    }

    if (active === undefined && typeof isHighlighted === 'undefined') {
      rectElement
        .transition()
        .duration(HOVER_EFFECT_DURATION)
        .attr('y', y || 0)
        .attr('height', barHeight)
        .attr('rx', radius || 0)
        .attr('ry', radius || 0);

      return;
    }

    if (!isHighlighted) {
      if (active) {
        rectElement
          .transition()
          .duration(HOVER_EFFECT_DURATION)
          .attr('y', y || 0)
          .attr('height', barHeight)
          .attr('rx', radius || 0)
          .attr('ry', radius || 0);
        return;
      }
      if (active === false) {
        rectElement
          .transition()
          .duration(HOVER_EFFECT_DURATION)
          .attr('y', (y || 0) + barHoverTransform)
          .attr('height', barHeightMapping.inactive)
          .attr('rx', barBorderRadiusMapping.inactive)
          .attr('ry', barBorderRadiusMapping.inactive);
        return;
      }

      rectElement
        .attr('y', (y || 0) + barHoverTransform)
        .attr('height', barHeightMapping.inactive)
        .attr('rx', barBorderRadiusMapping.inactive)
        .attr('ry', barBorderRadiusMapping.inactive);
    }
  }, [
    active,
    isHighlighted,
    barBorderRadiusMapping.inactive,
    barHeight,
    barHeightMapping.inactive,
    barHoverTransform,
    radius,
    y,
  ]);

  return (
    <rect
      ref={rectRef}
      onMouseOver={onMouseOver}
      onMouseMove={onMouseMove}
      onMouseOut={onMouseOut}
      onClick={onClick}
      className={className}
      data-test-id={dataTestId}
      y={y}
      height={barHeight}
      x={x}
      width={width < 0 ? 0 : width}
      fill={color}
      filter={
        active || isHighlighted ? `drop-shadow(${barDropShadow})` : undefined
      }
      cursor="pointer"
      rx={radius}
      ry={radius}
    />
  );
};

export default memo(Rect);
