// @hero-design/react/src/components/TagInput/StyledTagInput.tsx
import styled, { css } from 'styled-components';

import StyledInput, { InputWrapper } from '../Input/StyledInput';
import { StyledTag } from '../Tag/StyledTag';
import type { Size } from '../Input/utils';

const TagInputWrapper = styled(InputWrapper)<{
  themeSize: Size;
}>`
  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          padding-top: ${theme.__hd__.input.space.smallPadding};
          padding-bottom: ${theme.__hd__.input.space.smallPadding};

          ${StyledInput} {
            height: ${`calc(${theme.__hd__.input.sizes.small} - ${theme.__hd__.input.space.smallPadding})`};
          }
        `;
      case 'medium':
        return css`
          padding-top: ${theme.__hd__.input.space.mediumPadding};
          padding-bottom: ${theme.__hd__.input.space.mediumPadding};

          ${StyledInput} {
            height: ${`calc(${theme.__hd__.input.sizes.medium} - ${theme.__hd__.input.space.mediumPadding})`};
          }
        `;
      case 'large':
        return css`
          padding-top: ${theme.__hd__.input.space.largePadding};
          padding-bottom: ${theme.__hd__.input.space.largePadding};

          ${StyledInput} {
            height: ${`calc(${theme.__hd__.input.sizes.large} - ${theme.__hd__.input.space.largePadding})`};
          }
        `;
    }
  }};
`;

const TagInputContainer = styled.div<{
  themePaddingX: Size | 'none';
}>`
  display: inline-flex;
  flex-wrap: wrap;
  align-items: center;

  ${({ themePaddingX, theme }) => {
    switch (themePaddingX) {
      case 'small':
        return css`
          padding: 0 ${theme.__hd__.input.space.smallPadding};
        `;
      case 'medium':
        return css`
          padding: 0 ${theme.__hd__.input.space.mediumPadding};
        `;
      case 'large':
        return css`
          padding: 0 ${theme.__hd__.input.space.largePadding};
        `;
      case 'none':
        return css`
          padding: 0;
        `;
    }
  }};

  width: calc(100% + ${({ theme }) => theme.__hd__.input.space.tagMargin});
  margin: -${({ theme }) => theme.__hd__.input.space.tagMargin} 0 0 -${({
      theme,
    }) => theme.__hd__.input.space.tagMargin};

  > * {
    margin: ${({ theme }) => theme.__hd__.input.space.tagMargin} 0 0
      ${({ theme }) => theme.__hd__.input.space.tagMargin};
  }

  > ${StyledTag} {
    border: none;
    background-color: ${({ theme }) => theme.__hd__.input.colors.tagBackground};
  }

  > ${StyledInput} {
    min-width: ${({ theme }) => theme.__hd__.input.sizes.tagInputMinWidth};
    flex: 1;
  }
`;

export { TagInputWrapper, TagInputContainer };
