// @hero-design/react/src/components/Rate/StyledRate.tsx
import styled, { css } from 'styled-components';
import Icon from '../Icon';

export const StyledRateIconLabel = styled.label`
  display: inline-flex;
  align-items: center;
  margin: 0;
  padding: 0;
`;

export const StyledRateIconInput = styled.input`
  display: none;
`;

export const StyledRateIcon = styled(Icon)<{
  active: boolean;
}>`
  cursor: pointer;

  ${({ active, theme }) => {
    switch (active) {
      case false:
        return css`
          color: ${theme.__hd__.rate.colors.inactiveIcon};
        `;
      case true:
        return css`
          color: ${theme.__hd__.rate.colors.activeIcon};
        `;
    }
  }};
`;

type ThemeStatus = 'disabled' | 'readonly' | 'normal';

export const RateWrapper = styled.div<{
  themeLayout: 'medium' | 'large';
  themeStatus?: ThemeStatus;
}>`
  display: inline-flex;
  align-items: center;
  flex-wrap: wrap;
  padding: 0;

  ${({ themeLayout, theme }) => {
    switch (themeLayout) {
      case 'medium':
        return css`
          margin: ${theme.__hd__.rate.space.rateGroupMediumMargin};
          ${StyledRateIconLabel} {
            padding: ${theme.__hd__.rate.space.rateMediumMargin};
          }
        `;
      case 'large':
        return css`
          margin: ${theme.__hd__.rate.space.rateGroupLargeMargin};
          ${StyledRateIconLabel} {
            padding: ${theme.__hd__.rate.space.rateLargeMargin};
          }
        `;
    }
  }};

  ${({ themeStatus, theme }) => {
    switch (themeStatus) {
      case 'disabled':
        return css`
          ${StyledRateIcon} {
            cursor: not-allowed;
            color: ${theme.__hd__.rate.colors.disabledIcon};
          }
        `;
      case 'readonly':
        return css`
          ${StyledRateIcon} {
            cursor: default;
          }
        `;
      default:
        return css``;
    }
  }};
`;
