// @hero-design/react/src/components/Banner/StyledBanner.tsx
import styled, { css } from 'styled-components';

const StyledIconWrapper = styled.div<{
  themeIntent: 'success' | 'info' | 'warning' | 'danger' | 'error';
  themeSize: 'default' | 'compact';
}>`
  display: flex;
  align-items: center;
  margin: 0;
  padding: 0;

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'success':
        return css`
          color: ${theme.__hd__.banner.colors.success};
        `;
      case 'info':
        return css`
          color: ${theme.__hd__.banner.colors.info};
        `;
      case 'warning':
        return css`
          color: ${theme.__hd__.banner.colors.warning};
        `;
      case 'danger':
        return css`
          color: ${theme.__hd__.banner.colors.danger};
        `;
      case 'error':
        return css`
          color: ${theme.__hd__.banner.colors.error};
        `;
    }
  }};

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'default':
        return css`
          margin-top: ${theme.__hd__.banner.space.iconDefaultMarginTop};
          margin-right: ${theme.__hd__.banner.space.iconDefaultMarginRight};
          line-height: ${theme.__hd__.banner.lineHeights.iconDefault};
        `;
      case 'compact':
        return css`
          margin-top: ${theme.__hd__.banner.space.iconCompactMarginTop};
          margin-right: ${theme.__hd__.banner.space.iconCompactMarginRight};
          line-height: ${theme.__hd__.banner.lineHeights.iconCompact};
        `;
    }
  }};
`;

const StyledInner = styled.div`
  flex: 1;
  margin: 0;
  padding: 0;
`;

const StyledTitle = styled.div`
  font-size: ${({ theme }) => theme.__hd__.banner.fontSizes.title};
  line-height: ${({ theme }) => theme.__hd__.banner.lineHeights.title};
  font-weight: ${({ theme }) => theme.__hd__.banner.fontWeights.title};
  margin: 0;
  padding: 0;
  margin-bottom: ${({ theme }) => theme.__hd__.banner.space.titleMarginBottom};
`;

const StyledContent = styled.div`
  font-size: ${({ theme }) => theme.__hd__.banner.fontSizes.content};
  line-height: ${({ theme }) => theme.__hd__.banner.lineHeights.content};
  margin: 0;
  padding: 0;
`;

const StyledBanner = styled.div<{
  themeIntent: 'info' | 'success' | 'warning' | 'danger' | 'error';
  themeSize: 'default' | 'compact';
}>`
  display: flex;
  margin: 0;
  align-items: flex-start;
  border-radius: ${({ theme }) => theme.__hd__.banner.radii.default};
  border-width: ${({ theme }) => theme.__hd__.banner.borderWidths.default};
  border-style: solid;
  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'default':
        return css`
          padding: ${theme.__hd__.banner.space.defaultPadding};
        `;
      case 'compact':
        return css`
          padding: ${theme.__hd__.banner.space.compactPadding};
        `;
    }
  }};

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'info':
        return css`
          background-color: ${theme.__hd__.banner.colors.bgInfo};
          border-color: ${theme.__hd__.banner.colors.borderInfo};
          color: ${theme.__hd__.banner.colors.textInfo};
        `;
      case 'success':
        return css`
          background-color: ${theme.__hd__.banner.colors.bgSuccess};
          border-color: ${theme.__hd__.banner.colors.borderSuccess};
          color: ${theme.__hd__.banner.colors.textSuccess};
        `;
      case 'warning':
        return css`
          background-color: ${theme.__hd__.banner.colors.bgWarning};
          border-color: ${theme.__hd__.banner.colors.borderWarning};
          color: ${theme.__hd__.banner.colors.textWarning};
        `;
      case 'danger':
        return css`
          background-color: ${theme.__hd__.banner.colors.bgDanger};
          border-color: ${theme.__hd__.banner.colors.borderDanger};
          color: ${theme.__hd__.banner.colors.textDanger};
        `;
      case 'error':
        return css`
          background-color: ${theme.__hd__.banner.colors.bgError};
          border-color: ${theme.__hd__.banner.colors.borderError};
          color: ${theme.__hd__.banner.colors.textError};
        `;
    }
  }};
`;

const StyledCloseButton = styled.div<{ themeSize: 'default' | 'compact' }>`
  margin: 0;
  padding: 0;
  margin-left: auto;
  color: ${({ theme }) => theme.__hd__.banner.colors.closeButton};

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'default':
        return css`
          margin-top: ${theme.__hd__.banner.space.iconDefaultMarginTop};
          line-height: ${theme.__hd__.banner.lineHeights.iconDefault};
        `;
      case 'compact':
        return css`
          margin-top: ${theme.__hd__.banner.space.iconCompactMarginTop};
          line-height: ${theme.__hd__.banner.lineHeights.iconCompact};
        `;
    }
  }};
`;

export {
  StyledTitle,
  StyledInner,
  StyledContent,
  StyledIconWrapper,
  StyledCloseButton,
  StyledBanner,
};
