// @hero-design/react/src/theme/components/badge.ts
import type { GlobalTheme } from '../global';

const getBadgeTheme = (theme: GlobalTheme) => {
  const colors = {
    basicBackground: theme.colors.defaultLightBackground,
    basicText: theme.colors.text,
    basicStatusText: theme.colors.disabledText,
    text: theme.colors.lightText,
    border: theme.colors.defaultBorder,
    warning: theme.colors.warning,
    success: theme.colors.success,
    error: theme.colors.error,
    danger: theme.colors.danger,
    primary: theme.colors.primary,
    archived: theme.colors.bodyText,
    primaryBackground: theme.colors.primaryLightBackground,
    successBackground: theme.colors.successBackground,
    warningBackground: theme.colors.warningBackground,
    dangerBackground: theme.colors.dangerBackground,
    errorBackground: theme.colors.errorBackground,
    archivedBackground: theme.colors.neutralLightBackground,
    primaryText: theme.colors.primary,
    successText: theme.colors.activeSuccess,
    warningText: theme.colors.warningDark,
    dangerText: theme.colors.activeDanger,
    errorText: theme.colors.activeError,
    archivedText: theme.colors.bodyText,
  };

  const fontWeights = {
    default: theme.fontWeights.semiBold,
  };

  const fontSizes = {
    default: `${theme.fontSizes.medium}px`,
  };

  const shadows = {
    default: theme.shadows.default,
  };

  const sizes = {
    height: `${theme.sizes.large}px`,
    minWidth: `${theme.sizes.large}px`,
  };

  const space = {
    right: 0,
    narrowPadding: `${theme.space.xsmall}px`,
    widePadding: `${theme.space.xsmall}px ${theme.space.small}px`,
  };

  const radii = {
    default: `${theme.sizes.large / 2}px`,
    status: '50%',
  };

  const borderWidths = {
    default: `${theme.borderWidths.base}px`,
  };

  return {
    colors,
    fontWeights,
    fontSizes,
    shadows,
    sizes,
    space,
    radii,
    borderWidths,
  };
};

type BadgeTheme = ReturnType<typeof getBadgeTheme>;

export { getBadgeTheme };
export type { BadgeTheme };
