// @hero-design/react/src/theme/components/barChart.ts
import type { GlobalTheme } from '../global';

const getBarChart = (theme: GlobalTheme) => {
  const colors = {
    xAxisTickColor: theme.colors.subduedText,
    yAxisTickColor: theme.colors.subduedText,
    verticalTitleFillColor: theme.colors.subduedText,
    gridLineBackgroundColor: theme.colors.neutralBackground,
    tooltipBackgroundColor: theme.colors.defaultLightBackground,
    totalBackgroundColor: theme.colors.defaultLightBackground,
  };

  const radii = {
    tooltip: theme.radii.medium,
    bar: {
      large: theme.radii.base * 2,
      medium: theme.radii.base * 2,
      inactive: theme.radii.base,
    },
    tooltipArrow: theme.radii.base / 2,
    totalRect: theme.radii.base * 3,
  };

  const sizes = {
    tooltipArrow: theme.sizes.small,
    barHeight: {
      large: theme.sizes.xsmall * 3,
      medium: theme.sizes.small,
      inactive: theme.sizes.xsmall,
    },
    navigationButton: theme.sizes.large,
    totalSvgHeight: theme.sizes.large,
    totalLetterWidth: theme.sizes.small,
  };

  const space = {
    tooltipPadding: theme.space.small,
    tooltipChildMargin: theme.space.small,
    legendContainerMarginTop: theme.space.medium,
    legendContainerChildMarginBottom: theme.space.small,
    legendBoxGap: theme.space.medium,
    barChartMarginTop: theme.space.medium,
    barChartMarginRight: theme.space.large,
    barChartMarginBottom: theme.space.xlarge,
    barChartMarginLeft: theme.space.xxxxlarge + theme.space.large,
    barChartMarginLeftWithNavigation: theme.space.large * 2,
    chartContentMarginLeft: theme.space.xlarge,
    xAxisTransform: theme.space.medium,
    yAxisTransformHorizontal: -theme.space.small,
    barHoverTransform: {
      large: theme.space.xsmall,
      medium: theme.space.xxsmall,
    },
    tooltipTop: 50,
    barMarginLeft: theme.space.xsmall,
    moreTooltipPadding: theme.space.small,
    navigationButtonMarginLeft: theme.space.small,
    verticalTitleY: -theme.sizes.xlarge * 2,
    gridChartMarginTop: theme.space.small,
    barMarginTop: theme.space.xsmall,
    totalSvgHorizontalPadding: theme.space.small,
  };

  const fontWeights = {
    xAxisTick: theme.fontWeights.regular,
    yAxisTick: theme.fontWeights.regular,
    verticalTitle: theme.fontWeights.regular,
    totalText: theme.fontWeights.semiBold,
  };

  const fontSizes = {
    yAxisTick: `${theme.fontSizes.xsmall}px`,
    verticalTitle: `${theme.fontSizes.small}px`,
    xAxisTick: `${theme.fontSizes.xsmall}px`,
    totalText: `${theme.fontSizes.xsmall}px`,
  };

  const lineHeights = {
    yAxisTick: `${theme.lineHeights.xsmall}px`,
    verticalTitle: `${theme.lineHeights.small}px`,
    xAxisTick: `${theme.lineHeights.xsmall}px`,
  };

  const boxShadows = {
    tooltip: '0 0 10px 0 rgba(0, 0, 0, 0.2)',
    barDropShadow: '1px 1px 4px rgb(0 0 0 / 0.3)',
  };

  const borderWidth = {
    totalRect: theme.borderWidths.base,
  };

  return {
    colors,
    radii,
    space,
    fontWeights,
    fontSizes,
    lineHeights,
    boxShadows,
    sizes,
    borderWidth,
  };
};

type BarChartTheme = ReturnType<typeof getBarChart>;

export { getBarChart };
export type { BarChartTheme };
