// @hero-design/react/src/theme/components/card.ts
import type { GlobalTheme } from '../global';

const getCardTheme = (theme: GlobalTheme) => {
  const colors = {
    border: theme.colors.defaultBorder,
    background: theme.colors.defaultLightBackground,
    headerBasicBackground: theme.colors.defaultLightBackground,
    headerInfoBackground: theme.colors.infoHeaderBackground,
    headerGreyBackground: theme.colors.primaryHeaderBackground,
    headerPrimaryBackground: theme.colors.primary,
    headerDefaultText: theme.colors.text,
    headerInfoText: theme.colors.text,
    headerLightText: theme.colors.lightText,
    contentText: theme.colors.bodyText,
    contentBorder: theme.colors.defaultBorder,
    contentTextExtra: theme.colors.subduedText,
  };

  const borderWidths = {
    default: `${theme.borderWidths.base}px`,
  };

  const radii = {
    default: `${theme.radii.medium}px`,
  };

  const space = {
    headerMediumPadding: `${theme.space.medium}px`,
    headerSmallPadding: `${theme.space.small}px`,
    headerMargin: `-${theme.borderWidths.base}px -${theme.borderWidths.base}px 0`,
    contentMediumPadding: `${theme.space.medium}px`,
    contentSmallPadding: `${theme.space.small}px`,
    imgMargin: `-${theme.borderWidths.base}px -${theme.borderWidths.base}px 0`,
  };

  const fontWeights = {
    header: theme.fontWeights.regular,
    content: theme.fontWeights.regular,
  };

  const fontSizes = {
    header: `${theme.fontSizes.large}px`,
    content: `${theme.fontSizes.medium}px`,
    contentExtra: `${theme.fontSizes.small}px`,
  };

  const lineHeights = {
    header: `${theme.lineHeights.large}px`,
    content: `${theme.lineHeights.medium}px`,
    contentExtra: `${theme.lineHeights.small}px`,
  };

  return {
    colors,
    borderWidths,
    radii,
    space,
    fontWeights,
    fontSizes,
    lineHeights,
  };
};

type CardTheme = ReturnType<typeof getCardTheme>;

export { getCardTheme };
export type { CardTheme };
