// @hero-design/react/src/theme/components/carousel.ts
import type { GlobalTheme } from '../global';

const getCarouselTheme = (theme: GlobalTheme) => {
  const colors = {
    light: theme.colors.defaultLightBackground,
    dark: theme.colors.neutralBackground,
  };

  const space = {
    default: `${theme.space.large}px`,
    listRowMargin: `0 ${theme.space.xxsmall}px`,
    listColumnMargin: `${theme.space.xsmall}px ${theme.space.xxsmall}px`,
    navigationButtonColumnMargin: `${theme.space.large / 2}px`,
    slidePadding: `0px ${theme.space.small}px`,
    containerVerticalPadding: `${theme.space.xlarge}px`,
    containerHorizontalPadding: `${theme.space.large}px`,
  };

  const sizes = {
    basic: `${theme.sizes.xsmall}px`,
    default: `${theme.sizes.medium}px`,
    active: `${theme.sizes.large}px`,
  };

  const radii = {
    default: `${theme.radii.base}px`,
  };

  const fontSizes = {
    navigationButton: `${theme.fontSizes.xxxlarge}px`,
  };

  return { colors, space, sizes, radii, fontSizes };
};

type CarouselTheme = ReturnType<typeof getCarouselTheme>;

export { getCarouselTheme };
export type { CarouselTheme };
