// @hero-design/react/src/theme/components/chart.ts
import type { GlobalTheme } from '../global';

const getChartTheme = (theme: GlobalTheme) => {
  const colors = {
    tooltip: {
      foreground: 'transparent',
      background: theme.colors.defaultLightBackground,
      text: theme.colors.text,
    },
    pie: {
      background: theme.colors.defaultLightBackground,
      iconBackground: theme.colors.defaultGreyBackground,
    },
  };

  const sizes = {
    tooltip: {
      arrowSize: `${theme.sizes.xsmall}px`,
    },
    legend: {
      dot: theme.space.small,
    },
    pie: {
      small: 64,
      medium: 112,
      large: 200,
      stokes: {
        small: 6,
        medium: 10,
        large: 16,
      },
      hoverStrokes: {
        small: theme.space.xsmall,
        medium: theme.space.xsmall,
        large: theme.space.small,
      },
      badgeSize: `${theme.space.small}px`,
      iconContainerSize: '75%',
      restHoverStrokes: {
        small: 4,
        medium: 6,
        large: 10,
      },
    },

    multiplePie: {
      large: 200,
      stokes: {
        large: 10,
      },
      hoverStrokes: {
        large: theme.space.xsmall,
      },
      restHoverStrokes: {
        large: 4,
      },
    },
  };

  const spaces = {
    titleContainerMargin: `${theme.space.medium}px 0 ${theme.space.small}px 0 `,
    legend: {
      margin: `${theme.space.small}px 0 0 0 `,
      gap: `${theme.space.small}px`,
    },
    pie: {
      padding: theme.space.medium,
      tooltipTopPosition: 12,
      badgeMargin: `${theme.space.xsmall}px`,

      sliceGap: theme.space.xsmall,
      iconPadding: {
        small: `${theme.space.xsmall}px`,
        medium: `${theme.space.large}px`,
        large: `${theme.space.xlarge}px`,
      },
    },
    multiplePie: {
      badgeMargin: `${theme.space.xsmall}px`,
      titleMargin: `${theme.space.medium}px 0 0 0 `,
      sliceGap: theme.space.small,
      strokeGap: theme.space.small,
      iconPadding: {
        small: `${theme.space.xsmall}px`,
        medium: `${theme.space.large}px`,
        large: `${theme.space.xlarge}px`,
      },
      subTitleMargin: `${theme.space.small}px 0 0 0 `,
    },
  };

  const shadows = {
    pie: {
      slice: 'drop-shadow(1px 1px 4px rgb(0 0 0 / 0.3))',
    },
    multiplePie: {
      slice: 'drop-shadow(1px 1px 4px rgb(0 0 0 / 0.3))',
    },
  };

  const radii = {
    pie: {
      iconContainerRadius: '9999px',
    },
    legend: {
      dotRadius: `${theme.space.small}px`,
    },
  };

  return { colors, sizes, shadows, spaces, radii };
};

type ChartTheme = ReturnType<typeof getChartTheme>;

export { getChartTheme };
export type { ChartTheme };
