// @hero-design/react/src/theme/components/comment.ts
import type { GlobalTheme } from '../global';

const getCommentTheme = (theme: GlobalTheme) => {
  const colors = {
    author: theme.colors.text,
    authorDetails: theme.colors.disabledText,
    time: theme.colors.disabledText,
    highlight: theme.colors.defaultBackground,
    content: theme.colors.bodyText,
  };

  const sizes = {
    avatarSize: `${theme.sizes.xlarge}px`,
  };

  const space = {
    marginBottom: `${theme.space.medium}px`,
    nestedMargin: `${theme.space.medium}px 0 0 ${theme.space.xlarge}px`,
    avatarMarginRight: `${theme.space.medium}px`,
    avatarMarginLeft: `${theme.space.small}px`,
    padding: `${theme.space.small}px`,
    timeSeparatorPadding: `0 ${theme.space.small}px`,
    contentMarginTop: `${theme.space.xsmall}px`,
    reactionMarginTop: `${theme.space.xsmall}px`,
    editorActionsPadding: `${theme.space.small}px`,
  };

  const fontSizes = {
    author: `${theme.fontSizes.medium}px`,
    authorDetails: `${theme.fontSizes.small}px`,
    content: `${theme.fontSizes.medium}px`,
    time: `${theme.fontSizes.small}px`,
  };

  const fontWeights = {
    author: theme.fontWeights.semiBold,
    authorDetails: theme.fontWeights.semiBold,
    time: theme.fontWeights.regular,
    content: theme.fontWeights.regular,
  };

  const lineHeights = {
    author: `${theme.lineHeights.medium}px`,
    authorDetails: `${theme.lineHeights.small}px`,
    content: `${theme.lineHeights.medium}px`,
    time: `${theme.lineHeights.small}px`,
  };

  const radii = {
    container: `${theme.radii.base}px`,
  };

  return { colors, sizes, space, fontSizes, fontWeights, lineHeights, radii };
};

type CommentTheme = ReturnType<typeof getCommentTheme>;

export { getCommentTheme };
export type { CommentTheme };
