// @hero-design/react/src/theme/components/filters.ts
import { GlobalTheme } from '../global';

const getFiltersTheme = (theme: GlobalTheme) => {
  const sizes = {
    filterButton: {
      small: `${theme.sizes.large}px`,
      medium: `${theme.sizes.xlarge}px`,
    },
    filterDropdownWidth: `250px`,
    filterGroupWidth: '218px',
    filterGroupLargeDropdownWidth: '686px', // 3 * filterGroupWidth + 2 * filterGroupDropdownItemGap
    filterGroupXLargeDropdownWidth: '920px', // 4 * filterGroupWidth + 3 * filterGroupDropdownItemGap
  };

  const colors = {
    filterButtonBackground: theme.colors.defaultLightBackground,
    filterButtonBorder: theme.colors.defaultBorder,
    filterButtonHoverBorder: theme.colors.black,
    filterButtonActiveBackground: theme.colors.palette.smaltLight75,
    filterButtonActiveBorder: theme.colors.black,
    filterLabelText: theme.colors.subduedText,
    filterButtonText: theme.colors.text,
  };

  const space = {
    tagListMarginTop: `${theme.space.small}px`,
    filterButtonPadding: `${theme.space.small}px ${theme.space.medium}px`,
    filterButtonGap: `${theme.space.small}px`,
    filterButtonIconPadding: `${theme.space.small}px`,
    filterDropdownPadding: `${theme.space.medium}px`,
    filterLabelContentPadding: `${theme.space.xsmall}px`,
    filterGroupDropdownPadding: `${theme.space.medium}px`,
    filterGroupDropdownItemGap: `${theme.space.medium}px`,
    filterGroupFilterMarginBottom: `${theme.space.medium}px`,
  };

  const fontSizes = {
    filterButton: {
      small: `${theme.fontSizes.small}px`,
      medium: `${theme.fontSizes.medium}px`,
    },
  };

  const fontWeights = {
    filterButton: theme.fontWeights.semiBold,
    filterLabel: theme.fontWeights.semiBold,
    filterBy: theme.fontWeights.semiBold,
  };

  const lineHeights = {
    filterButton: {
      small: `${theme.lineHeights.small}px`,
      medium: `${theme.lineHeights.medium}px`,
    },
  };

  const borderWidths = {
    filterButton: `${theme.borderWidths.base}px`,
  };

  const radii = {
    filterButton: `${theme.radii.base * 2}px`,
  };

  const shadows = {};

  return {
    sizes,
    colors,
    space,
    fontSizes,
    fontWeights,
    lineHeights,
    borderWidths,
    radii,
    shadows,
  };
};

type FiltersTheme = ReturnType<typeof getFiltersTheme>;

export { getFiltersTheme };
export type { FiltersTheme };
