// @hero-design/react/src/theme/components/lineChart.ts
import type { GlobalTheme } from '../global';

const getLineChartTheme = (theme: GlobalTheme) => {
  const colors = {
    xAxisTickColor: theme.colors.subduedText,
    yAxisTickColor: theme.colors.subduedText,
    verticalTitleFillColor: theme.colors.subduedText,
    gridLineBackgroundColor: theme.colors.neutralBackground,
    tooltipBackgroundColor: theme.colors.defaultLightBackground,
  };

  const borderWidth = {
    line: {
      default: theme.borderWidths.base * 2,
      highlighted: theme.borderWidths.base * 4,
      noHighlighted: theme.borderWidths.base * 2,
    },
  };

  const sizes = {
    navigationButton: theme.sizes.large,
    yAxisTitleWidth: theme.sizes.large,
    yAxisWidth: theme.sizes.xxlarge,
    xAxisHeight: theme.sizes.xlarge,
    xAxisLabelMinWidth: theme.sizes.medium,
  };

  const space = {
    tooltipBottomMargin: theme.space.medium,
    tooltipChildMargin: theme.space.small,
    legendContainerMarginTop: theme.space.medium,
    legendContainerChildMarginBottom: theme.space.small,
    xAxisTransform: theme.space.medium,
    xAxisLabelHorizontalPadding: theme.space.small,
    yAxisTransform: -theme.space.xsmall,
    moreTooltipPadding: theme.space.small,
    navigateButtonPadding: theme.space.xsmall,
  };

  const fontWeights = {
    xAxisTick: theme.fontWeights.regular,
    yAxisTick: theme.fontWeights.regular,
    verticalTitle: theme.fontWeights.light,
  };

  const fontSizes = {
    xAxisTick: `${theme.fontSizes.xsmall}px`,
    yAxisTick: `${theme.fontSizes.xsmall}px`,
  };

  const lineHeights = {
    xAxisTick: `${theme.lineHeights.xsmall}px`,
    yAxisTick: `${theme.lineHeights.xsmall}px`,
  };

  const boxShadows = {
    tooltip: theme.shadows.default,
  };

  const opacity = {
    gridLine: 0.3,
  };

  const strokeDasharray = {
    gridLine: '5 5',
    line: '4 4',
  };

  return {
    strokeDasharray,
    colors,
    space,
    fontWeights,
    fontSizes,
    lineHeights,
    boxShadows,
    sizes,
    borderWidth,
    opacity,
  };
};

type LineChartTheme = ReturnType<typeof getLineChartTheme>;

export { getLineChartTheme };
export type { LineChartTheme };
