// @hero-design/react/src/theme/components/menu.ts
import type { GlobalTheme } from '../global';

const getMenuTheme = (theme: GlobalTheme) => {
  const colors = {
    background: theme.colors.defaultLightBackground,
    itemActiveBackground: theme.colors.defaultGreyBackground,
    item: theme.colors.text,
    itemDisabled: theme.colors.disabledText,
    itemDanger: theme.colors.danger,
  };

  const space = {
    itemPadding: `${theme.space.xsmall}px ${theme.space.medium}px`,
    iconMarginRight: `${theme.space.small}px`,
    rightElementMarginLeft: `${theme.space.xlarge}px`,
  };

  const fontSizes = {
    item: `${theme.fontSizes.medium}px`,
    icon: `${theme.fontSizes.small}px`,
  };

  const fontWeights = {
    item: theme.fontWeights.regular,
    itemActive: theme.fontWeights.semiBold,
  };

  const lineHeights = {
    item: `${theme.lineHeights.medium}px`,
    icon: `${theme.lineHeights.small}px`,
  };

  return { colors, space, fontSizes, fontWeights, lineHeights };
};

type MenuTheme = ReturnType<typeof getMenuTheme>;

export { getMenuTheme };
export type { MenuTheme };
