// @hero-design/react/src/theme/components/notification.ts
import type { GlobalTheme } from '../global';

const getNotificationTheme = (theme: GlobalTheme) => {
  const space = {
    iconMarginTop: `${theme.space.xxsmall}px`,
    iconMarginRight: `${theme.space.medium}px`,
    padding: `${theme.space.medium}px`,
    titleMarginBottom: `${theme.space.xsmall}px`,
  };

  const fontSizes = {
    title: `${theme.fontSizes.large}px`,
    content: `${theme.fontSizes.medium}px`,
  };

  const fontWeights = {
    title: theme.fontWeights.semiBold,
    content: theme.fontWeights.regular,
  };

  const lineHeights = {
    title: `${theme.lineHeights.large}px`,
    content: `${theme.lineHeights.medium}px`,
    icon: `${theme.lineHeights.large}px`,
  };

  const radii = {
    default: `${theme.radii.medium}px`,
  };

  const shadows = {
    default: theme.shadows.default,
  };

  const colors = {
    title: theme.colors.text,
    content: theme.colors.bodyText,
    bg: theme.colors.defaultLightBackground,
  };

  return { space, fontSizes, fontWeights, lineHeights, radii, shadows, colors };
};

type NotificationTheme = ReturnType<typeof getNotificationTheme>;

export { getNotificationTheme };
export type { NotificationTheme };
