// @hero-design/react/src/theme/components/pagination.ts
import type { GlobalTheme } from '../global';

const getPaginationTheme = (theme: GlobalTheme) => {
  const space = {
    leftRightPadding: `${theme.space.xsmall}px`,
  };

  const colors = {
    border: theme.colors.defaultBorder,
    background: theme.colors.defaultLightBackground,
    number: theme.colors.text,
    numberActive: theme.colors.primary,
    numberDisabled: theme.colors.subduedText,
  };

  const borderWidths = {
    default: `${theme.borderWidths.base}px`,
  };

  const sizes = {
    height: `${theme.sizes.large}px`,
    minWidth: `${theme.sizes.large}px`,
  };

  const fontSizes = {
    default: `${theme.fontSizes.medium}px`,
  };

  const fontWeights = {
    default: theme.fontWeights.semiBold,
  };

  const radii = {
    default: `${theme.radii.base}px`,
  };

  return { space, colors, borderWidths, sizes, fontSizes, fontWeights, radii };
};

type PaginationTheme = ReturnType<typeof getPaginationTheme>;

export { getPaginationTheme };
export type { PaginationTheme };
