// @hero-design/react/src/theme/components/selectButton.ts
import type { GlobalTheme } from '../global';

const getSelectButtonTheme = (theme: GlobalTheme) => {
  const sizes = {
    small: `${theme.sizes.large}px`,
    medium: `${theme.sizes.xlarge}px`,
    large: `${theme.sizes.xxxlarge}px`,
    groupSmallWidth: `calc(100% + ${theme.space.xsmall}px)`,
    groupMediumWidth: `calc(100% + ${theme.space.small}px)`,
    groupLargeWidth: `calc(100% + ${theme.space.small}px)`,
  };

  const space = {
    smallPadding: `${theme.space.xsmall}px ${theme.space.small}px`,
    mediumPadding: `${theme.space.small}px ${theme.space.medium}px`,
    largePadding: `${theme.space.small}px ${theme.space.medium}px`,
    groupSmallMargin: `-${theme.space.xsmall}px 0 0 -${theme.space.xsmall}px`,
    groupMediumMargin: `-${theme.space.small}px 0 0 -${theme.space.small}px`,
    groupLargeMargin: `-${theme.space.small}px 0 0 -${theme.space.small}px`,
    smallMargin: `${theme.space.xsmall}px 0 0 ${theme.space.xsmall}px`,
    mediumMargin: `${theme.space.small}px 0 0 ${theme.space.small}px`,
    largeMargin: `${theme.space.small}px 0 0 ${theme.space.small}px`,
  };

  const fontSizes = {
    small: `${theme.fontSizes.small}px`,
    medium: `${theme.fontSizes.medium}px`,
    large: `${theme.fontSizes.xlarge}px`,
  };

  const fontWeights = {
    default: theme.fontWeights.regular,
  };

  const borderWidths = {
    default: `${theme.borderWidths.base}px`,
  };

  const radii = {
    default: `${theme.radii.base}px`,
  };

  const colors = {
    background: theme.colors.defaultLightBackground,
    selectedBackground: theme.colors.primary,
    hover: theme.colors.activePrimary,
    defaultBorder: theme.colors.defaultBorder,
    defaultText: theme.colors.text,
    disabledText: theme.colors.disabledText,
    disabledSelectedText: theme.colors.lightText,
    disabledBackground: theme.colors.defaultBackground,
    disabledSelectedBackground: theme.colors.neutralBackground,
    selectedText: theme.colors.lightText,
  };

  return { sizes, space, fontSizes, fontWeights, borderWidths, radii, colors };
};

type SelectButtonTheme = ReturnType<typeof getSelectButtonTheme>;

export { getSelectButtonTheme };
export type { SelectButtonTheme };
