// @hero-design/react/src/theme/components/sideBar.ts
import type { GlobalTheme } from '../global';

const getSideBarTheme = (theme: GlobalTheme) => {
  const colors = {
    sideBarBackground: theme.colors.primaryNeutralBackground,
    sideBarItemHoverBackground: `${theme.colors.defaultLightBackground}26`,
    sideBarItemActiveBackground: theme.colors.defaultLightBackground,
    sideBarText: theme.colors.lightText,
    sideBarActiveText: theme.colors.text,
    sideBarCountBackground: theme.colors.defaultLightBackground,
    sideBarCountText: theme.colors.text,
    panelBackground: theme.colors.defaultLightBackground,
    panelItemHoverBackground: theme.colors.primary,
    panelItemActiveBackground: theme.colors.primaryLightBackground,
    sidebarItemIconBackground: theme.colors.hoverPrimary,
    scrollBarThumb: `${theme.colors.neutralDarkBackground}b3`,
  };

  const space = {
    sideBarHeaderItemsContainerPadding: `${theme.space.small}px ${theme.space.small}px`,
    sideBarItemsContainerPadding: `${theme.space.medium}px ${theme.space.xsmall}px`,
    sideBarItemPadding: `${theme.space.small}px`,
    sideBarItemMarginBottom: `${theme.space.medium}px`,
    sideBarNotiBadge: `${theme.space.xxsmall}px ${theme.space.xsmall}px`,
    panelContainerPadding: `${theme.space.medium}px ${theme.space.xsmall}px ${theme.space.medium}px ${theme.space.large}px `,
    panelBackButtonPadding: `${theme.space.small}px ${theme.space.medium}px ${theme.space.medium}px`,
    panelSubHeaderPadding: `${theme.space.small}px ${theme.space.medium}px`,
    backIconMarginRight: `${theme.space.xsmall}px`,
    panelItemContainerMarginTop: `${theme.space.large}px`,
    panelItemPadding: `${theme.space.small}px ${theme.space.medium}px`,
    panelItemMarginTop: `${theme.space.xxsmall}px`,
    scrollBarMargin: `${theme.space.medium}px 0px`,
  };

  const sizes = {
    sideBarWidth: '92px',
    sideBarScrollWrapperWidth: '103px',
    itemWidth: '84px',
    itemIconWidth: '40px',
    itemIconHeight: '28px',
    sideBarNotiBadge: '20px',
    panelWidth: '255px',
    panelWrapperWidth: '266px',
    panelHeaderHeight: '44px',
    panelItemBadge: '22px',
    scrollBar: '11px',
  };

  const radii = {
    sideBarWrapperRight: `${theme.radii.large}px`,
    sideBarItem: `${theme.radii.medium}px`,
    panelContainter: `0px ${theme.radii.large}px ${theme.radii.large}px 0px`,
    panelItem: `8px`, // TODO: Remove hard code
  };

  const fontSizes = {
    headerIcon: `${theme.fontSizes.xxxxxlarge}px`,
    sideBarNotiBadge: `${theme.fontSizes.xsmall}px`,
    panelHeader: `${theme.fontSizes.xlarge}px`,
    panelSubheader: `${theme.fontSizes.large}px`,
  };

  const fontWeights = {
    panelHeader: theme.fontWeights.bold,
  };

  const borderWidths = {
    scrollBar: `${theme.borderWidths.base * 2}px`,
  };

  const shadows = {
    panelContainerShadow: `3px 3px 6px ${theme.colors.palette.black}1A`,
  };

  return {
    colors,
    space,
    sizes,
    fontSizes,
    fontWeights,
    radii,
    borderWidths,
    shadows,
  };
};

type SideBarTheme = ReturnType<typeof getSideBarTheme>;

export { getSideBarTheme };
export type { SideBarTheme };
