// @hero-design/react/src/theme/components/slider.ts
import type { GlobalTheme } from '../global';

const getSliderTheme = (theme: GlobalTheme) => {
  const colors = {
    background: theme.colors.defaultGreyBackground,
    thumbBackground: theme.colors.defaultLightBackground,
    thumbBorder: theme.colors.primaryRingBackground,
    track: theme.colors.primary,
    disabled: theme.colors.neutralDarkBackground,
  };

  const borderWidths = {
    thumb: `${theme.borderWidths.base * 2}px`,
  };

  const radii = {
    default: `${theme.radii.base}px`,
  };

  const sizes = {
    trackHeight: `${theme.sizes.xsmall}px`,
    thumb: `${theme.sizes.medium}px`,
    inputHeight: `${theme.sizes.medium}px`,
    tooltipMinWidth: `${theme.sizes.medium}px`,
  };

  const space = {
    trackOffsetTop: `calc((${sizes.inputHeight} - ${sizes.trackHeight}) / 2)`,
    tooltipOffsetLeft: `calc(${sizes.thumb} / 2 - ${borderWidths.thumb})`,
    tooltipPointerOffsetTop: `${theme.sizes.large + theme.space.xxsmall}px`,
  };

  const shadows = {
    thumb: `0 0 0 3px ${theme.colors.primaryBackground}`,
  };

  return { colors, borderWidths, radii, sizes, space, shadows };
};

type SliderTheme = ReturnType<typeof getSliderTheme>;

export { getSliderTheme };
export type { SliderTheme };
