// @hero-design/react/src/theme/components/spinner.ts
import type { GlobalTheme } from '../global';

const getSpinnerTheme = (theme: GlobalTheme) => {
  const sizes = {
    small: `${theme.sizes.small}px`,
    medium: `${theme.sizes.medium}px`,
    large: `${theme.sizes.large}px`,
    smallContainer: `${theme.sizes.small + theme.space.medium * 2}px`,
    mediumContainer: `${theme.sizes.medium + theme.space.xlarge * 2}px`,
    largeContainer: `${theme.sizes.large + theme.space.xxxxlarge * 2}px`,
  };

  const colors = {
    default: theme.colors.primary,
  };

  const space = {
    smallMargin: `${theme.space.medium}px auto`,
    mediumMargin: `${theme.space.xlarge}px auto`,
    largeMargin: `${theme.space.xxxxlarge}px auto`,
  };

  const shadows = {
    small: `${theme.sizes.small}px 0 0 ${theme.colors.primary},
    0 ${theme.sizes.small}px 0 ${theme.colors.primary},
    -${theme.sizes.small}px 0 ${theme.colors.primary}50,
    0 -${theme.sizes.small}px 0 ${theme.colors.primary}80`,

    medium: `${theme.sizes.medium}px 0 0 ${theme.colors.primary},
    0 ${theme.sizes.medium}px 0 ${theme.colors.primary},
    -${theme.sizes.medium}px 0 0 ${theme.colors.primary}50,
    0 -${theme.sizes.medium}px 0 ${theme.colors.primary}80`,

    large: `${theme.sizes.large}px 0 0 ${theme.colors.primary},
    0 ${theme.sizes.large}px 0 ${theme.colors.primary},
    -${theme.sizes.large}px 0 0 ${theme.colors.primary}50,
    0 -${theme.sizes.large}px 0 ${theme.colors.primary}80`,
  };

  const fontSizes = {
    small: `${theme.fontSizes.small}px`,
    medium: `${theme.fontSizes.medium}px`,
    large: `${theme.fontSizes.large}px`,
  };

  return { sizes, colors, space, shadows, fontSizes };
};

type SpinnerTheme = ReturnType<typeof getSpinnerTheme>;

export { getSpinnerTheme };
export type { SpinnerTheme };
