// @hero-design/react/src/theme/components/switch.ts
import type { GlobalTheme } from '../global';

const getSwitchTheme = (theme: GlobalTheme) => {
  const sizes = {
    smallWidth: `${theme.sizes.xlarge}px`,
    smallHeight: `${theme.sizes.medium}px`,
    sliderSmall: `${theme.sizes.medium - theme.sizes.xsmall}px`,
    mediumWidth: `${theme.sizes.xxxlarge}px`,
    mediumHeight: `${theme.sizes.large}px`,
    sliderMedium: `${theme.sizes.large - theme.sizes.small}px`,
  };

  const colors = {
    background: theme.colors.defaultDarkGreyBackground,
    checkedBackground: theme.colors.primary,
    disabledBackground: theme.colors.defaultDarkGreyBackground,
    checkedDisabledBackground: theme.colors.neutralBackground,
    handleBackgroundColor: theme.colors.defaultLightBackground,
    iconColor: theme.colors.lightText,
    loadingIconColor: theme.colors.defaultBorder,
    disabledLoadingIconColor: theme.colors.neutralBackground,
    checkedLoadingIconColor: theme.colors.primary,
    checkedDisabledLoadingIconColor: theme.colors.primaryBackground,
  };

  const radii = {
    small: `${theme.sizes.small}px`,
    medium: `${theme.sizes.medium}px`,
  };

  const space = {
    sliderSmallTop: `${theme.sizes.xxsmall}px`,
    sliderSmallLeft: `${theme.sizes.xxsmall}px`,
    sliderSmallCheckedLeft: `${
      theme.sizes.xlarge - theme.sizes.medium + theme.sizes.xsmall / 2
    }px`,
    sliderMediumTop: `${theme.sizes.xsmall}px`,
    sliderMediumLeft: `${theme.sizes.xsmall}px`,
    sliderMediumCheckedLeft: `${
      theme.sizes.xxxlarge - theme.sizes.large + theme.sizes.small / 2
    }px`,

    iconSmallTop: 0,
    iconSmallLeft: `${theme.sizes.xsmall}px`,
    iconSmallMarginLeft: 0,
    iconMediumTop: `${theme.sizes.xsmall}px`,
    iconMediumLeft: `${theme.sizes.xsmall}px`,
    iconMediumMarginLeft: `${theme.sizes.xxsmall}px`,

    loadingIconSmallTop: 0,
    loadingIconSmallLeft: `${theme.sizes.xxsmall}px`,
    loadingIconSmallCheckedLeft: `${
      theme.sizes.xlarge - theme.sizes.medium + theme.sizes.xxsmall / 2
    }px`,
    loadingIconMediumTop: `${theme.sizes.xsmall}px`,
    loadingIconMediumLeft: `${theme.sizes.xsmall}px`,

    loadingIconMediumCheckedLeft: `${
      theme.sizes.xxxlarge - theme.sizes.large + theme.sizes.small / 2
    }px`,
  };

  const fontSizes = {
    iconSmall: `${theme.fontSizes.small}px`,
    iconMedium: `${theme.fontSizes.large}px`,
    disabledIconSmall: `${theme.fontSizes.small}px`,
    disabledIconMedium: `${theme.fontSizes.large}px`,
    loadingIconSmall: `${theme.fontSizes.xsmall}px`,
    loadingIconMedium: `${theme.fontSizes.medium}px`,
  };

  return { sizes, colors, radii, space, fontSizes };
};

type SwitchTheme = ReturnType<typeof getSwitchTheme>;

export { getSwitchTheme };
export type { SwitchTheme };
