// @hero-design/react/src/theme/components/timePicker.ts
import type { GlobalTheme } from '../global';

const getTimePickerTheme = (theme: GlobalTheme) => {
  const colors = {
    columnBorder: theme.colors.defaultBorder,
    cell: theme.colors.text,
    cellActive: theme.colors.defaultBackground,
    cellHover: theme.colors.primaryLightBackground,
    cellBackground: theme.colors.defaultLightBackground,
  };

  const space = {
    cellPadding: `0 0 0 ${theme.space.medium}px`,
  };

  const fontSizes = {
    cell: `${theme.fontSizes.medium}px`,
  };

  const fontWeights = {
    cellActive: theme.fontWeights.semiBold,
  };

  const lineHeights = {
    cell: `${theme.lineHeights.xxxlarge}px`,
  };

  const sizes = {
    columnWidth: '56px',
    columnAfterHeight: '160px',
    columnHeight: '192px',
    cellHeight: `${theme.sizes.xlarge}px`,
  };

  const borderWidths = {
    column: `${theme.borderWidths.base}px`,
  };

  return {
    colors,
    space,
    fontSizes,
    fontWeights,
    lineHeights,
    sizes,
    borderWidths,
  };
};

type TimePickerTheme = ReturnType<typeof getTimePickerTheme>;

export { getTimePickerTheme };
export type { TimePickerTheme };
