// @hero-design/react/src/theme/components/tooltip.ts
import type { GlobalTheme } from '../global';

const getTooltipTheme = (theme: GlobalTheme) => {
  const sizes = {
    arrow: `${theme.sizes.small}px`,
  };

  const colors = {
    dark: {
      text: theme.colors.lightText,
      background: theme.colors.defaultDarkBackground,
      border: theme.colors.defaultDarkBackground,
    },
    light: {
      text: theme.colors.text,
      background: theme.colors.defaultLightBackground,
      border: theme.colors.defaultBorder,
    },
  };

  const space = {
    wrapperPadding: `${theme.space.xsmall}px ${theme.space.small}px`,
    margin: theme.space.small,
  };

  const fontSizes = {
    default: `${theme.fontSizes.small}px`,
  };

  const fontWeights = {
    default: theme.fontWeights.regular,
  };

  const lineHeights = {
    default: `${theme.lineHeights.small}px`,
  };

  const radii = {
    default: `${theme.radii.base}px`,
  };

  const borderWidths = {
    default: '0px',
  };

  const shadows = {
    dark: 'unset',
    light: theme.shadows.default,
  };

  return {
    sizes,
    colors,
    space,
    fontSizes,
    fontWeights,
    lineHeights,
    radii,
    borderWidths,
    shadows,
  };
};

type TooltipTheme = ReturnType<typeof getTooltipTheme>;

export { getTooltipTheme };
export type { TooltipTheme };
