// @hero-design/react/src/theme/components/widget.ts
import type { GlobalTheme } from '../global';

const getWidgetTheme = (theme: GlobalTheme) => {
  const colors = {
    containerBackground: theme.colors.defaultLightBackground,
    containerBorder: theme.colors.defaultBorder,
    containerActiveBorder: theme.colors.primary,
  };

  const space = {
    headerPadding: `${theme.space.large}px ${theme.space.large}px ${theme.space.medium}px`,
    headerItemGap: `${theme.space.small}px`,
    contentPadding: `0px ${theme.space.large}px ${theme.space.large}px`,
    resizeHandlePadding: `${theme.space.xsmall}px ${theme.space.small}px ${theme.space.small}px ${theme.space.xsmall}px`,
    resizeHandleOffset: `-${theme.space.xsmall}px`,
  };

  const borderWidths = {
    container: `${theme.borderWidths.base * 4}px`,
  };

  const radii = {
    container: `${theme.radii.large}px`,
  };

  return {
    colors,
    space,
    borderWidths,
    radii,
  };
};

type WidgetTheme = ReturnType<typeof getWidgetTheme>;

export { getWidgetTheme };

export type { WidgetTheme };
