// @hero-design/react/src/components/Chart/ColumnChart/StyledColumnChart.tsx
import styled from 'styled-components';
import Box from '../../Box';
import { Arrow, TooltipWrapper } from '../../Tooltip/StyledTooltip';
import IconButton from '../../Button/IconButton';

const StyledTooltipContentWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  & > *:not(:last-child) {
    margin-right: ${({ theme }) =>
      `${theme.__hd__.columnChart.space.tooltipChildMargin}px`};
  }
`;

const StyledColumnChartContainer = styled.div`
  position: relative;
  height: 100%;
  width: 100%;
  svg {
    position: relative;
    z-index: 2;
  }
  .x-axis {
    text {
      font-weight: ${({ theme }) =>
        theme.__hd__.columnChart.fontWeights.xAxisTick};
      font-size: ${({ theme }) => theme.__hd__.columnChart.fontSizes.xAxisTick};
      line-height: ${({ theme }) =>
        theme.__hd__.columnChart.lineHeights.xAxisTick};
      color: ${({ theme }) => theme.__hd__.columnChart.colors.xAxisTickColor};
    }
  }
  .y-axis {
    text {
      font-weight: ${({ theme }) =>
        theme.__hd__.columnChart.fontWeights.yAxisTick};
      font-size: ${({ theme }) => theme.__hd__.columnChart.fontSizes.yAxisTick};
      line-height: ${({ theme }) =>
        theme.__hd__.columnChart.lineHeights.yAxisTick};
      color: ${({ theme }) => theme.__hd__.columnChart.colors.yAxisTickColor};
    }
  }

  .vertical-title {
    font-size: ${({ theme }) =>
      theme.__hd__.columnChart.fontSizes.verticalTitle};
    font-weight: ${({ theme }) =>
      theme.__hd__.columnChart.fontWeights.verticalTitle};
    fill: ${({ theme }) =>
      theme.__hd__.columnChart.colors.verticalTitleFillColor};
  }
`;

const StyledLegendContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  > *:not(:last-child) {
    margin-bottom: ${({ theme }) =>
      `${theme.__hd__.columnChart.space.legendContainerChildMarginBottom}px`};
  }
  ${TooltipWrapper} {
    background: ${({ theme }) =>
      theme.__hd__.columnChart.colors.tooltipBackgroundColor};
    box-shadow: ${({ theme }) => theme.__hd__.columnChart.boxShadows.tooltip};
    padding: ${({ theme }) =>
      theme.__hd__.columnChart.space.moreTooltipPadding};
  }

  ${Arrow} {
    &::before {
      background: ${({ theme }) =>
        theme.__hd__.columnChart.colors.tooltipBackgroundColor};
    }
  }
`;

const StyledNavigateButton = styled(IconButton)`
  width: ${({ theme }) =>
    `${theme.__hd__.columnChart.sizes.navigationButton}px`};
  height: ${({ theme }) =>
    `${theme.__hd__.columnChart.sizes.navigationButton}px`};
`;

const StyledNavigateButtonWrapper = styled(Box)<{
  themeTop: string | number;
  themePosition: 'right' | 'left';
  themeLeft?: number;
}>`
  position: absolute;
  display: grid;
  place-items: center;
  top: ${({ themeTop }) =>
    typeof themeTop === 'number' ? `${themeTop}px` : themeTop};
  ${({ themePosition, themeLeft }) => {
    const positionMapping = {
      left: `left: ${themeLeft}px`,
      right: `right: 0`,
    };

    return positionMapping[themePosition];
  }};
  z-index: 3;
`;
export {
  StyledTooltipContentWrapper,
  StyledColumnChartContainer,
  StyledLegendContainer,
  StyledNavigateButton,
  StyledNavigateButtonWrapper,
};
