// @hero-design/react/src/components/Chart/Legend/StyledLegend.tsx
import styled, { css } from 'styled-components';
import Box from '../../Box';
import Checkbox from '../../Checkbox/Checkbox';
import { HiddenInput, LabelText } from '../../Checkbox/StyledCheckbox';

export const StyledLegendDot = styled.span<{ themeColor: string }>`
  width: ${({ theme }) => theme.__hd__.chart.sizes.legend.dot}px;
  height: ${({ theme }) => theme.__hd__.chart.sizes.legend.dot}px;
  background-color: ${({ themeColor }) => themeColor};
  border-radius: ${({ theme }) => theme.__hd__.chart.radii.legend.dotRadius};
`;

const StyledLegendCircle = styled(Box)<{ themeColor: string }>`
  width: ${({ theme }) => theme.__hd__.legend.sizes.legendWidth};
  height: ${({ theme }) => theme.__hd__.legend.sizes.legendHeight};
  border-radius: ${({ theme }) => theme.__hd__.legend.radii.legendCircle};
  display: inline-flex;
  background-color: ${({ themeColor }) => themeColor};
  flex-shrink: 0;
`;

const StyledLegendContainer = styled(Box)`
  display: flex;
  align-items: center;
  & > ${StyledLegendCircle}:not(:last-child) {
    margin-right: ${({ theme }) =>
      theme.__hd__.legend.space.legendCircleMarginRight};
  }
`;

const StyledLegendCheckBox = styled(Checkbox)<{ themeColor: string }>`
  > ${LabelText} {
    &:hover {
      &::before {
        border-color: ${({ themeColor }) => themeColor};
      }
    }
  }

  > ${HiddenInput} {
    &:checked ~ ${LabelText} {
      &::before {
        border-color: ${({ themeColor }) => themeColor};
        background: ${({ themeColor }) => themeColor};
      }
    }
  }
`;

const StyledLegendBox = styled(Box)<{
  themeDirection?: 'vertical' | 'horizontal';
}>`
  display: flex;
  align-items: center;
  gap: ${({ theme }) => `${theme.__hd__.columnChart.space.legendBoxGap}px`};
  ${({ themeDirection }) =>
    themeDirection === 'vertical' &&
    css`
      flex-direction: column;
    `};
`;

export {
  StyledLegendCircle,
  StyledLegendContainer,
  StyledLegendCheckBox,
  StyledLegendBox,
};
