// @hero-design/react/src/components/Steps/StyledSteps.tsx
import styled, { css } from 'styled-components';

import { StyledIcon } from '../Icon/StyledIcon';

const Wrapper = styled.div`
  display: flex;
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.steps.space.padding};
`;

const StepWrapper = styled.div<{
  themeNumberOfSteps: number;
  themeVariant: 'basic' | 'compact';
}>`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
  margin: 0;
  padding: 0;

  width: ${(props): number => 100 / props.themeNumberOfSteps}%;
  max-width: ${({ theme }) => theme.__hd__.steps.sizes.maxWidthMobile};

  ${({ themeVariant }) =>
    themeVariant === 'basic' &&
    css`
      @media (min-width: ${({ theme }) => theme.breakpoints.small}) {
        max-width: ${({ theme }) => theme.__hd__.steps.sizes.maxWidthDesktop};
      }
    `};
`;

const StepIconWrapper = styled.div<{ themeClickable: boolean }>`
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
  margin: 0;
  padding: 0;

  ${({ themeClickable }) => {
    switch (themeClickable) {
      case true:
        return css`
          pointer-events: auto;
          cursor: pointer;
        `;
      case false:
        return css`
          pointer-events: none;
        `;
    }
  }};

  width: ${({ theme }) => theme.__hd__.steps.sizes.iconSize};
  height: ${({ theme }) => theme.__hd__.steps.sizes.iconSize};

  & > ${StyledIcon} {
    border-radius: 50%;
    background-color: ${({ theme }) => theme.__hd__.steps.colors.completeBg};
    color: ${({ theme }) => theme.__hd__.steps.colors.labelCompleteText};
  }
`;

const StepLine = styled.div<{ themeStatus: 'complete' | 'incomplete' }>`
  position: absolute;
  height: ${({ theme }) => theme.__hd__.steps.sizes.lineHeight};
  margin: 0;
  padding: 0;

  top: ${({ theme }) => theme.__hd__.steps.space.linePositionTop};
  left: ${({ theme }) => theme.__hd__.steps.space.linePositionLeft};
  right: ${({ theme }) => theme.__hd__.steps.space.linePositionRight};

  ${({ themeStatus, theme }) => {
    switch (themeStatus) {
      case 'complete':
        return css`
          background-color: ${theme.__hd__.steps.colors.lineCompleteBg};
        `;
      case 'incomplete':
        return css`
          background-color: ${theme.__hd__.steps.colors.lineIncompleteBg};
        `;
    }
  }};
`;

const ActiveIcon = styled.div<{
  themeStatus: 'complete' | 'incomplete';
}>`
  z-index: 1;
  position: absolute;
  box-sizing: content-box;

  display: flex;
  align-items: center;
  justify-content: center;

  width: ${({ theme }) => theme.__hd__.steps.sizes.activeIconSize};
  height: ${({ theme }) => theme.__hd__.steps.sizes.activeIconSize};

  border: ${({ theme }) => theme.__hd__.steps.borderWidths.activeIcon} solid
    ${({ theme }) => theme.__hd__.steps.colors.activeIconBorder};
  border-radius: ${({ theme }) => theme.__hd__.steps.radii.icon};
  color: ${({ theme }) => theme.__hd__.steps.colors.activeIconBorder};

  margin: 0;
  padding: 0;

  ${({ themeStatus, theme }) => {
    switch (themeStatus) {
      case 'complete':
        return css`
          background-color: ${theme.__hd__.steps.colors.activeCompleteBg};
          font-size: ${theme.__hd__.steps.fontSizes.activeIcon};
        `;

      case 'incomplete':
        return css`
          background-color: ${theme.__hd__.steps.colors.activeIncompleteBg};
        `;
    }
  }};
`;

const IncompleteIcon = styled.div`
  box-sizing: border-box;

  width: ${({ theme }) => theme.__hd__.steps.sizes.iconSize};
  height: ${({ theme }) => theme.__hd__.steps.sizes.iconSize};

  border: ${({ theme }) => theme.__hd__.steps.borderWidths.incompleteIcon} solid
    ${({ theme }) => theme.__hd__.steps.colors.incompleteIconBorder};
  border-radius: ${({ theme }) => theme.__hd__.steps.radii.icon};

  margin: 0;
  padding: 0;
`;

const StepLabel = styled.div<{
  themeActive: boolean;
  themeClickable: boolean;
  themeStatus: 'complete' | 'incomplete';
  themeVariant: 'basic' | 'compact';
}>`
  text-align: center;
  margin: ${({ theme }) => theme.__hd__.steps.space.labelMargin};
  padding: 0;

  ${({ themeClickable }) => {
    switch (themeClickable) {
      case true:
        return css`
          pointer-events: auto;
          cursor: pointer;
        `;
      case false:
        return css`
          pointer-events: none;
        `;
    }
  }};

  ${({ themeVariant }) =>
    themeVariant === 'compact' &&
    css`
      display: none;
    `};

  @media (max-width: ${({ theme }) => theme.breakpoints.small}) {
    display: none;
  }

  ${({ themeStatus, theme }) => {
    switch (themeStatus) {
      case 'complete':
        return css`
          color: ${theme.__hd__.steps.colors.labelCompleteText};
        `;
      case 'incomplete':
        return css`
          color: ${theme.__hd__.steps.colors.labelIncompleteText};
        `;
    }
  }};

  ${({ themeActive, theme }) => {
    switch (themeActive) {
      case true:
        return css`
          color: ${theme.__hd__.steps.colors.labelActiveText};
          font-size: ${theme.__hd__.steps.fontSizes.labelActive};
          font-weight: ${theme.__hd__.steps.fontWeights.labelActive};
          line-height: ${theme.__hd__.steps.lineHeights.labelActive};
        `;
      case false:
        return css`
          font-size: ${theme.__hd__.steps.fontSizes.labelInactive};
          line-height: ${theme.__hd__.steps.lineHeights.labelInactive};
        `;
    }
  }};
`;

export {
  Wrapper,
  StepWrapper,
  StepLine,
  StepLabel,
  StepIconWrapper,
  ActiveIcon,
  IncompleteIcon,
};
