// @hero-design/react/src/components/Select/SingleSelect/Styled.tsx
import styled, { css } from 'styled-components';
import type { Size } from '../../Input/utils';

export const InputContainer = styled.div<{
  themePaddingLeft: Size | 'none';
}>`
  display: inline-flex;
  flex-wrap: wrap;
  align-items: center;
  flex: 1;
  position: relative;
  overflow: hidden;
  margin: 0;

  ${({ themePaddingLeft, theme }) => {
    switch (themePaddingLeft) {
      case 'small':
        return css`
          padding: 0 0 0 ${theme.__hd__.input.space.smallPadding};
        `;
      case 'medium':
        return css`
          padding: 0 0 0 ${theme.__hd__.input.space.mediumPadding};
        `;
      case 'large':
        return css`
          padding: 0 0 0 ${theme.__hd__.input.space.largePadding};
        `;
      case 'none':
        return css`
          padding: 0;
        `;
    }
  }};
`;

export const SelectedItemText = styled.div<{
  themeState: 'enabled' | 'disabled';
}>`
  position: absolute;
  margin: 0;
  padding: 0;
  white-space: nowrap;

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'enabled':
        return css`
          color: ${theme.__hd__.input.colors.text};
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.input.colors.disabledText};
        `;
    }
  }};
`;

export const StyledHiddenInput = styled.input`
  position: fixed;
  width: 0;
  height: 0;
  opacity: 0;
  margin: 0;
  padding: 0;
`;
