// @hero-design/react/src/components/Filters/Group/StyledGroupFilter.tsx
import styled from 'styled-components';
import Box from '../../Box';
import { DropdownWrapper, DropdownPopper } from '../../Dropdown/StyledDropdown';

const GroupWrapper = styled(Box)`
  > ${DropdownWrapper} > ${DropdownPopper} {
    z-index: 9999;
  }
`;

const GroupFilterWrapper = styled(Box)``;

const DropdownContentWrapper = styled(Box)`
  padding: ${({ theme }) =>
    theme.__hd__.filters.space.filterGroupDropdownPadding};
  min-width: ${({ theme }) => theme.__hd__.filters.sizes.filterDropdownWidth};

  & > ${GroupFilterWrapper}:not(:last-child) {
    margin-bottom: ${({ theme }) =>
      theme.__hd__.filters.space.filterGroupFilterMarginBottom};
  }
`;

const WideDropdownContentWrapper = styled(Box)<{
  themeSize: 'large' | 'xlarge';
}>`
  display: flex;
  flex-wrap: wrap;
  gap: ${({ theme }) => theme.__hd__.filters.space.filterGroupDropdownItemGap};
  & > ${GroupFilterWrapper} {
    width: ${({ theme }) => theme.__hd__.filters.sizes.filterGroupWidth};
  }

  ${({ theme, themeSize }) => {
    switch (themeSize) {
      case 'large':
        return `
          width: ${theme.__hd__.filters.sizes.filterGroupLargeDropdownWidth};
        `;
      case 'xlarge':
        return `
          width: ${theme.__hd__.filters.sizes.filterGroupXLargeDropdownWidth};
        `;
    }
  }};
`;

export {
  GroupWrapper,
  DropdownContentWrapper,
  GroupFilterWrapper,
  WideDropdownContentWrapper,
};
