// @hero-design/react/src/components/Carousel/StyledCarousel.tsx
import styled, { css } from 'styled-components';

import { StyledIcon } from '../Icon/StyledIcon';

export const CarouselContainer = styled.div<{
  themeIsSingleSlideCarousel: boolean;
  themeDotPlacement: 'top' | 'right' | 'bottom' | 'left';
}>`
  position: relative;
  overflow: hidden;
  padding: 0;
  margin: 0;
  width: 100%;
  height: 100%;

  ${({ themeIsSingleSlideCarousel, themeDotPlacement, theme }) => {
    if (!themeIsSingleSlideCarousel) {
      switch (themeDotPlacement) {
        case 'top':
          return css`
            padding: ${theme.__hd__.carousel.space.containerHorizontalPadding}
              ${theme.__hd__.carousel.space.containerVerticalPadding} 0px;
          `;
        case 'bottom':
          return css`
            padding: 0px ${theme.__hd__.carousel.space.containerVerticalPadding}
              ${theme.__hd__.carousel.space.containerHorizontalPadding};
          `;
        default:
          return css``;
      }
    }
  }};
`;

export const CarouselWrapper = styled.div`
  overflow: hidden;
  padding: 0;
  margin: 0;
  width: 100%;
  height: 100%;
`;

export const StyledButtonWrapper = styled.ul<{
  themeDotDirection: 'row' | 'column';
  themeDotPlacement: 'top' | 'right' | 'bottom' | 'left';
  themeIsSingleSlideCarousel: boolean;
}>`
  position: absolute;
  z-index: 15;
  display: flex !important;
  justify-content: center;
  list-style: none;
  margin: 0;
  padding: 0;

  ${({ themeDotDirection }) => {
    switch (themeDotDirection) {
      case 'row':
        return css`
          flex-direction: row;
        `;
      case 'column':
        return css`
          flex-direction: column;
          transform: translateY(-50%);
        `;
    }
  }};

  ${({ themeDotPlacement, themeIsSingleSlideCarousel, theme }) => {
    switch (themeDotPlacement) {
      case 'top':
        return css`
          bottom: auto;
          top: ${themeIsSingleSlideCarousel
            ? theme.__hd__.carousel.space.default
            : 0};
          right: 0;
          left: 0;
        `;
      case 'right':
        return css`
          top: 50%;
          bottom: auto;
          left: auto;
          right: ${theme.__hd__.carousel.space.default};
        `;
      case 'bottom':
        return css`
          top: auto;
          bottom: ${themeIsSingleSlideCarousel
            ? theme.__hd__.carousel.space.default
            : 0};
          right: 0;
          left: 0;
        `;
      case 'left':
        return css`
          top: 50%;
          bottom: auto;
          left: ${theme.__hd__.carousel.space.default};
          right: auto;
        `;
    }
  }};
`;

export const StyledButtonList = styled.li<{
  themeDotState: 'activeRow' | 'defaultRow' | 'activeColumn' | 'defaultColumn';
}>`
  position: relative;
  display: inline-block;
  flex: 0 1 auto;
  box-sizing: content-box;
  padding: 0;
  text-align: center;
  vertical-align: top;
  transition: width 0.5s;

  ${({ themeDotState, theme }) => {
    switch (themeDotState) {
      case 'activeRow':
        return css`
          width: ${theme.__hd__.carousel.sizes.active};
          margin: ${theme.__hd__.carousel.space.listRowMargin};
        `;
      case 'defaultRow':
        return css`
          width: ${theme.__hd__.carousel.sizes.default};
          margin: ${theme.__hd__.carousel.space.listRowMargin};
        `;
      case 'activeColumn':
        return css`
          width: ${theme.__hd__.carousel.sizes.basic};
          height: ${theme.__hd__.carousel.sizes.active};
          margin: ${theme.__hd__.carousel.space.listColumnMargin};
        `;
      case 'defaultColumn':
        return css`
          width: ${theme.__hd__.carousel.sizes.basic};
          height: ${theme.__hd__.carousel.sizes.default};
          margin: ${theme.__hd__.carousel.space.listColumnMargin};
        `;
    }
  }};
`;

export const StyledButton = styled.button<{
  active: boolean;
  themeDotDirection: 'row' | 'column';
  themeVariant: 'light' | 'dark';
}>`
  display: block;
  padding: 0;
  margin: 0;
  outline: none;
  cursor: pointer;
  border: 0;
  border-radius: ${({ theme }) => theme.__hd__.carousel.radii.default};
  opacity: ${(props: { active: boolean }): number => (props.active ? 1 : 0.2)};

  ${({ themeDotDirection, theme }) => {
    switch (themeDotDirection) {
      case 'row':
        return css`
          height: ${theme.__hd__.carousel.sizes.basic};
          width: 100%;
        `;
      case 'column':
        return css`
          height: 100%;
          width: ${theme.__hd__.carousel.sizes.basic};
        `;
    }
  }};

  ${({ themeVariant, theme }) => {
    switch (themeVariant) {
      case 'light':
        return css`
          background: ${theme.__hd__.carousel.colors.light};
        `;
      case 'dark':
        return css`
          background: ${theme.__hd__.carousel.colors.dark};
        `;
    }
  }};
`;

export const CarouselSingleSlide = styled.div`
  flex: 0 0 auto;
  padding: 0;
  margin: 0;
  width: 100%;
  height: 100%;
`;

export const MultipleSlideItem = styled.div<{ themeSlideWidth: number }>`
  flex: 0 0 auto;
  padding: 0;
  margin: 0;
  height: 100%;
  width: ${({ themeSlideWidth }) => `${themeSlideWidth}%`};
  padding: ${({ theme }) => theme.__hd__.carousel.space.slidePadding};
`;

export const CarouselSlideCollection = styled.div<{
  currentSlide: number;
  themeDotDirection: 'row' | 'column';
  themeSlideWidth: number;
}>`
  display: flex;
  transition: all 0.5s ease-in-out;
  cursor: pointer;
  padding: 0;
  margin: 0;
  height: 100%;
  width: 100%;

  ${({ themeDotDirection }) => {
    switch (themeDotDirection) {
      case 'row':
        return css`
          flex-direction: row;
          ${(props: {
            currentSlide: number;
            themeSlideWidth: number;
          }): string =>
            `transform:translateX(-${
              props.currentSlide * props.themeSlideWidth
            }%)`};
        `;
      case 'column':
        return css`
          flex-direction: column;
          ${(props: { currentSlide: number }): string =>
            `transform:translateY(-${props.currentSlide * 100}%)`};
        `;
    }
  }};
`;

export const StyledButtonIconWrapper = styled.div<{
  themeButtonPlacement: 'middleRow' | 'leftColumn' | 'rightColumn';
  themeVariant: 'light' | 'dark';
  themeIsSingleSlideCarousel: boolean;
}>`
  position: absolute;
  z-index: 15;
  display: flex !important;
  justify-content: center;
  padding: 0;
  justify-content: space-between;

  ${({ themeButtonPlacement, themeIsSingleSlideCarousel, theme }) => {
    switch (themeButtonPlacement) {
      case 'middleRow':
        return css`
          flex-direction: row;
          transform: translate(0, -50%);
          width: ${themeIsSingleSlideCarousel ? '90%' : '100%'};
          margin: ${themeIsSingleSlideCarousel ? '0 auto' : '0'};
          top: 50%;
          left: 0;
          right: 0;
        `;
      case 'rightColumn':
        return css`
          flex-direction: column;
          height: 90%;
          margin: auto 0;
          right: ${theme.__hd__.carousel.space.navigationButtonColumnMargin};
          top: 0;
          bottom: 0;
        `;
      case 'leftColumn':
        return css`
          flex-direction: column;
          height: 90%;
          margin: auto 0;
          left: ${theme.__hd__.carousel.space.navigationButtonColumnMargin};
          top: 0;
          bottom: 0;
        `;
    }
  }};
  & ${StyledIcon} {
    font-size: ${({ theme }) =>
      theme.__hd__.carousel.fontSizes.navigationButton};

    ${({ themeVariant, theme }) => {
      switch (themeVariant) {
        case 'light':
          return css`
            color: ${theme.__hd__.carousel.colors.light};
          `;
        case 'dark':
          return css`
            color: ${theme.__hd__.carousel.colors.dark};
          `;
      }
    }};
  }
`;
