// @hero-design/react/src/components/DatePicker/Single/Month.tsx
import React, { ReactElement, useCallback, useMemo } from 'react';
import { IconName } from '../../Icon';
import { CommonProps } from '../../common';
import { generateMonths, MonthKey } from './util';
import { StyledSingleSelect } from './StyledSingle';

interface SingleMonth extends CommonProps {
  /**
   * Specify the [automated assistance](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) in filling out form field values by the browser.
   */
  autoComplete?: string;
  /**
   * Whether the picker is disabled.
   */
  disabled?: boolean;
  /**
   * Whether the input is invalid.
   */
  invalid?: boolean;
  /**
   * Name of <input> element, is used to refer to the form data for submission.
   */
  name?: string;
  /**
   * Blur event handler.
   */
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void;
  /**
   * onChange event handler.
   */
  onChange?: (value: number) => void;
  /**
   * Placeholder text in the absence of any value.
   */
  placeholder?: string;
  /**
   * Name of Icon or an Icon element to render on the left side of the input, before the user's cursor.
   */
  prefix?: IconName | ReactElement;
  /**
   * The size of the input box.
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * Current selected date which must be in correct format. If value is invalid, it will be skipped.
   */
  value?: number;
  /**
   * The earliest index of month user can select.
   * The value can be (0 - 11). It represent for the month number.
   */
  minMonth?: MonthKey;
  /**
   * The latest index of month user can select.
   * The value can be (0 - 11).It represent for the month number.
   */
  maxMonth?: MonthKey;
  /**
   * Month format. Following date-fns's format (https://date-fns.org/v2.16.1/docs/format).
   */
  format?: string;
}

const SingleMonth = ({
  autoComplete,
  className,
  disabled,
  invalid,
  id,
  name,
  onBlur,
  onChange,
  placeholder,
  prefix,
  size,
  style,
  sx,
  value,
  maxMonth = 11,
  minMonth = 0,
  format = 'LLLL',
  'data-test-id': dataTestId,
}: SingleMonth) => {
  const handleChange = useCallback(
    (_value) => {
      onChange?.(_value);
    },
    [onChange]
  );

  const options = useMemo(
    () => generateMonths(minMonth, maxMonth, format),
    [format, maxMonth, minMonth]
  );

  return (
    <StyledSingleSelect
      autoComplete={autoComplete}
      className={className}
      disabled={disabled}
      invalid={invalid}
      id={id}
      name={name}
      onBlur={onBlur}
      options={options}
      onChange={handleChange}
      placeholder={placeholder}
      prefix={prefix}
      size={size}
      style={style}
      sx={sx}
      value={value}
      data-test-id={dataTestId}
    />
  );
};

export default SingleMonth;
