// @hero-design/react/src/components/Dropdown/StyledDropdown.tsx
import styled from 'styled-components';

const DropdownWrapper = styled.div<{ themeDisplay: 'inline-block' | 'block' }>`
  position: relative;
  display: ${({ themeDisplay }) => themeDisplay};
  margin: 0;
  padding: 0;
`;

const DropdownPopper = styled.div<{ open: boolean }>`
  visibility: ${({ open }): string => (open === true ? 'visible' : 'hidden')};
  z-index: 99;
  margin: 0;
  padding: 0;
`;

const ContentWrapper = styled.div`
  box-shadow: ${({ theme }) => theme.__hd__.dropdown.shadows.default};
  border-radius: ${({ theme }) => theme.__hd__.dropdown.radii.default};
  overflow: hidden;
  background: ${({ theme }) => theme.__hd__.dropdown.colors.background};
  margin: 0;
  padding: 0;

  > *:first-child {
    border-top-right-radius: ${({ theme }) =>
      theme.__hd__.dropdown.radii.default};
    border-top-left-radius: ${({ theme }) =>
      theme.__hd__.dropdown.radii.default};
  }

  > *:last-child {
    border-bottom-right-radius: ${({ theme }) =>
      theme.__hd__.dropdown.radii.default};
    border-bottom-left-radius: ${({ theme }) =>
      theme.__hd__.dropdown.radii.default};
  }
`;

export { DropdownWrapper, DropdownPopper, ContentWrapper };
