// @hero-design/react/src/components/DatePicker/Calendar/DoubleCalendar.tsx
import React, { useState, useCallback, useEffect, useMemo } from 'react';
import getDate from 'date-fns/fp/getDate';
import getMonth from 'date-fns/fp/getMonth';
import getYear from 'date-fns/fp/getYear';
import addMonths from 'date-fns/fp/addMonths';
import addYears from 'date-fns/fp/addYears';
import subMonths from 'date-fns/fp/subMonths';
import subYears from 'date-fns/fp/subYears';
import setMonth from 'date-fns/fp/setMonth';
import setYear from 'date-fns/fp/setYear';

import type { ReactElement, MouseEvent } from 'react';
import Select from '../../Select';
import Divider from '../../Divider';
import Button from '../../Button';

import {
  Container,
  Navigation,
  DoubleCalendarContainer,
  CalendarContainer,
  CalendarWrapper,
  CalendarRow,
  DayWrapper,
  Day,
  MobileRightArrowsDateRangeWrapper,
  DesktopRightArrowsDateRangeWrapper,
} from './StyledCalendar';

import {
  getMonthMatrix,
  getDoubleCalendarDayState,
  generateMonthOptions,
  generateYearOptions,
  disableYear,
  disableMonthsNavigation,
  getWeekDays,
} from './utils';
import type { YearType } from './utils';

export interface DoubleCalendarProps {
  dateClickCount: number;
  endDate?: Date;
  isSettingEndDate: boolean;
  isSettingStartDate: boolean;
  maxDate?: Date;
  minDate?: Date;
  onSelectDate?: (date: Date) => void;
  startDate?: Date;
}

const DoubleCalendar = ({
  startDate,
  endDate,
  minDate,
  maxDate,
  onSelectDate,
  isSettingStartDate,
  isSettingEndDate,
  dateClickCount,
}: DoubleCalendarProps): ReactElement => {
  const currentDate = useMemo(() => new Date(), []);
  const initialDate = startDate !== undefined ? startDate : currentDate;
  const [date, setDate] = useState<Date>(initialDate);
  const [tempDate, setTempDate] = useState<Date>();
  const month = getMonth(date);
  const year = getYear(date);
  const dateInNextMonth = useMemo(() => addMonths(1, date), [date]);
  const nextMonth = getMonth(dateInNextMonth);
  const nextYear = getYear(dateInNextMonth);

  const calendarMinDate = useMemo(
    () => (minDate !== undefined ? minDate : subYears(100, currentDate)),
    [minDate, currentDate]
  );

  const calendarMaxDate = useMemo(
    () => (maxDate !== undefined ? maxDate : addYears(100, currentDate)),
    [maxDate, currentDate]
  );

  const initialYearOptions = useMemo(
    () =>
      generateYearOptions({
        minDate: calendarMinDate,
        maxDate: calendarMaxDate,
      }),
    [calendarMinDate, calendarMaxDate]
  );

  const [yearOptions, setYearOptions] =
    useState<YearType[]>(initialYearOptions);

  useEffect(() => {
    const minEdgeYear = initialYearOptions[0];
    const maxEdgeYear = initialYearOptions[initialYearOptions.length - 1];
    if (minEdgeYear === undefined || maxEdgeYear === undefined) return;

    if (getYear(date) < minEdgeYear.value) {
      setYearOptions([
        {
          text: (minEdgeYear.value - 1).toString(),
          value: minEdgeYear.value - 1,
        },
        ...initialYearOptions,
      ]);
    }

    if (getYear(dateInNextMonth) > maxEdgeYear.value) {
      setYearOptions([
        ...initialYearOptions,
        {
          text: (maxEdgeYear.value + 1).toString(),
          value: maxEdgeYear.value + 1,
        },
      ]);
    }

    if (
      getYear(date) >= minEdgeYear.value &&
      getYear(dateInNextMonth) <= maxEdgeYear.value
    ) {
      setYearOptions(initialYearOptions);
    }
  }, [date, dateInNextMonth, initialYearOptions]);

  const monthOptions = generateMonthOptions({
    minDate: calendarMinDate,
    maxDate: calendarMaxDate,
    year,
  });

  const nextMonthOptions = generateMonthOptions({
    minDate: calendarMinDate,
    maxDate: calendarMaxDate,
    year: nextYear,
  });

  const { disabledPreviousMonth, disabledNextMonth } = disableMonthsNavigation(
    date,
    calendarMinDate,
    calendarMaxDate
  );

  const disabledNextYear = disableYear(year + 1, yearOptions);
  const disabledPreviousYear = disableYear(year - 1, yearOptions);

  const onChangeMonth = useCallback(
    (newMonth) => {
      setDate(setMonth(newMonth, date));
    },
    [date, setDate]
  );

  const onChangeYear = useCallback(
    (newYear) => {
      setDate(setYear(newYear, date));
    },
    [date, setDate]
  );

  const onChangeNextMonth = useCallback(
    (newMonth) => {
      setDate(subMonths(1, setMonth(newMonth, dateInNextMonth)));
    },
    [dateInNextMonth, setDate]
  );

  const onChangeNextYear = useCallback(
    (newYear) => {
      setDate(subMonths(1, setYear(newYear, dateInNextMonth)));
    },
    [dateInNextMonth, setDate]
  );

  const onMouseOverDay = useCallback(
    (hoveredDate: Date): void => {
      setTempDate(hoveredDate);
    },
    [setTempDate]
  );

  const onDateClick = useCallback(
    (e: MouseEvent<HTMLSpanElement>, clickedDate: Date): void => {
      if (onSelectDate !== undefined) {
        onSelectDate(clickedDate);
      }
      e.preventDefault();
    },
    [onSelectDate]
  );

  const onNavigationClick = useCallback(
    ({ isMoveBack }: { isMoveBack: boolean }): void => {
      const newDate = isMoveBack ? subMonths(1, date) : addMonths(1, date);
      setDate(newDate);
    },
    [date, setDate]
  );

  const onDoubleNavigationClick = useCallback(
    ({ isMoveBack }: { isMoveBack: boolean }): void => {
      const newDate = isMoveBack ? subYears(1, date) : addYears(1, date);
      setDate(newDate);
    },
    [date, setDate]
  );

  const renderCalendar = useCallback(
    (monthMatrix: { date: Date; disabled: boolean }[][]): ReactElement => (
      <CalendarWrapper>
        <CalendarRow>
          {getWeekDays().map((day) => (
            <DayWrapper key={day}>
              <Day themeState="dayLabel">{day}</Day>
            </DayWrapper>
          ))}
        </CalendarRow>
        {monthMatrix.map((week, weekNumber) => (
          // eslint-disable-next-line react/no-array-index-key
          <CalendarRow key={weekNumber}>
            {week.map((dateOfWeek) => {
              const day = getDate(dateOfWeek.date);
              const dayState = getDoubleCalendarDayState({
                dateOfWeek,
                isSettingStartDate,
                isSettingEndDate,
                dateClickCount,
                startDate,
                endDate,
                tempDate,
              });
              return (
                // eslint-disable-next-line react/no-array-index-key
                <DayWrapper key={`${weekNumber}-${day}`} themeState={dayState}>
                  <Day
                    themeState={dayState}
                    onClick={
                      dayState !== 'disabled'
                        ? (e): void => onDateClick(e, dateOfWeek.date)
                        : undefined
                    }
                    onMouseOver={(): void => onMouseOverDay(dateOfWeek.date)}
                    onFocus={(): void => onMouseOverDay(dateOfWeek.date)}
                  >
                    {day}
                  </Day>
                </DayWrapper>
              );
            })}
          </CalendarRow>
        ))}
      </CalendarWrapper>
    ),
    [
      isSettingStartDate,
      isSettingEndDate,
      dateClickCount,
      startDate,
      endDate,
      tempDate,
      onDateClick,
      onMouseOverDay,
    ]
  );

  return (
    <Container>
      <DoubleCalendarContainer>
        <CalendarContainer>
          <Navigation>
            <Button.Icon
              size="small"
              icon="double-left-arrows"
              intent="primary"
              disabled={disabledPreviousYear}
              onClick={(): void =>
                onDoubleNavigationClick({ isMoveBack: true })
              }
              sx={{
                mr: 'small',
              }}
              data-test-id="sub-year-date-range-picker"
            />
            <Button.Icon
              size="small"
              icon="single-left-arrow"
              intent="primary"
              disabled={disabledPreviousMonth}
              onClick={(): void => onNavigationClick({ isMoveBack: true })}
              sx={{
                mr: 'small',
              }}
              data-test-id="sub-month-date-range-picker"
            />
            <Select
              options={monthOptions}
              value={month}
              onChange={onChangeMonth}
              data-test-id="first-month-select"
            />
            <Select
              options={yearOptions}
              value={year}
              onChange={onChangeYear}
              data-test-id="first-year-select"
            />
            <MobileRightArrowsDateRangeWrapper>
              <Button.Icon
                size="small"
                icon="single-right-arrow"
                intent="primary"
                disabled={disabledNextMonth}
                onClick={(): void => onNavigationClick({ isMoveBack: false })}
                sx={{
                  mr: 'small',
                  ml: 'small',
                }}
                data-test-id="add-month-date-range-picker-mobile"
              />
              <Button.Icon
                size="small"
                icon="double-right-arrows"
                intent="primary"
                disabled={disabledNextYear}
                onClick={(): void =>
                  onDoubleNavigationClick({ isMoveBack: false })
                }
                data-test-id="add-year-date-range-picker-mobile"
              />
            </MobileRightArrowsDateRangeWrapper>
          </Navigation>
          {renderCalendar(
            getMonthMatrix({
              month,
              year,
              minDate: calendarMinDate,
              maxDate: calendarMaxDate,
            })
          )}
        </CalendarContainer>
        <Divider marginX="small" />
        <CalendarContainer>
          <Navigation>
            <Select
              options={nextMonthOptions}
              value={nextMonth}
              onChange={onChangeNextMonth}
              data-test-id="second-month-select"
            />
            <Select
              options={yearOptions}
              value={nextYear}
              onChange={onChangeNextYear}
              data-test-id="second-year-select"
            />
            <DesktopRightArrowsDateRangeWrapper>
              <Button.Icon
                size="small"
                icon="single-right-arrow"
                intent="primary"
                disabled={disabledNextMonth}
                onClick={(): void => onNavigationClick({ isMoveBack: false })}
                sx={{
                  mr: 'small',
                  ml: 'small',
                }}
                data-test-id="add-month-date-range-picker-desktop"
              />
              <Button.Icon
                size="small"
                icon="double-right-arrows"
                intent="primary"
                disabled={disabledNextYear}
                onClick={(): void =>
                  onDoubleNavigationClick({ isMoveBack: false })
                }
                data-test-id="add-year-date-range-picker-desktop"
              />
            </DesktopRightArrowsDateRangeWrapper>
          </Navigation>
          {renderCalendar(
            getMonthMatrix({
              month: nextMonth,
              year: nextYear,
              minDate: calendarMinDate,
              maxDate: calendarMaxDate,
            })
          )}
        </CalendarContainer>
      </DoubleCalendarContainer>
    </Container>
  );
};

export default DoubleCalendar;
