// @hero-design/react/src/components/DatePicker/Calendar/StyledCalendar.tsx
import styled, { css } from 'styled-components';

import { SelectWrapper } from '../../Select/StyledSelect';
import Divider from '../../Divider/StyledDivider';

export type DayState =
  | 'selectedRangeDate'
  | 'selectedWeekDate'
  | 'selectedFirstDateOfWeek'
  | 'selectedLastDateOfWeek'
  | 'selectedStartDate'
  | 'selectedEndDate'
  | 'inRange'
  | 'inHoveringRange'
  | 'inHoveringWeekDate'
  | 'inHoveringFirstDateOfWeek'
  | 'inHoveringLastDateOfWeek'
  | 'enabled'
  | 'disabled'
  | 'selected'
  | 'today'
  | 'dayLabel';

export type WeekState = 'disabled' | 'enabled' | 'selected' | 'inHoveringWeek';

const Container = styled.div`
  width: fit-content;
  color: ${({ theme }) => theme.__hd__.datePicker.colors.text};
  margin: 0;
  padding: 0;
`;

const NavigationWrapper = styled.div`
  display: flex;
  justify-content: space-around;
  margin: 0;
  padding: 0;
`;

const MobileRightArrowsDateRangeWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const DesktopRightArrowsDateRangeWrapper = styled.div`
  margin: 0;
  padding: 0;
`;

const Navigation = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.datePicker.space.navigationPadding};

  & > ${SelectWrapper} {
    width: ${({ theme }) => theme.__hd__.datePicker.sizes.monthYear};
    padding: ${({ theme }) => theme.__hd__.datePicker.space.monthYearPadding};
  }
  & > ${MobileRightArrowsDateRangeWrapper} {
    @media (min-width: ${({ theme }) => theme.breakpoints.medium}) {
      display: none;
    }
  }
  & > ${DesktopRightArrowsDateRangeWrapper} {
    @media (max-width: ${({ theme }) => theme.breakpoints.medium}) {
      display: none;
    }
  }
`;

const DoubleCalendarContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin: 0;
  padding: 0;

  @media (min-width: ${({ theme }) => theme.breakpoints.medium}) {
    flex-direction: row;
  }

  & > ${Divider} {
    display: block;

    @media (min-width: ${({ theme }) => theme.breakpoints.medium}) {
      display: none;
    }
  }
`;

const CalendarContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin: 0;
  padding: 0;
`;

const CalendarWrapper = styled.div`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.datePicker.space.calendarPadding};
`;

const CalendarRow = styled.div`
  display: flex;
  justify-content: space-between;
  font-size: ${({ theme }) => theme.__hd__.datePicker.fontSizes.day};
  font-weight: ${({ theme }) => theme.__hd__.datePicker.fontWeights.day};
  line-height: ${({ theme }) => theme.__hd__.datePicker.lineHeights.day};
  margin: 0;
  padding: 0;
`;

const DayWrapper = styled.div<{
  themeState?: DayState;
}>`
  margin: ${({ theme }) => theme.__hd__.datePicker.space.dayMargin};
  padding: ${({ theme }) => theme.__hd__.datePicker.space.dayPadding};

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'selectedRangeDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-radius: ${theme.__hd__.datePicker.radii.day};
        `;
      case 'selectedWeekDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          &:hover {
            cursor: pointer;
          }
        `;
      case 'selectedFirstDateOfWeek':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-left-radius: ${theme.__hd__.datePicker.radii.day};
          border-bottom-left-radius: ${theme.__hd__.datePicker.radii.day};
          &:hover {
            cursor: pointer;
          }
        `;
      case 'selectedLastDateOfWeek':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-right-radius: ${theme.__hd__.datePicker.radii.day};
          border-bottom-right-radius: ${theme.__hd__.datePicker.radii.day};
          &:hover {
            cursor: pointer;
          }
        `;
      case 'selectedStartDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-left-radius: ${theme.__hd__.datePicker.radii.day};
          border-bottom-left-radius: ${theme.__hd__.datePicker.radii.day};
        `;
      case 'selectedEndDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-right-radius: ${theme.__hd__.datePicker.radii.day};
          border-bottom-right-radius: ${theme.__hd__.datePicker.radii.day};
        `;
      case 'inHoveringFirstDateOfWeek':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-top-left-radius: ${theme.__hd__.datePicker.radii.day};
          border-bottom-left-radius: ${theme.__hd__.datePicker.radii.day};
          &:hover {
            cursor: pointer;
          }
        `;
      case 'inHoveringLastDateOfWeek':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-top-right-radius: ${theme.__hd__.datePicker.radii.day};
          border-bottom-right-radius: ${theme.__hd__.datePicker.radii.day};
          &:hover {
            cursor: pointer;
          }
        `;
      case 'inRange':
      case 'inHoveringRange':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
        `;
      case 'inHoveringWeekDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          &:hover {
            cursor: pointer;
          }
        `;
      default:
        return css``;
    }
  }};
`;

const Day = styled.span<{
  themeState: DayState;
}>`
  display: inline-flex;
  align-items: center;
  justify-content: space-evenly;
  margin: 0;
  padding: 0;
  width: ${({ theme }) => theme.__hd__.datePicker.sizes.day};
  height: ${({ theme }) => theme.__hd__.datePicker.sizes.day};
  border-radius: ${({ theme }) => theme.__hd__.datePicker.radii.day};
  box-sizing: border-box;
  overflow: visible;
  white-space: nowrap;
  cursor: pointer;

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'enabled':
        return css`
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors.hoverBackground};
          }
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.datePicker.colors.disabledText};
          pointer-events: none;
          cursor: not-allowed;
        `;
      case 'selected':
      case 'selectedRangeDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
        `;
      case 'selectedStartDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-right-radius: 0;
          border-bottom-right-radius: 0;
        `;
      case 'selectedEndDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-left-radius: 0;
          border-bottom-left-radius: 0;
        `;
      case 'today':
        return css`
          color: ${theme.__hd__.datePicker.colors.todayText};
          border-width: ${theme.__hd__.datePicker.borderWidths.day};
          border-color: ${theme.__hd__.datePicker.colors.todayBorder};
          border-style: solid;
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors.hoverBackground};
          }
        `;
      case 'dayLabel':
        return css`
          cursor: default;
        `;
      case 'inRange':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-radius: 0;
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors
              .inRangeHoverBackground};
            color: ${theme.__hd__.datePicker.colors.inRangeHoverText};
          }
        `;
      case 'selectedFirstDateOfWeek':
      case 'selectedLastDateOfWeek':
      case 'selectedWeekDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-radius: 0;
        `;
      case 'inHoveringFirstDateOfWeek':
      case 'inHoveringLastDateOfWeek':
      case 'inHoveringWeekDate':
      case 'inHoveringRange':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-radius: 0;
        `;
    }
  }};
`;

const WeekWrapper = styled.div`
  margin: ${({ theme }) => theme.__hd__.datePicker.space.dayMargin};
  padding: ${({ theme }) => theme.__hd__.datePicker.space.dayPadding};
`;

const Week = styled.span<{
  themeState: WeekState;
}>`
  display: inline-flex;
  align-items: center;
  justify-content: space-evenly;
  margin: 0;
  padding: 0;
  width: ${({ theme }) => theme.__hd__.datePicker.sizes.day};
  height: ${({ theme }) => theme.__hd__.datePicker.sizes.day};
  border-radius: ${({ theme }) => theme.__hd__.datePicker.radii.day};
  box-sizing: border-box;
  cursor: pointer;

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'enabled':
        return css`
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors.hoverBackground};
          }
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.datePicker.colors.disabledText};
          pointer-events: none;
          cursor: not-allowed;
        `;
      case 'selected':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
        `;
      case 'inHoveringWeek':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          &:hover {
            cursor: pointer;
          }
        `;
    }
  }};
`;

const MonthNavigation = styled.div`
  margin: 0;
  padding: ${({ theme }) => theme.__hd__.datePicker.space.navigationPadding};
`;

type MonthState =
  | 'enabled'
  | 'disabled'
  | 'selectedRangeDate'
  | 'selectedStartDate'
  | 'selectedEndDate'
  | 'today'
  | 'inRange'
  | 'inHoveringRange'
  | 'outOfTheMonth'
  | 'selected';

const Month = styled.div<{
  themeState: MonthState;
}>`
  margin: 0;
  border-radius: ${({ theme }) => theme.__hd__.datePicker.radii.month};
  box-sizing: border-box;
  cursor: pointer;
  padding: ${({ theme }) => theme.__hd__.datePicker.space.monthPadding};
  font-size: ${({ theme }) => theme.__hd__.datePicker.fontSizes.month};
  font-weight: ${({ theme }) => theme.__hd__.datePicker.fontWeights.month};
  line-height: ${({ theme }) => theme.__hd__.datePicker.lineHeights.month};

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'enabled':
        return css`
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors.hoverBackground};
          }
        `;
      case 'disabled':
        return css`
          color: ${theme.__hd__.datePicker.colors.disabledText};
          pointer-events: none;
          cursor: not-allowed;
        `;
      case 'selected':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedMonth};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-width: ${theme.__hd__.datePicker.borderWidths.month};
          border-color: ${theme.__hd__.datePicker.colors.selectedBackground};
          border-style: solid;
        `;
      case 'today':
        return css`
          color: ${theme.__hd__.datePicker.colors.todayText};
          border-width: ${theme.__hd__.datePicker.borderWidths.month};
          border-color: ${theme.__hd__.datePicker.colors.todayBorder};
          border-style: solid;
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors.hoverBackground};
          }
        `;
      case 'selectedRangeDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
        `;
      case 'selectedStartDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-right-radius: 0px;
          border-bottom-right-radius: 0px;
        `;
      case 'selectedEndDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-left-radius: 0px;
          border-bottom-left-radius: 0px;
        `;
      case 'inRange':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-radius: 0;
          &:hover {
            background-color: ${theme.__hd__.datePicker.colors
              .inRangeHoverBackground};
            color: ${theme.__hd__.datePicker.colors.inRangeHoverText};
          }
        `;
      case 'inHoveringRange':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-radius: 0;
        `;
      case 'outOfTheMonth': {
        return css`
          color: ${theme.__hd__.datePicker.colors.disabledText};
          cursor: default;
        `;
      }
    }
  }};
`;

const MonthGrid = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 0px;
  margin: 0px;
  width: ${({ theme }) => theme.__hd__.datePicker.sizes.month};
`;

const MonthWrapper = styled.div<{
  themeState: MonthState;
}>`
  padding: ${({ theme }) => theme.__hd__.datePicker.space.monthSelectedPadding};
  margin: ${({ theme }) => theme.__hd__.datePicker.space.monthSelectedMargin};
  border-radius: ${({ theme }) => theme.__hd__.datePicker.radii.month};
  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'selectedRangeDate':
        return css`
          color: ${theme.__hd__.datePicker.colors.selectedText};
          font-weight: ${theme.__hd__.datePicker.fontWeights.selectedDay};
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
        `;
      case 'selectedStartDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-right-radius: 0px;
          border-bottom-right-radius: 0px;
        `;
      case 'selectedEndDate':
        return css`
          background-color: ${theme.__hd__.datePicker.colors
            .selectedBackground};
          border-top-left-radius: 0px;
          border-bottom-left-radius: 0px;
        `;
      case 'inRange':
      case 'inHoveringRange':
        return css`
          background-color: ${theme.__hd__.datePicker.colors.inRangeBackground};
          border-radius: 0;
        `;
      default: {
        return css``;
      }
    }
  }};
`;

const MonthGridItem = styled.div`
  text-align: center;
  margin: 0;
  padding: 0;
  width: 33.3%;
`;

export {
  Container,
  NavigationWrapper,
  Navigation,
  DoubleCalendarContainer,
  CalendarContainer,
  CalendarWrapper,
  CalendarRow,
  DayWrapper,
  Day,
  Week,
  WeekWrapper,
  Month,
  MonthNavigation,
  MonthGrid,
  MonthGridItem,
  DesktopRightArrowsDateRangeWrapper,
  MobileRightArrowsDateRangeWrapper,
  MonthWrapper,
};
