// @hero-design/react/src/components/Statistic/StyledStatistic.tsx
import styled, { css } from 'styled-components';

type StatisticIntent = 'text' | 'white' | 'inherit';

export const StatisticContainer = styled.div<{
  themeAlign: 'left' | 'right';
}>`
  padding: 0;
  margin: 0;
  width: 100%;
  ${({ themeAlign }) => {
    switch (themeAlign) {
      case 'left':
        return css`
          text-align: left;
        `;
      case 'right':
        return css`
          text-align: right;
        `;
    }
  }};
`;

export const StyledTitleWrapper = styled.div<{
  themeIntent: StatisticIntent;
}>`
  padding: 0;
  margin: ${({ theme }) => theme.__hd__.statistic.space.titleWrapperMargin};
  font-size: ${({ theme }) => theme.__hd__.statistic.fontSizes.title};
  line-height: ${({ theme }) => theme.__hd__.statistic.lineHeights.default};
  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'text':
        return css`
          color: ${theme.__hd__.statistic.colors.titleText};
        `;
      case 'white':
        return css`
          color: ${theme.__hd__.statistic.colors.titleWhite};
        `;
      case 'inherit':
        return css`
          color: inherit;
        `;
    }
  }};
`;

export const StyledContentWrapper = styled.div`
  padding: 0;
  margin: 0;
`;

export const StyledValueWrapper = styled.span<{
  themeIntent: StatisticIntent;
}>`
  padding: 0;
  margin: 0;
  font-size: ${({ theme }) => theme.__hd__.statistic.fontSizes.content};
  font-weight: ${({ theme }) => theme.__hd__.statistic.fontWeights.content};
  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'text':
        return css`
          color: ${theme.__hd__.statistic.colors.contentText};
        `;
      case 'white':
        return css`
          color: ${theme.__hd__.statistic.colors.contentWhite};
        `;
      case 'inherit':
        return css`
          color: inherit;
        `;
    }
  }};
`;

export const StyledPrefixWrapper = styled.span<{
  themeIntent: StatisticIntent;
}>`
  padding: 0;
  margin: ${({ theme }) => theme.__hd__.statistic.space.prefixWrapperMargin};

  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'text':
        return css`
          color: ${theme.__hd__.statistic.colors.contentText};
        `;
      case 'white':
        return css`
          color: ${theme.__hd__.statistic.colors.contentWhite};
        `;
      case 'inherit':
        return css`
          color: inherit;
        `;
    }
  }};
`;

export const StyledSuffixWrapper = styled.span<{
  themeIntent: StatisticIntent;
}>`
  padding: 0;
  margin: ${({ theme }) => theme.__hd__.statistic.space.suffixWrapperMargin};
  ${({ themeIntent, theme }) => {
    switch (themeIntent) {
      case 'text':
        return css`
          color: ${theme.__hd__.statistic.colors.contentText};
        `;
      case 'white':
        return css`
          color: ${theme.__hd__.statistic.colors.contentWhite};
        `;
      case 'inherit':
        return css`
          color: inherit;
        `;
    }
  }};
`;
