// @hero-design/react/src/components/Chart/LineChart/StyledLineChart.tsx
import styled from 'styled-components';
import { Arrow, TooltipWrapper } from '../../Tooltip/StyledTooltip';
import IconButton from '../../Button/IconButton';

const StyledTooltipContentWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  & > *:not(:last-child) {
    margin-right: ${({ theme }) =>
      `${theme.__hd__.lineChart.space.tooltipChildMargin}px`};
  }
`;

const StyledLineChartContainer = styled.div`
  position: relative;
  height: 100%;
  width: 100%;
  svg {
    position: relative;
    z-index: 2;
  }
  .x-axis {
    text {
      font-weight: ${({ theme }) =>
        theme.__hd__.lineChart.fontWeights.xAxisTick};
      font-size: ${({ theme }) => theme.__hd__.lineChart.fontSizes.xAxisTick};
      line-height: ${({ theme }) =>
        theme.__hd__.lineChart.lineHeights.xAxisTick};
      color: ${({ theme }) => theme.__hd__.lineChart.colors.xAxisTickColor};
    }
  }
  .y-axis {
    text {
      font-weight: ${({ theme }) =>
        theme.__hd__.lineChart.fontWeights.yAxisTick};
      font-size: ${({ theme }) => theme.__hd__.lineChart.fontSizes.yAxisTick};
      line-height: ${({ theme }) =>
        theme.__hd__.lineChart.lineHeights.yAxisTick};
      color: ${({ theme }) => theme.__hd__.lineChart.colors.yAxisTickColor};
    }
  }
`;

const StyledLegendContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  flex-grow: 0;
  margin-top: ${({ theme }) =>
    `${theme.__hd__.lineChart.space.legendContainerMarginTop}px`};
  > *:not(:last-child) {
    margin-bottom: ${({ theme }) =>
      `${theme.__hd__.lineChart.space.legendContainerChildMarginBottom}px`};
  }
  ${TooltipWrapper} {
    background: ${({ theme }) =>
      theme.__hd__.lineChart.colors.tooltipBackgroundColor};
    box-shadow: ${({ theme }) => theme.__hd__.lineChart.boxShadows.tooltip};
    padding: ${({ theme }) => theme.__hd__.lineChart.space.moreTooltipPadding};
  }

  ${Arrow} {
    &::before {
      background: ${({ theme }) =>
        theme.__hd__.lineChart.colors.tooltipBackgroundColor};
    }
  }
`;

const StyledNavigateButton = styled(IconButton)`
  width: ${({ theme }) => `${theme.__hd__.lineChart.sizes.navigationButton}px`};
  height: ${({ theme }) =>
    `${theme.__hd__.lineChart.sizes.navigationButton}px`};
`;

const StyledNavigateButtonWrapper = styled.div`
  display: flex;
  flex-grow: 0;
  place-items: center;
  padding: ${({ theme }) =>
    `${theme.__hd__.lineChart.space.navigateButtonPadding}px`};
`;

const StyledNavigationButtonsWrapper = styled.div`
  display: flex;
  flex: 1;
  min-width: 0;
`;
const StyledNavigationButtonsChildrenWrapper = styled.div`
  display: flex;
  flex: 1;
  min-width: 0;
`;

const StyledYAxisTitleWrapper = styled.div`
  width: ${({ theme }) => `${theme.__hd__.lineChart.sizes.yAxisTitleWidth}px`};
`;

const StyledLineChartChildrenWrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  height: 100%;
  min-width: 0;
`;

const StyledYAxisTitleAndNavigationButtonsAndChartContentWrapper = styled.div`
  display: flex;
  flex: 1;
  min-width: 0;
`;

export {
  StyledTooltipContentWrapper,
  StyledLineChartContainer,
  StyledLegendContainer,
  StyledNavigateButton,
  StyledNavigateButtonWrapper,
  StyledNavigationButtonsWrapper,
  StyledNavigationButtonsChildrenWrapper,
  StyledYAxisTitleWrapper,
  StyledLineChartChildrenWrapper,
  StyledYAxisTitleAndNavigationButtonsAndChartContentWrapper,
};
