// @hero-design/react/src/components/Chart/StyledChart.tsx
import styled from 'styled-components';
import { TooltipWrapper } from '../Tooltip/StyledTooltip';
import Typography from '../Typography';

const StyledTooltip = styled(TooltipWrapper)`
  position: absolute;
  transform: translate(-50%, -100%);
  white-space: nowrap;
  &:after {
    content: ' ';
    position: absolute;
    top: 100%;
    left: 50%;
    transform: translateX(-50%);
    border-width: ${({ theme }) => theme.__hd__.chart.sizes.tooltip.arrowSize};
    border-style: solid;
    border-color: ${({ theme }) =>
      theme.__hd__.chart.colors.tooltip.foreground};
    border-top-color: ${({ theme }) =>
      theme.__hd__.chart.colors.tooltip.background};
  }
`;

const StyledTitleContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: ${({ theme }) => theme.__hd__.chart.spaces.titleContainerMargin};
`;

const StyledChartTitle = styled(Typography.Text)``;

const StyledChartSubTitle = styled(Typography.Text)`
  margin: ${({ theme }) =>
    theme.__hd__.chart.spaces.multiplePie.subTitleMargin};
`;

export {
  StyledTooltip,
  StyledChartTitle,
  StyledChartSubTitle,
  StyledTitleContainer,
};
