// @hero-design/react/src/components/DatePicker/index.tsx
import type { FunctionComponent } from 'react';
import DatePicker from './DatePicker';
import DateRangePicker from './DateRangePicker';
import MonthPicker from './MonthPicker';
import WeekPicker from './WeekPicker';
import MonthRangePicker from './MonthRangePicker';

import type { DatePickerProps } from './DatePicker';
import FortnightlyPicker from './FortnightlyPicker';
import SingleMonth from './Single/Month';
import SingleYear from './Single/Year';

interface CompoundDatePickerProps extends FunctionComponent<DatePickerProps> {
  Month: typeof MonthPicker;
  Range: typeof DateRangePicker;
  Week: typeof WeekPicker;
  MonthRange: typeof MonthRangePicker;
  Fortnightly: typeof FortnightlyPicker;
  SingleMonth: typeof SingleMonth;
  SingleYear: typeof SingleYear;
}

const CompoundDatePicker = DatePicker as CompoundDatePickerProps;

CompoundDatePicker.Month = MonthPicker;
CompoundDatePicker.Week = WeekPicker;
CompoundDatePicker.Range = DateRangePicker;
CompoundDatePicker.MonthRange = MonthRangePicker;
CompoundDatePicker.Fortnightly = FortnightlyPicker;
CompoundDatePicker.SingleMonth = SingleMonth;
CompoundDatePicker.SingleYear = SingleYear;

export default CompoundDatePicker;
