// @hero-design/react/src/components/DatePicker/StyledDatePicker.tsx
import styled, { css } from 'styled-components';

import { InputWrapper } from '../Input/StyledInput';
import { DropdownWrapper } from '../Dropdown/StyledDropdown';
import { DateRangePickerState } from './utils';

const DatePickerContainer = styled.div`
  margin: 0;
  padding: 0;

  &:focus {
    outline: none;
  }

  & > ${DropdownWrapper} {
    display: block;
  }
`;

const DateRangeInputWrapper = styled(DatePickerContainer)<{
  themeState: DateRangePickerState;
}>`
  display: flex;
  position: relative;
  align-items: center;
  border: ${({ theme }) => theme.__hd__.datePicker.borderWidths.range} solid
    ${({ theme }) => theme.__hd__.datePicker.colors.rangeBorder};
  border-radius: ${({ theme }) => theme.__hd__.datePicker.radii.range};

  & > ${InputWrapper} {
    flex: 1 1;
    width: 0;
    border: none;
    box-shadow: none;
  }

  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'invalid':
        return css`
          border-color: ${theme.__hd__.datePicker.colors.invalidBorder};
          &:focus-within {
            outline: none;
            border-color: ${theme.__hd__.datePicker.colors.invalidBorder};
            box-shadow: ${theme.__hd__.datePicker.shadows.invalidFocus};
          }
        `;
      case 'enabled':
        return css`
          &:focus-within {
            border-color: ${theme.__hd__.datePicker.colors.rangeFocusBorder};
            box-shadow: ${theme.__hd__.datePicker.shadows.range};
          }
        `;
    }
  }};
`;

const DateSeparator = styled.span`
  margin: 0;
  padding: 0;
  color: ${({ theme }) => theme.__hd__.datePicker.colors.text};
`;

const FocusBar = styled.div<{
  themeState: DateRangePickerState;
}>`
  position: absolute;
  margin: 0;
  padding: 0;
  bottom: 0;
  height: ${({ theme }) => theme.__hd__.datePicker.sizes.rangeFocusBar};
  ${({ themeState, theme }) => {
    switch (themeState) {
      case 'invalid':
        return css`
          background: ${theme.__hd__.datePicker.colors.invalidRangeFocusBar};
        `;
      case 'enabled':
        return css`
          background: ${theme.__hd__.datePicker.colors.rangeFocusBar};
        `;
    }
  }};
  transition: width 0.3s, left 0.3s, right 0.3s;
`;

export { DatePickerContainer, DateRangeInputWrapper, DateSeparator, FocusBar };
