// @hero-design/react/src/components/Card/index.tsx
import React, { useMemo } from 'react';
import type { AriaAttributes, ReactElement, ReactNode } from 'react';
import { useCss } from '../../utils/hooks';

import CardContext from './CardContext';
import CardContent from './CardContent';
import CardHeader from './CardHeader';
import { StyledCard, StyledCardImg } from './StyledCard';
import { pipe } from '../../fp/function';
import { fromUndefinedable, getOrElse, map } from '../../fp/Option';
import type { CommonProps } from '../common';

export interface CardProps extends CommonProps, AriaAttributes {
  /**
   * Display border.
   */
  border?: boolean;
  /**
   * Custom content.
   */
  children?: ReactNode;
  /**
   * Short hand for Card content.
   */
  content?: ReactNode;
  /**
   * Short hand for Card content with format as extra content.
   */
  extra?: ReactNode;
  /**
   * Short hand for Card header.
   */
  header?: ReactNode;
  /**
   * Card image.
   */
  imageSrc?: string;
  /**
   * Card size. This will affect the padding of card's header and content.
   */
  size?: 'small' | 'medium';
  /**
   * Card variant.
   */
  variant?: 'basic' | 'primary' | 'info' | 'grey';
}

const Card = ({
  header,
  content,
  extra,
  variant = 'basic',
  size = 'medium',
  imageSrc,
  border = true,
  children,
  id,
  className,
  style,
  sx = {},
  'data-test-id': dataTestId,
  ...ariaProps
}: CardProps): ReactElement => {
  const cardVariant = border === true ? 'withBorder' : 'withoutBorder';
  const css = useCss(sx);

  const contextValue = useMemo(
    () => ({
      variant,
      size,
    }),
    [variant, size]
  );

  const wrapWithStyledCard = (cardInner: ReactNode): ReactElement => (
    <CardContext.Provider value={contextValue}>
      <StyledCard
        themeStatus={cardVariant}
        id={id}
        className={className}
        style={{ ...style, ...css }}
        data-test-id={dataTestId}
        {...ariaProps}
      >
        {cardInner}
      </StyledCard>
    </CardContext.Provider>
  );

  return pipe(
    children,
    fromUndefinedable,
    map(wrapWithStyledCard),
    getOrElse(() =>
      wrapWithStyledCard(
        <>
          {imageSrc !== undefined && (
            <StyledCardImg src={imageSrc} alt="card-img" />
          )}
          {header !== undefined && <CardHeader content={header} />}
          {content !== undefined && <CardContent content={content} />}
          {extra !== undefined && <CardContent content={extra} extra />}
        </>
      )
    )
  );
};

Card.Header = CardHeader;
Card.Content = CardContent;

export default Card;
