// @hero-design/react/src/components/Card/StyledCard.tsx
import styled, { css } from 'styled-components';

const StyledCard = styled.div<{ themeStatus: 'withBorder' | 'withoutBorder' }>`
  display: inline-block;
  box-sizing: border-box;
  width: 100%;
  background: ${({ theme }) => theme.__hd__.card.colors.background};
  border-radius: ${({ theme }) => theme.__hd__.card.radii.default};
  margin: 0;
  padding: 0;

  > *:first-child {
    border-top-left-radius: ${({ theme }) => theme.__hd__.card.radii.default};
    border-top-right-radius: ${({ theme }) => theme.__hd__.card.radii.default};
  }

  > *:last-child {
    border-bottom-left-radius: ${({ theme }) =>
      theme.__hd__.card.radii.default};
    border-bottom-right-radius: ${({ theme }) =>
      theme.__hd__.card.radii.default};
  }

  ${({ themeStatus, theme }) => {
    switch (themeStatus) {
      case 'withBorder':
        return css`
          border-width: ${theme.__hd__.card.borderWidths.default};
          border-style: solid;
          border-color: ${theme.__hd__.card.colors.border};
        `;
      case 'withoutBorder':
        return css`
          border-width: ${theme.__hd__.card.borderWidths.default};
          border-style: solid;
          border-color: transparent;
        `;
    }
  }};
`;

const StyledCardHeader = styled.div<{
  themeSize: 'small' | 'medium';
  themeVariant: 'basic' | 'primary' | 'info' | 'grey';
}>`
  box-sizing: border-box;
  position: relative;
  font-size: ${({ theme }) => theme.__hd__.card.fontSizes.header};
  font-weight: ${({ theme }) => theme.__hd__.card.fontWeights.header};
  line-height: ${({ theme }) => theme.__hd__.card.lineHeights.header};
  border-bottom: ${({ theme }) => theme.__hd__.card.borderWidths.default} solid;
  margin: 0;

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          padding: ${theme.__hd__.card.space.headerSmallPadding};
        `;
      case 'medium':
        return css`
          padding: ${theme.__hd__.card.space.headerMediumPadding};
        `;
    }
  }};

  ${({ themeVariant, theme }) => {
    switch (themeVariant) {
      case 'basic':
        return css`
          background: ${theme.__hd__.card.colors.headerBasicBackground};
          color: ${theme.__hd__.card.colors.headerDefaultText};
          border-bottom-color: ${theme.__hd__.card.colors.border};
        `;
      case 'info':
        return css`
          background: ${theme.__hd__.card.colors.headerInfoBackground};
          color: ${theme.__hd__.card.colors.headerInfoText};
          margin: ${theme.__hd__.card.space.headerMargin};
          border-bottom-color: ${theme.__hd__.card.colors.headerInfoBackground};
        `;
      case 'grey':
        return css`
          background: ${theme.__hd__.card.colors.headerGreyBackground};
          color: ${theme.__hd__.card.colors.headerDefaultText};
          margin: ${theme.__hd__.card.space.headerMargin};
          border-bottom-color: ${theme.__hd__.card.colors.headerGreyBackground};
        `;
      case 'primary':
        return css`
          background: ${theme.__hd__.card.colors.headerPrimaryBackground};
          color: ${theme.__hd__.card.colors.headerLightText};
          margin: ${theme.__hd__.card.space.headerMargin};
          border-bottom-color: ${theme.__hd__.card.colors
            .headerPrimaryBackground};
        `;
    }
  }};
`;

const StyledCardContent = styled.div<{
  themeSize: 'small' | 'medium';
  themeVariant: 'mainContent' | 'extra';
}>`
  font-weight: ${({ theme }) => theme.__hd__.card.fontWeights.content};
  box-sizing: border-box;
  margin: 0;

  ${({ themeVariant, theme }) => {
    switch (themeVariant) {
      case 'mainContent':
        return css`
          padding: ${theme.__hd__.card.space.headerSmallPadding};
          line-height: ${theme.__hd__.card.lineHeights.content};
          color: ${theme.__hd__.card.colors.contentText};
          font-size: ${theme.__hd__.card.fontSizes.content};
        `;
      case 'extra':
        return css`
          padding: ${theme.__hd__.card.space.headerMediumPadding};
          line-height: ${theme.__hd__.card.lineHeights.contentExtra};
          color: ${theme.__hd__.card.colors.contentTextExtra};
          font-size: ${theme.__hd__.card.fontSizes.contentExtra};
          border-top-width: ${theme.__hd__.card.borderWidths.default};
          border-top-style: solid;
          border-top-color: ${theme.__hd__.card.colors.border};
        `;
    }
  }};

  ${({ themeSize, theme }) => {
    switch (themeSize) {
      case 'small':
        return css`
          padding: ${theme.__hd__.card.space.contentSmallPadding};
        `;
      case 'medium':
        return css`
          padding: ${theme.__hd__.card.space.contentMediumPadding};
        `;
    }
  }};
`;

const StyledCardImg = styled.img`
  display: block;
  width: calc(
    100% + 2 * ${({ theme }) => theme.__hd__.card.borderWidths.default}
  );
  max-width: calc(
    100% + 2 * ${({ theme }) => theme.__hd__.card.borderWidths.default}
  );
  margin: ${({ theme }) => theme.__hd__.card.space.imgMargin};
  padding: 0;
`;

export { StyledCard, StyledCardHeader, StyledCardContent, StyledCardImg };
